{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "34cf0daf-8ab6-4571-a6d8-ef271c8cf2c7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n",
      "ArchR : Version 1.0.2\n",
      "For more information see our website : www.ArchRProject.com\n",
      "If you encounter a bug please report : https://github.com/GreenleafLab/ArchR/issues\n",
      "\n",
      "Loading Required Packages...\n",
      "\n",
      "\tLoading Package : grid v4.1.3\n",
      "\n",
      "\tLoading Package : gridExtra v2.3\n",
      "\n",
      "\tLoading Package : gtools v3.9.4\n",
      "\n",
      "\tLoading Package : gtable v0.3.3\n",
      "\n",
      "\tLoading Package : ggplot2 v3.4.2\n",
      "\n",
      "\tLoading Package : magrittr v2.0.3\n",
      "\n",
      "\tLoading Package : plyr v1.8.8\n",
      "\n",
      "\tLoading Package : stringr v1.5.0\n",
      "\n",
      "\tLoading Package : data.table v1.14.8\n",
      "\n",
      "\tLoading Package : matrixStats v1.0.0\n",
      "\n",
      "\tLoading Package : S4Vectors v0.32.4\n",
      "\n",
      "\tLoading Package : GenomicRanges v1.46.1\n",
      "\n",
      "\tLoading Package : BiocGenerics v0.40.0\n",
      "\n",
      "\tLoading Package : Matrix v1.6.0\n",
      "\n",
      "\tLoading Package : Rcpp v1.0.11\n",
      "\n",
      "\tLoading Package : SummarizedExperiment v1.24.0\n",
      "\n",
      "\tLoading Package : rhdf5 v2.38.1\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:plyr’:\n",
      "\n",
      "    arrange, count, desc, failwith, id, mutate, rename, summarise,\n",
      "    summarize\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:gridExtra’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in library(ggpubr): there is no package called ‘ggpubr’\n",
     "output_type": "error",
     "traceback": [
      "Error in library(ggpubr): there is no package called ‘ggpubr’\nTraceback:\n",
      "1. library(ggpubr)"
     ]
    }
   ],
   "source": [
    "library(ArchR)\n",
    "library(stringr)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(glue)\n",
    "library(parallel)\n",
    "library(ggpubr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9087321e-95d1-4609-a249-1585e08fbc74",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 20.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "addArchRThreads(threads = 20) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec992951-646d-4017-8582-0c8779104b16",
   "metadata": {},
   "source": [
    "# Figure 1f: ATAC tracks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66c9a11b-5417-467d-aa7d-5a29b35443a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "                                                   / |\n",
      "                                                 /    \\\n",
      "            .                                  /      |.\n",
      "            \\\\\\                              /        |.\n",
      "              \\\\\\                          /           `|.\n",
      "                \\\\\\                      /              |.\n",
      "                  \\                    /                |\\\n",
      "                  \\\\#####\\           /                  ||\n",
      "                ==###########>      /                   ||\n",
      "                 \\\\##==......\\    /                     ||\n",
      "            ______ =       =|__ /__                     ||      \\\\\\\n",
      "        ,--' ,----`-,__ ___/'  --,-`-===================##========>\n",
      "       \\               '        ##_______ _____ ,--,__,=##,__   ///\n",
      "        ,    __==    ___,-,__,--'#'  ==='      `-'    | ##,-/\n",
      "        -,____,---'       \\\\####\\\\________________,--\\\\_##,/\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "proj <- loadArchRProject('/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/ATAC/data/ATAC_FINAL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9eb4c468-1e15-4f26-bca6-f3c859c689cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 2 × 55</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AllClasses.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationGenome.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationPeaks.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-CRHR1.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-FKBP5.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser3.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-CAMKV.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-KCNIP4.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-xiangyu.R</th><th scope=col>⋯</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/MultiModal.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/plot_footprints.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ProjectMethods.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/QualityControl.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RcppExports.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ReproduciblePeakSet.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RNAIntegration.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/Trajectory.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ValidationUtils.R</th><th scope=col>/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/VisualizeData.R</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;fn&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>value</th><td>rownames</td><td><span style=white-space:pre-wrap>function (genome = NULL) , {,     .validInput(input = genome, name = \"genome\", valid = \"character\"),     if (toupper(genome) == \"HG19\" | toupper(genome) == \"HG38\") {,         if (suppressWarnings(!require(org.Hs.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Hs.eg.db\", update = FALSE),         },         library(org.Hs.eg.db),         annodb &lt;- org.Hs.eg.db,     },     else if (toupper(genome) == \"MM9\" | toupper(genome) == \"MM10\") {,         if (suppressWarnings(!require(org.Mm.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Mm.eg.db\", update = FALSE),         },         library(org.Mm.eg.db),         annodb &lt;- org.Mm.eg.db,     },     else {,         stop(\"Genome not recognized!\"),     },     return(annodb), }</span></td><td><span style=white-space:pre-wrap>function (seEnrich = NULL, pal = paletteContinuous(set = \"comet\", ,     n = 100), n = 10, cutOff = 20, pMax = Inf, clusterCols = TRUE, ,     binaryClusterRows = TRUE, labelRows = TRUE, rastr = TRUE, ,     transpose = FALSE, returnMatrix = FALSE, logFile = createLogFile(\"plotEnrichHeatmap\")) , {,     .validInput(input = seEnrich, name = \"seEnrich\", valid = c(\"SummarizedExperiment\")),     .validInput(input = pal, name = \"pal\", valid = c(\"character\")),     .validInput(input = n, name = \"n\", valid = c(\"integer\")),     .validInput(input = cutOff, name = \"cutOff\", valid = c(\"numeric\")),     .validInput(input = pMax, name = \"pMax\", valid = c(\"numeric\")),     .validInput(input = clusterCols, name = \"clusterCols\", valid = c(\"boolean\")),     .validInput(input = binaryClusterRows, name = \"binaryClusterRows\", ,         valid = c(\"boolean\")),     .validInput(input = labelRows, name = \"labelRows\", valid = c(\"boolean\")),     .validInput(input = rastr, name = \"rastr\", valid = c(\"boolean\")),     .validInput(input = transpose, name = \"transpose\", valid = c(\"boolean\")),     .validInput(input = returnMatrix, name = \"returnMatrix\", ,         valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotEnrichHeatmap Input-Parameters\", logFile = logFile),     mat &lt;- assays(seEnrich)[[\"mlog10Padj\"]],     .logThis(mat, \"mat-mlog10Padj\", logFile = logFile),     keep &lt;- lapply(seq_len(ncol(mat)), function(x) {,         idx &lt;- head(order(mat[, x], decreasing = TRUE), n),         rownames(mat)[idx[which(mat[idx, x] &gt; cutOff)]],     }) %&gt;% unlist %&gt;% unique,     mat &lt;- mat[keep, , drop = FALSE],     .logThis(mat, \"mat-mlog10Padj-Filter\", logFile = logFile),     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     passMat &lt;- lapply(seq_len(nrow(mat)), function(x) {,         (mat[x, ] &gt;= 0.9 * max(mat[x, ])) * 1,     }) %&gt;% Reduce(\"rbind\", .) %&gt;% data.frame,     colnames(passMat) &lt;- colnames(mat),     .logThis(passMat, \"passMat\", logFile = logFile),     mat[mat &gt; pMax] &lt;- pMax,     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     mat &lt;- .rowScale(as.matrix(mat), min = 0),     if (pMax != 100) {,         rownames(mat[[1]]) &lt;- paste0(rownames(mat[[1]]), \" (\", ,             round(mat$max), \")\"),         rownames(passMat) &lt;- rownames(mat[[1]]),     },     .logThis(mat, \"mat-mlog10Padj-RowScale\", logFile = logFile),     mat2 &lt;- mat[[1]] * 100,     if (nrow(mat2) &gt; 1 &amp; ncol(mat2) &gt; 1) {,         if (binaryClusterRows) {,             bS &lt;- .binarySort(mat2, lmat = passMat[rownames(mat2), ,                 colnames(mat2)], clusterCols = clusterCols),             mat2 &lt;- bS[[1]][, colnames(mat2)],             clusterRows &lt;- FALSE,             clusterCols &lt;- bS[[2]],         },         else {,             clusterRows &lt;- TRUE,         },     },     else {,         clusterCols &lt;- NULL,         clusterRows &lt;- FALSE,     },     if (nrow(mat2) &gt; 100) {,         borderColor &lt;- FALSE,     },     else {,         borderColor &lt;- TRUE,     },     .logThis(mat2, \"mat-mlog10Padj-RowScale-Plot\", logFile = logFile),     if (transpose) {,         if (!is.null(clusterCols)) {,             mat2 &lt;- t(mat2[, clusterCols$order, drop = FALSE]),         },         else {,             mat2 &lt;- t(mat2),         },         if (returnMatrix) {,             return(mat2),         },         ht &lt;- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     else {,         if (returnMatrix) {,             return(mat2),         },         ht &lt;- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     return(ht), }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td><span style=white-space:pre-wrap>function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }</span></td><td>⋯</td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL, name = \"CombinedDims\", reducedDims = NULL, ,     dimWeights = NULL, dimsToUse = NULL, scaleDims = NULL, corCutOff = 0.75) , {,     if (is.null(dimWeights)) {,         dimWeights &lt;- rep(1, length(reducedDims)),     },     combinedDims &lt;- lapply(seq_along(reducedDims), function(x) {,         rD &lt;- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims[x], ,             dimsToUse = dimsToUse, scaleDims = scaleDims, corCutOff = corCutOff),         cV &lt;- apply(as.matrix(rD), 2, var),         normV &lt;- 1/sqrt(sum(cV)),         rD * normV * dimWeights[x],     }) %&gt;% Reduce(\"cbind\", .),     ArchRProj@reducedDims[[name]] &lt;- SimpleList(matRD = combinedDims, ,         scaleDims = NA, corToDepth = list(scaled = rep(0, ncol(combinedDims)), ,             none = rep(0, ncol(combinedDims)))),     ArchRProj, }</span></td><td><span style=white-space:pre-wrap>function (seFoot = NULL, names = NULL, pal = NULL, flank = 250, ,     flankNorm = 50, normMethod = \"Subtract\", smoothWindow = NULL, ,     baseSize = 6, plot = TRUE, ArchRProj = NULL, plotName = paste0(\"Plot-Footprints-\", ,         normMethod), height = 6, width = 4, addDOC = TRUE, force = FALSE, ,     logFile = createLogFile(\"plotFootprints\")) , {,     tstart &lt;- Sys.time(),     if (is.null(names)) {,         names &lt;- names(assays(seFoot)),     },     if (plot) {,         name &lt;- gsub(\"\\\\.pdf\", \"\", plotName),         if (is.null(ArchRProj)) {,             outDir &lt;- \"Plots\",         },         else {,             outDir &lt;- file.path(getOutputDirectory(ArchRProj), ,                 \"Plots\"),         },         dir.create(outDir, showWarnings = FALSE),         if (addDOC) {,             doc &lt;- gsub(\":\", \"-\", stringr::str_split(Sys.time(), ,                 pattern = \" \", simplify = TRUE)[1, 2]),             filename &lt;- file.path(outDir, paste0(name, \"_Date-\", ,                 Sys.Date(), \"_Time-\", doc, \".pdf\")),         },         else {,             filename &lt;- file.path(outDir, paste0(name, \".pdf\")),         },         pdf(filename, width = width, height = height, useDingbats = FALSE),     },     ggList &lt;- lapply(seq_along(names), function(x) {,         gg &lt;- .ggFootprint(seFoot = seFoot, name = names[x], ,             pal = pal, smoothWindow = smoothWindow, flank = flank, ,             flankNorm = flankNorm, baseSize = baseSize, normMethod = normMethod, ,             logFile = logFile),         if (plot) {,             if (x != 1) {,                 grid::grid.newpage(),             },             grid::grid.draw(gg),             return(0),         },         else {,             return(gg),         },     }),     if (!plot) {,         names(ggList) &lt;- names,         ggList,     },     else {,         dev.off(),         return(invisible(0)),     }, }</span></td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL, features = NULL, name = NULL, addRatio = TRUE, ,     threads = getArchRThreads(), logFile = createLogFile(\"addFeatureCounts\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = features, name = \"features\", valid = c(\"granges\")),     .validInput(input = name, name = \"name\", valid = c(\"character\")),     .validInput(input = addRatio, name = \"addRatio\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     ArrowFiles &lt;- getArrowFiles(ArchRProj),     cellNames &lt;- ArchRProj$cellNames,     featuresList &lt;- split(features, seqnames(features)),     h5disableFileLocking(),     countsDF &lt;- .safelapply(seq_along(featuresList), function(i) {,         chri &lt;- names(featuresList)[i],         cellTotal &lt;- rep(0, length(cellNames)),         names(cellTotal) &lt;- cellNames,         featuresi &lt;- ranges(featuresList[[i]]),         .logDiffTime(paste0(\"Counting in \", chri, \" (\", i, \" of \", ,             length(featuresList), \")\"), t1 = tstart, logFile = logFile),         for (j in seq_along(ArrowFiles)) {,             fragmentsij &lt;- .getFragsFromArrow(ArrowFile = ArrowFiles[j], ,                 chr = chri, out = \"IRanges\", cellNames = cellNames),             if (length(fragmentsij) &gt; 0) {,                 mcols(fragmentsij)$RG@values &lt;- match(mcols(fragmentsij)$RG@values, ,                   cellNames),                 for (y in seq_len(2)) {,                   if (y == 1) {,                     temp &lt;- IRanges(start(fragmentsij), width = 1),                   },                   else if (y == 2) {,                     temp &lt;- IRanges(end(fragmentsij), width = 1),                   },                   stopifnot(length(temp) == length(fragmentsij)),                   tabSum &lt;- S4Vectors:::tabulate(mcols(fragmentsij)$RG[queryHits(findOverlaps(temp, ,                     featuresi))]),                   cellTotal[seq_along(tabSum)] &lt;- cellTotal[seq_along(tabSum)] + ,                     tabSum,                 },             },             if (j%%3 == 0) {,                 gc(),             },         },         cellTotal,     }, threads = threads) %&gt;% Reduce(\"rbind\", .),     totalCounts &lt;- colSums(countsDF),     countRatio &lt;- totalCounts/(2 * ArchRProj$nFrags),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Counts\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj, data = totalCounts, ,         cells = names(totalCounts), name = paste0(name, \"Counts\"), ,         force = TRUE),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Ratio\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj, data = countRatio, ,         cells = names(totalCounts), name = paste0(name, \"Ratio\"), ,         force = TRUE),     ArchRProj, }</span></td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL, groupBy = \"Sample\", chromSizes = getChromSizes(ArchRProj), ,     maxSize = 750, pal = NULL, returnDF = FALSE, threads = getArchRThreads(), ,     logFile = createLogFile(\"plotFragmentSizes\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = maxSize, name = \"maxSize\", valid = c(\"integer\")),     .validInput(input = returnDF, name = \"returnDF\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotFragmentSizes Input-Parameters\", logFile = logFile),     chr &lt;- paste0(seqnames(chromSizes)),     groups &lt;- getCellColData(ArchRProj = ArchRProj, select = groupBy, ,         drop = FALSE),     uniqGroups &lt;- gtools::mixedsort(unique(groups[, 1])),     if (threads &gt; 1) {,         h5disableFileLocking(),     },     dfFS &lt;- .safelapply(seq_along(uniqGroups), function(x) {,         .logDiffTime(paste0(uniqGroups[x], \" Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         cellx &lt;- rownames(groups)[which(paste0(groups[, 1]) == ,             uniqGroups[x])],         for (i in seq_along(chr)) {,             if (i == 1) {,                 fsi &lt;- unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %&gt;% width %&gt;% tabulate(nbins = maxSize),             },             else {,                 fsi &lt;- fsi + unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %&gt;% width %&gt;% tabulate(nbins = maxSize),             },             .logThis(fsi, paste0(uniqGroups[x], \" : FragSizes : \", ,                 chr[i]), logFile = logFile),         },         df &lt;- DataFrame(group = uniqGroups[x], fragmentSize = seq_along(fsi), ,             fragmentPercent = round(100 * fsi/sum(fsi), 4)),         .logThis(df, paste0(uniqGroups[x], \" : Frag DF\"), logFile = logFile),         .logDiffTime(paste0(uniqGroups[x], \" Finished Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         df,     }, threads = threads) %&gt;% Reduce(\"rbind\", .),     .logThis(dfFS, paste0(\"All : FragSizes DF\"), logFile = logFile),     .endLogging(logFile = logFile),     if (threads &gt; 1) {,         h5enableFileLocking(),     },     if (returnDF) {,         return(dfFS),     },     else {,         plotDF &lt;- data.frame(dfFS),         if (is.null(pal)) {,             pal &lt;- paletteDiscrete(values = unique(plotDF$group)),         },         p &lt;- ggplot(plotDF, aes(fragmentSize, fragmentPercent, ,             color = group)) + geom_line(size = 1) + theme_ArchR() + ,             xlab(\"ATAC-seq Fragment Size (bp)\") + ylab(\"Percentage of Fragments\") + ,             scale_color_manual(values = pal) + scale_y_continuous(limits = c(0, ,             max(plotDF$fragmentPercent) * 1.05), expand = c(0, ,             0)) + scale_x_continuous(limits = c(min(plotDF$fragmentSize), ,             max(plotDF$fragmentSize)), expand = c(0, 0)),         p,     }, }</span></td><td><span style=white-space:pre-wrap>function (string_vector, id_vector, n_id, window, w, kmer) , {,     .Call(\"_ArchR_kmerIDFrequencyCpp\", PACKAGE = \"ArchR\", string_vector, ,         id_vector, n_id, window, w, kmer), }</span></td><td><span style=white-space:pre-wrap>function () , {,     message(\"Searching For MACS2..\"),     if (.suppressAll(.checkPath(\"macs2\", throwError = FALSE))) {,         message(\"Found with $path!\"),         return(\"macs2\"),     },     message(\"Not Found in $PATH\"),     search2 &lt;- suppressWarnings(tryCatch({,         system2(\"pip\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     search3 &lt;- suppressWarnings(tryCatch({,         system2(\"pip3\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     if (length(search2) &gt; 0) {,         if (search2[1] != \"ERROR\") {,             path2Install &lt;- gsub(\"Location: \", \"\", search2[grep(\"Location\", ,                 search2, ignore.case = TRUE)]),             path2Bin &lt;- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip\"),     if (length(search3) &gt; 0) {,         if (search3[1] != \"ERROR\") {,             path2Install &lt;- gsub(\"Location: \", \"\", search3[grep(\"Location\", ,                 search3, ignore.case = TRUE)]),             path2Bin &lt;- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip3!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip3\"),     stop(\"Could Not Find Macs2! Please install w/ pip, add to your $PATH, or just supply the macs2 path directly and avoid this function!\"), }</span></td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL, useMatrix = \"GeneScoreMatrix\", matrixName = \"GeneIntegrationMatrix\", ,     reducedDims = \"IterativeLSI\", seRNA = NULL, groupATAC = NULL, ,     groupRNA = NULL, groupList = NULL, sampleCellsATAC = 10000, ,     sampleCellsRNA = 10000, embeddingATAC = NULL, embeddingRNA = NULL, ,     dimsToUse = 1:30, scaleDims = NULL, corCutOff = 0.75, plotUMAP = TRUE, ,     UMAPParams = list(n_neighbors = 40, min_dist = 0.4, metric = \"cosine\", ,         verbose = FALSE), nGenes = 2000, useImputation = TRUE, ,     reduction = \"cca\", addToArrow = TRUE, scaleTo = 10000, genesUse = NULL, ,     nameCell = \"predictedCell\", nameGroup = \"predictedGroup\", ,     nameScore = \"predictedScore\", transferParams = list(), threads = getArchRThreads(), ,     verbose = TRUE, force = FALSE, logFile = createLogFile(\"addGeneIntegrationMatrix\"), ,     ...) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = useMatrix, name = \"useMatrix\", valid = c(\"character\")),     .validInput(input = matrixName, name = \"matrixName\", valid = c(\"character\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = seRNA, name = \"seRNA\", valid = c(\"SummarizedExperiment\", ,         \"Seurat\")),     .validInput(input = groupATAC, name = \"groupATAC\", valid = c(\"character\", ,         \"null\")),     .validInput(input = groupRNA, name = \"groupRNA\", valid = c(\"character\")),     .validInput(input = groupList, name = \"groupList\", valid = c(\"list\", ,         \"null\")),     .validInput(input = sampleCellsATAC, name = \"sampleCellsATAC\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = sampleCellsRNA, name = \"sampleCellsRNA\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = embeddingATAC, name = \"embeddingATAC\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = embeddingRNA, name = \"embeddingRNA\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = dimsToUse, name = \"dimsToUse\", valid = c(\"numeric\", ,         \"null\")),     .validInput(input = scaleDims, name = \"scaleDims\", valid = c(\"boolean\", ,         \"null\")),     .validInput(input = plotUMAP, name = \"plotUMAP\", valid = c(\"boolean\")),     .validInput(input = UMAPParams, name = \"UMAPParams\", valid = c(\"list\")),     .validInput(input = nGenes, name = \"nGenes\", valid = c(\"integer\")),     .validInput(input = useImputation, name = \"useImputation\", ,         valid = c(\"boolean\")),     .validInput(input = reduction, name = \"reduction\", valid = c(\"character\")),     .validInput(input = addToArrow, name = \"addToArrow\", valid = c(\"boolean\")),     .validInput(input = scaleTo, name = \"scaleTo\", valid = c(\"numeric\")),     .validInput(input = genesUse, name = \"genesUse\", valid = c(\"character\", ,         \"null\")),     .validInput(input = nameCell, name = \"nameCell\", valid = c(\"character\")),     .validInput(input = nameGroup, name = \"nameGroup\", valid = c(\"character\")),     .validInput(input = nameScore, name = \"nameScore\", valid = c(\"character\")),     .validInput(input = transferParams, name = \"transferParams\", ,         valid = c(\"list\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = verbose, name = \"verbose\", valid = c(\"boolean\")),     .validInput(input = force, name = \"force\", valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logDiffTime(\"Running Seurat's Integration Stuart* et al 2019\", ,         tstart, verbose = verbose, logFile = logFile),     .requirePackage(\"Seurat\", source = \"cran\"),     .logThis(append(args, mget(names(formals()), sys.frame(sys.nframe()))), ,         \"Input-Parameters\", logFile = logFile),     if (is.null(groupList)) {,         groupList &lt;- SimpleList(),         groupList[[1]] &lt;- SimpleList(ATAC = ArchRProj$cellNames, ,             RNA = colnames(seRNA)),     },     .logDiffTime(\"Checking ATAC Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (useMatrix %ni% getAvailableMatrices(ArchRProj)) {,         .logMessage(paste0(\"Matrix \", useMatrix, \" does not exist in the provided ArchRProject. See available matrix names from getAvailableMatrices()!\"), ,             logFile = logFile),         stop(\"Matrix name provided to useMatrix does not exist in ArchRProject!\"),     },     if (!is.null(groupATAC)) {,         dfATAC &lt;- getCellColData(ArchRProj = ArchRProj, select = groupATAC, ,             drop = FALSE),     },     nCell &lt;- rep(0, length(ArchRProj$cellNames)),     names(nCell) &lt;- ArchRProj$cellNames,     groupList &lt;- lapply(seq_along(groupList), function(x) {,         ATAC &lt;- groupList[[x]]$ATAC,         if (!is.null(groupATAC)) {,             if (any(ATAC %in% dfATAC[, 1])) {,                 idx &lt;- which(ATAC %in% dfATAC[, 1]),                 ATAC2 &lt;- rownames(dfATAC)[which(dfATAC[, 1] %in% ,                   ATAC[idx])],                 if (length(idx) == length(ATAC)) {,                   ATAC &lt;- ATAC2,                 },                 else {,                   ATAC &lt;- c(ATAC[-idx], ATAC2),                 },             },         },         SimpleList(ATAC = ATAC, RNA = groupList[[x]]$RNA),     }) %&gt;% SimpleList,     for (i in seq_along(groupList)) {,         nCell[groupList[[i]]$ATAC] &lt;- nCell[groupList[[i]]$ATAC] + ,             1,     },     if (!all(nCell == 1)) {,         .logMessage(paste0(\"Missing \", length(which(nCell == ,             0)), \" cells. Found \", length(which(nCell &gt; 1)), ,             \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"), ,             logFile = logFile),         stop(\"Missing \", length(which(nCell == 0)), \" cells. Found \", ,             length(which(nCell &gt; 1)), \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"),     },     .logDiffTime(\"Checking RNA Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (inherits(seRNA, \"SummarizedExperiment\")) {,         seuratRNA &lt;- CreateSeuratObject(counts = assay(seRNA)),         if (groupRNA %ni% colnames(colData(seRNA))) {,             .logMessage(\"groupRNA not in colData of seRNA\", logFile = logFile),             stop(\"groupRNA not in colData of seRNA\"),         },         seuratRNA$Group &lt;- paste0(colData(seRNA)[, groupRNA, ,             drop = TRUE]),         rm(seRNA),     },     else {,         if (groupRNA %ni% colnames(seRNA@meta.data)) {,             .logMessage(\"groupRNA not in meta.data of Seurat Object\", ,                 logFile = logFile),             stop(\"groupRNA not in meta.data of Seurat Object\"),         },         seuratRNA &lt;- seRNA,         seuratRNA$Group &lt;- paste0(seRNA@meta.data[, groupRNA]),         rm(seRNA),     },     if (\"RNA\" %in% names(seuratRNA@assays)) {,         DefaultAssay(seuratRNA) &lt;- \"RNA\",     },     else {,         stop(\"'RNA' is not present in Seurat Object's Assays! Please make sure that this assay is present!\"),     },     gc(),     if (!is.null(groupRNA)) {,         dfRNA &lt;- DataFrame(row.names = colnames(seuratRNA), Group = seuratRNA$Group),     },     groupList &lt;- lapply(seq_along(groupList), function(x) {,         RNA &lt;- groupList[[x]]$RNA,         if (!is.null(groupRNA)) {,             if (any(RNA %in% dfRNA[, 1])) {,                 idx &lt;- which(RNA %in% dfRNA[, 1]),                 RNA2 &lt;- rownames(dfRNA)[which(dfRNA[, 1] %in% ,                   RNA[idx])],                 if (length(idx) == length(RNA)) {,                   RNA &lt;- RNA2,                 },                 else {,                   RNA &lt;- c(RNA[-idx], RNA2),                 },             },         },         SimpleList(ATAC = groupList[[x]]$ATAC, RNA = RNA),     }) %&gt;% SimpleList,     cellRNA &lt;- unlist(lapply(groupList, function(x) x$RNA)),     if (!all(cellRNA %in% colnames(seuratRNA))) {,         .logMessage(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\", ,             logFile = logFile),         stop(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\"),     },     seuratRNA &lt;- seuratRNA[, unique(cellRNA)],     seuratRNA &lt;- NormalizeData(object = seuratRNA, verbose = FALSE),     geneDF &lt;- .getFeatureDF(getArrowFiles(ArchRProj), useMatrix),     sumOverlap &lt;- sum(unique(geneDF$name) %in% unique(rownames(seuratRNA))),     if (sumOverlap &lt; 5) {,         stop(\"Error not enough overlaps (\", sumOverlap, \") between gene names from gene scores (ArchR) and rna matrix (seRNA)!\"),     },     .logDiffTime(paste0(\"Found \", sumOverlap, \" overlapping gene names from gene scores and rna matrix!\"), ,         tstart, verbose = TRUE, logFile = logFile),     .logDiffTime(\"Creating Integration Blocks\", tstart, verbose = verbose, ,         logFile = logFile),     blockList &lt;- SimpleList(),     for (i in seq_along(groupList)) {,         gLi &lt;- groupList[[i]],         if (length(gLi$ATAC) &gt; sampleCellsATAC) {,             if (!is.null(embeddingATAC)) {,                 probATAC &lt;- .getDensity(embeddingATAC[gLi$ATAC, ,                   1], embeddingATAC[gLi$ATAC, 2])$density,                 probATAC &lt;- probATAC/max(probATAC),                 cellsATAC &lt;- gLi$ATAC[order(probATAC, decreasing = TRUE)],             },             else {,                 cellsATAC &lt;- sample(gLi$ATAC, length(gLi$ATAC)),             },             cutoffs &lt;- lapply(seq_len(1000), function(x) length(gLi$ATAC)/x) %&gt;% ,                 unlist,             blockSize &lt;- ceiling(min(cutoffs[order(abs(cutoffs - ,                 sampleCellsATAC))[1]] + 1, length(gLi$ATAC))),             nBlocks &lt;- ceiling(length(gLi$ATAC)/blockSize),             blocks &lt;- lapply(seq_len(nBlocks), function(x) {,                 cellsATAC[seq(x, length(cellsATAC), nBlocks)],             }) %&gt;% SimpleList,         },         else {,             blocks &lt;- list(gLi$ATAC),         },         if (!is.null(embeddingRNA)) {,             probRNA &lt;- .getDensity(embeddingRNA[gLi$RNA, 1], ,                 embeddingRNA[gLi$RNA, 2])$density,             probRNA &lt;- probRNA/max(probRNA),         },         else {,             probRNA &lt;- rep(1, length(gLi$RNA)),         },         blockListi &lt;- lapply(seq_along(blocks), function(x) {,             SimpleList(ATAC = blocks[[x]], RNA = sample(x = gLi$RNA, ,                 size = min(sampleCellsRNA, length(gLi$RNA)), ,                 prob = probRNA)),         }) %&gt;% SimpleList,         blockList &lt;- c(blockList, blockListi),     },     rm(groupList),     .logDiffTime(\"Prepping Interation Data\", tstart, verbose = verbose, ,         logFile = logFile),     subProj &lt;- ArchRProj,     subProj@imputeWeights &lt;- SimpleList(),     geneDF &lt;- .getFeatureDF(getArrowFiles(subProj), useMatrix),     geneDF &lt;- geneDF[geneDF$name %in% rownames(seuratRNA), , ,         drop = FALSE],     splitGeneDF &lt;- S4Vectors::split(geneDF, geneDF$seqnames),     featureDF &lt;- lapply(splitGeneDF, function(x) {,         x$idx &lt;- seq_len(nrow(x)),         return(x),     }) %&gt;% Reduce(\"rbind\", .),     dfParams &lt;- data.frame(reduction = reduction),     allChr &lt;- unique(featureDF$seqnames),     tmpFile &lt;- .tempfile(),     o &lt;- suppressWarnings(file.remove(paste0(tmpFile, \"-IntegrationBlock-\", ,         seq_along(blockList), \".h5\"))),     if (threads &gt; 1) {,         h5disableFileLocking(),     },     rD &lt;- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims, ,         corCutOff = corCutOff, dimsToUse = dimsToUse),     outDir1 &lt;- getOutputDirectory(ArchRProj),     outDir2 &lt;- file.path(outDir1, \"RNAIntegration\"),     outDir3 &lt;- file.path(outDir2, matrixName),     dir.create(outDir1, showWarnings = FALSE),     dir.create(outDir2, showWarnings = FALSE),     dir.create(outDir3, showWarnings = FALSE),     prevFiles &lt;- list.files(outDir3, full.names = TRUE),     prevFiles &lt;- .suppressAll(file.remove(prevFiles)),     tstart &lt;- Sys.time(),     threads2 &lt;- max(ceiling(threads * 0.75), 1),     .logDiffTime(paste0(\"Computing Integration in \", length(blockList), ,         \" Integration Blocks!\"), tstart, verbose = verbose, logFile = logFile),     dfAll &lt;- .safelapply(seq_along(blockList), function(i) {,         prefix &lt;- sprintf(\"Block (%s of %s) :\", i, length(blockList)),         .logDiffTime(sprintf(\"%s Computing Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         blocki &lt;- blockList[[i]],         subProj@cellColData &lt;- subProj@cellColData[blocki$ATAC, ,             ],         subProj@sampleColData &lt;- subProj@sampleColData[unique(subProj$Sample), ,             , drop = FALSE],         subRNA &lt;- seuratRNA[, blocki$RNA],         subRNA &lt;- subRNA[rownames(subRNA) %in% geneDF$name, ],         .logDiffTime(sprintf(\"%s Identifying Variable Genes\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         subRNA &lt;- FindVariableFeatures(object = subRNA, nfeatures = nGenes, ,             verbose = FALSE),         subRNA &lt;- ScaleData(object = subRNA, verbose = FALSE),         if (is.null(genesUse)) {,             genesUse &lt;- VariableFeatures(object = subRNA),         },         .logDiffTime(sprintf(\"%s Getting GeneScoreMatrix\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         mat &lt;- .getPartialMatrix(getArrowFiles(subProj), featureDF = geneDF[geneDF$name %in% ,             genesUse, ], threads = 1, cellNames = subProj$cellNames, ,             useMatrix = useMatrix, verbose = FALSE),         rownames(mat) &lt;- geneDF[geneDF$name %in% genesUse, \"name\"],         .logThis(mat, paste0(\"GeneScoreMat-Block-\", i), logFile = logFile),         if (useImputation) {,             .logDiffTime(sprintf(\"%s Imputing GeneScoreMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             imputeParams &lt;- list(),             imputeParams$ArchRProj &lt;- subProj,             imputeParams$randomSuffix &lt;- TRUE,             imputeParams$reducedDims &lt;- reducedDims,             imputeParams$dimsToUse &lt;- dimsToUse,             imputeParams$scaleDims &lt;- scaleDims,             imputeParams$corCutOff &lt;- corCutOff,             imputeParams$threads &lt;- 1,             imputeParams$logFile &lt;- logFile,             subProj &lt;- suppressMessages(do.call(addImputeWeights, ,                 imputeParams)),             mat &lt;- suppressMessages(imputeMatrix(mat = mat, imputeWeights = getImputeWeights(subProj), ,                 verbose = FALSE, logFile = logFile)),             o &lt;- suppressWarnings(file.remove(unlist(getImputeWeights(subProj)[[1]]))),             .logThis(mat, paste0(\"GeneScoreMat-Block-Impute-\", ,                 i), logFile = logFile),         },         mat &lt;- log(mat + 1),         seuratATAC &lt;- Seurat::CreateSeuratObject(counts = mat[head(seq_len(nrow(mat)), ,             5), , drop = FALSE]),         seuratATAC[[\"GeneScore\"]] &lt;- Seurat::CreateAssayObject(counts = mat),         rm(mat),         DefaultAssay(seuratATAC) &lt;- \"GeneScore\",         seuratATAC &lt;- Seurat::ScaleData(seuratATAC, verbose = FALSE),         .logDiffTime(sprintf(\"%s Seurat FindTransferAnchors\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferAnchors &lt;- .retryCatch({,             gc(),             Seurat::FindTransferAnchors(reference = subRNA, query = seuratATAC, ,                 reduction = reduction, features = genesUse, verbose = FALSE, ,                 ...),         }, maxAttempts = 2, logFile = logFile),         .logThis(paste0(utils::capture.output(transferAnchors), ,             collapse = \"\\n\"), paste0(\"transferAnchors-\", i), ,             logFile = logFile),         rDSub &lt;- rD[colnames(seuratATAC), , drop = FALSE],         .logThis(rDSub, paste0(\"rDSub-\", i), logFile = logFile),         transferParams$anchorset &lt;- transferAnchors,         transferParams$weight.reduction &lt;- CreateDimReducObject(embeddings = rDSub, ,             key = \"LSI_\", assay = DefaultAssay(seuratATAC)),         transferParams$verbose &lt;- FALSE,         transferParams$dims &lt;- seq_len(ncol(rDSub)),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Group Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata &lt;- subRNA$Group,         rnaLabels &lt;- do.call(Seurat::TransferData, transferParams),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Names Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata &lt;- colnames(subRNA),         rnaLabels2 &lt;- do.call(Seurat::TransferData, transferParams)[, ,             1],         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Seurat TransferData GeneMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             transferParams$refdata &lt;- GetAssayData(subRNA, assay = \"RNA\", ,                 slot = \"data\"),             gc(),             matchedRNA &lt;- do.call(Seurat::TransferData, transferParams),             matchedRNA &lt;- matchedRNA@data,         },         matchDF &lt;- DataFrame(cellNames = colnames(seuratATAC), ,             predictionScore = rnaLabels$prediction.score.max, ,             predictedGroup = rnaLabels$predicted.id, predictedCell = rnaLabels2),         rownames(matchDF) &lt;- matchDF$cellNames,         .logDiffTime(sprintf(\"%s Saving TransferAnchors Joint CCA\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         jointCCA &lt;- DataFrame(transferAnchors@object.list[[1]]@reductions$cca@cell.embeddings),         jointCCA$Assay &lt;- ifelse(endsWith(rownames(jointCCA), ,             \"_reference\"), \"RNA\", \"ATAC\"),         jointCCA$Group &lt;- NA,         jointCCA$Score &lt;- NA,         jointCCA[paste0(colnames(subRNA), \"_reference\"), \"Group\"] &lt;- subRNA$Group,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Group\"] &lt;- matchDF$predictedGroup,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Score\"] &lt;- matchDF$predictionScore,         .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,             paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),         rm(transferParams, transferAnchors),         gc(),         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Transferring Paired RNA to Temp File\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             tmpFilei &lt;- paste0(tmpFile, \"-IntegrationBlock-\", ,                 i, \".h5\"),             o &lt;- h5createFile(tmpFilei),             sampleNames &lt;- getCellColData(subProj, \"Sample\")[matchDF$cellNames, ,                 ],             uniqueSamples &lt;- unique(sampleNames),             matchedRNA &lt;- .safeSubset(mat = matchedRNA, subsetRows = paste0(featureDF$name), ,                 subsetCols = matchDF$cellNames),             for (z in seq_along(uniqueSamples)) {,                 mat &lt;- matchedRNA[, which(sampleNames == uniqueSamples[z]), ,                   drop = FALSE],                 Group &lt;- uniqueSamples[z],                 o &lt;- tryCatch({,                   h5delete(tmpFilei, paste0(Group)),                 }, error = function(x) {,                 }),                 o &lt;- h5createGroup(tmpFilei, paste0(Group)),                 j &lt;- Rle(findInterval(seq(mat@x) - 1, mat@p[-1]) + ,                   1),                 lengthRle &lt;- length(j@lengths),                 lengthI &lt;- length(mat@i),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/i\"), storage.mode = \"integer\", dims = c(lengthI, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jLengths\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jValues\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/x\"), storage.mode = \"double\", dims = c(lengthI, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5write(obj = mat@i + 1, file = tmpFilei, ,                   name = paste0(Group, \"/i\"))),                 o &lt;- .suppressAll(h5write(obj = j@lengths, file = tmpFilei, ,                   name = paste0(Group, \"/jLengths\"))),                 o &lt;- .suppressAll(h5write(obj = j@values, file = tmpFilei, ,                   name = paste0(Group, \"/jValues\"))),                 o &lt;- .suppressAll(h5write(obj = mat@x, file = tmpFilei, ,                   name = paste0(Group, \"/x\"))),                 o &lt;- .suppressAll(h5write(obj = colnames(mat), ,                   file = tmpFilei, name = paste0(Group, \"/cellNames\"))),             },             rm(matchedRNA, mat, j),         },         .logDiffTime(sprintf(\"%s Completed Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         gc(),         matchDF$Block &lt;- Rle(i),         matchDF,     }, threads = threads2) %&gt;% Reduce(\"rbind\", .),     if (plotUMAP) {,         for (i in seq_along(blockList)) {,             o &lt;- tryCatch({,                 prefix &lt;- sprintf(\"Block (%s of %s) :\", i, length(blockList)),                 .logDiffTime(sprintf(\"%s Plotting Joint UMAP\", ,                   prefix), tstart, verbose = verbose, logFile = logFile),                 jointCCA &lt;- readRDS(file.path(outDir3, paste0(\"Save-Block\", ,                   i, \"-JointCCA.rds\"))),                 set.seed(1),                 UMAPParams &lt;- .mergeParams(UMAPParams, list(n_neighbors = 40, ,                   min_dist = 0.4, metric = \"cosine\", verbose = FALSE)),                 UMAPParams$X &lt;- as.data.frame(jointCCA[, grep(\"CC_\", ,                   colnames(jointCCA))]),                 UMAPParams$ret_nn &lt;- FALSE,                 UMAPParams$ret_model &lt;- FALSE,                 UMAPParams$n_threads &lt;- 1,                 uwotUmap &lt;- tryCatch({,                   do.call(uwot::umap, UMAPParams),                 }, error = function(e) {,                   errorList &lt;- UMAPParams,                   .logError(e, fn = \"uwot::umap\", info = prefix, ,                     errorList = errorList, logFile = logFile),                 }),                 jointCCA$UMAP1 &lt;- uwotUmap[, 1],                 jointCCA$UMAP2 &lt;- uwotUmap[, 2],                 .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,                   paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),                 p1 &lt;- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Assay, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by Assay\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 p2 &lt;- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Group, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by scRNA Group\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 pdf(file.path(outDir3, paste0(\"Save-Block\", i, ,                   \"-JointCCA-UMAP.pdf\")), width = 12, height = 6, ,                   useDingbats = FALSE),                 ggAlignPlots(p1, p2, type = \"h\"),                 dev.off(),             }, error = function(e) {,             }),         },     },     if (addToArrow) {,         .logDiffTime(\"Transferring Data to ArrowFiles\", tstart, ,             verbose = verbose, logFile = logFile),         matrixName &lt;- .isProtectedArray(matrixName),         integrationFiles &lt;- paste0(tmpFile, \"-IntegrationBlock-\", ,             seq_along(blockList), \".h5\"),         if (!all(file.exists(integrationFiles))) {,             .logMessage(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\", ,                 logFile = logFile),             stop(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\"),         },         h5list &lt;- .safelapply(seq_along(integrationFiles), function(x) {,             h5ls(integrationFiles[x]),         }, threads = threads),         ArrowFiles &lt;- getArrowFiles(ArchRProj),         allSamples &lt;- names(ArrowFiles),         o &lt;- .safelapply(seq_along(allSamples), function(y) {,             sample &lt;- allSamples[y],             prefix &lt;- sprintf(\"%s (%s of %s)\", sample, y, length(ArrowFiles)),             .logDiffTime(sprintf(\"%s Getting GeneIntegrationMatrix From TempFiles!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             sampleIF &lt;- lapply(seq_along(h5list), function(x) {,                 if (any(h5list[[x]]$group == paste0(\"/\", sample))) {,                   integrationFiles[x],                 },                 else {,                   NULL,                 },             }) %&gt;% unlist,             sampleMat &lt;- lapply(seq_along(sampleIF), function(x) {,                 cellNames &lt;- .h5read(sampleIF[x], paste0(sample, ,                   \"/cellNames\")),                 mat &lt;- sparseMatrix(i = .h5read(sampleIF[x], ,                   paste0(sample, \"/i\"))[, 1], j = as.vector(Rle(.h5read(sampleIF[x], ,                   paste0(sample, \"/jValues\"))[, 1], .h5read(sampleIF[x], ,                   paste0(sample, \"/jLengths\"))[, 1])), x = .h5read(sampleIF[x], ,                   paste0(sample, \"/x\"))[, 1], dims = c(nrow(featureDF), ,                   length(cellNames))),                 colnames(mat) &lt;- cellNames,                 mat,             }) %&gt;% Reduce(\"cbind\", .),             sampleMat@x &lt;- exp(sampleMat@x) - 1,             sampleMat &lt;- .normalizeCols(sampleMat, scaleTo = scaleTo),             sampleMat &lt;- drop0(sampleMat),             rownames(sampleMat) &lt;- paste0(featureDF$name),             sampleMat &lt;- sampleMat[, ArchRProj$cellNames[BiocGenerics::which(ArchRProj$Sample == ,                 sample)], drop = FALSE],             o &lt;- .createArrowGroup(ArrowFile = ArrowFiles[sample], ,                 group = matrixName, force = force),             o &lt;- .initializeMat(ArrowFile = ArrowFiles[sample], ,                 Group = matrixName, Class = \"double\", Units = \"NormCounts\", ,                 cellNames = colnames(sampleMat), params = dfParams, ,                 featureDF = featureDF, force = force),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictionScore\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictionScore\")),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictedGroup\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedGroup\")),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictedCell\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedCell\")),             .logDiffTime(sprintf(\"%s Adding GeneIntegrationMatrix to ArrowFile!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             for (z in seq_along(allChr)) {,                 chrz &lt;- allChr[z],                 .logDiffTime(sprintf(\"Adding GeneIntegrationMatrix to %s for Chr (%s of %s)!\", ,                   sample, z, length(allChr)), tstart, verbose = FALSE, ,                   logFile = logFile),                 idz &lt;- BiocGenerics::which(featureDF$seqnames %bcin% ,                   chrz),                 matz &lt;- sampleMat[idz, , drop = FALSE],                 stopifnot(identical(paste0(featureDF$name[idz]), ,                   paste0(rownames(matz)))),                 o &lt;- .addMatToArrow(mat = matz, ArrowFile = ArrowFiles[sample], ,                   Group = paste0(matrixName, \"/\", chrz), binarize = FALSE, ,                   addColSums = TRUE, addRowSums = TRUE, addRowVarsLog2 = TRUE, ,                   logFile = logFile),                 rm(matz),                 if (z%%3 == 0 | z == length(allChr)) {,                   gc(),                 },             },             0,         }, threads = threads),         o &lt;- suppressWarnings(file.remove(integrationFiles)),     },     .logDiffTime(\"Completed Integration with RNA Matrix\", tstart, ,         verbose = verbose, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedCell, name = nameCell, force = TRUE),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedGroup, name = nameGroup, force = TRUE),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictionScore, name = nameScore, force = TRUE),     .endLogging(logFile = logFile),     return(ArchRProj), }</span></td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL, useSeqnames = NULL, verbose = TRUE, ,     binarize = FALSE, threads = getArchRThreads(), logFile = createLogFile(\"exportMatrixForSTREAM\")) , {,     mat &lt;- getMatrixFromProject(ArchRProj = ArchRProj, useMatrix = \"PeakMatrix\", ,         useSeqnames = useSeqnames, verbose = verbose, binarize = binarize, ,         threads = threads, logFile = logFile),     featureDF &lt;- ArchR:::.getFeatureDF(getArrowFiles(ArchRProj)[1], ,         \"PeakMatrix\"),     stopifnot(all(featureDF$idx == rowData(mat)$idx)),     countsDF &lt;- Matrix::summary(assay(mat)),     peaksDF &lt;- data.frame(as.vector(featureDF[, 1]), featureDF[, ,         3], featureDF[, 4]),     cellsDF &lt;- data.frame(colnames(mat)),     data.table::fwrite(countsDF, file = \"STREAM_Counts.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(peaksDF, file = \"STREAM_Regions.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(cellsDF, file = \"STREAM_Sample.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     return(0), }</span></td><td><span style=white-space:pre-wrap>function (ArchRProj = NULL) , {,     if (!inherits(ArchRProj, \"ArchRProject\")) {,         stop(\"Not a valid ArchRProject as input!\"),     },     else {,         ArchRProj,     }, }</span></td><td><span style=white-space:pre-wrap>function (p = NULL, plotWidth = unit(6, \"in\"), plotHeight = unit(6, ,     \"in\"), margin = 0.25, height = 1, it = 0.05, newPage = FALSE) , {,     .requirePackage(\"grid\", source = \"cran\"),     .requirePackage(\"gridExtra\", source = \"cran\"),     if (!inherits(plotWidth, \"unit\")) {,         plotWidth &lt;- unit(plotWidth, \"in\"),     },     if (!inherits(plotHeight, \"unit\")) {,         plotHeight &lt;- unit(plotHeight, \"in\"),     },     g &lt;- ggplotGrob(p),     legend &lt;- grep(\"guide-box\", g$layout$name),     if (length(legend) != 0) {,         gl &lt;- g$grobs[[legend]],         g &lt;- ggplotGrob(p + theme(legend.position = \"none\")),     },     else {,         gl &lt;- NULL,         g &lt;- ggplotGrob(p),     },     panels &lt;- grep(\"panel\", g$layout$name),     panel_index_w &lt;- unique(g$layout$l[panels]),     panel_index_h &lt;- unique(g$layout$t[panels]),     nw &lt;- length(panel_index_w),     nh &lt;- length(panel_index_h),     pw &lt;- convertWidth(plotWidth, unitTo = \"in\", valueOnly = TRUE),     ph &lt;- convertWidth(plotHeight, unitTo = \"in\", valueOnly = TRUE),     pw &lt;- pw * 0.95,     ph &lt;- ph * 0.95,     x &lt;- 0,     width &lt;- 1,     sm &lt;- FALSE,     while (!sm) {,         x &lt;- x + it,         w &lt;- unit(x * width, \"in\"),         h &lt;- unit(x * height/width, \"in\"),         m &lt;- unit(x * margin/width, \"in\"),         g$widths[panel_index_w] &lt;- rep(w, nw),         g$heights[panel_index_h] &lt;- rep(h, nh),         sw &lt;- convertWidth(x = sum(g$widths) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sh &lt;- convertHeight(x = sum(g$heights) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sm &lt;- sw &gt; pw | sh &gt; ph,     },     if (length(legend) != 0) {,         sgh &lt;- convertHeight(x = sum(g$heights), unitTo = \"in\", ,             valueOnly = TRUE),         sgw &lt;- convertWidth(x = sum(g$widths), unitTo = \"in\", ,             valueOnly = TRUE),         slh &lt;- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,             valueOnly = TRUE),         slw &lt;- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,             valueOnly = TRUE),         size &lt;- 6,         wh &lt;- 0.1,         it &lt;- 0,         while (slh &gt; 0.2 * ph | slw &gt; pw) {,             it &lt;- it + 1,             if (it &gt; 3) {,                 break,             },             size &lt;- size * 0.8,             wh &lt;- wh * 0.8,             gl &lt;- ggplotGrob(p + theme(legend.key.width = unit(wh, ,                 \"cm\"), legend.key.height = unit(wh, \"cm\"), legend.spacing.x = unit(0, ,                 \"cm\"), legend.spacing.y = unit(0, \"cm\"), legend.text = element_text(size = max(size, ,                 2))) + guides(fill = guide_legend(ncol = 4), ,                 color = guide_legend(ncol = 4)))$grobs[[legend]],             slh &lt;- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,                 valueOnly = TRUE),             slw &lt;- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,                 valueOnly = TRUE),         },         p &lt;- grid.arrange(g, gl, ncol = 1, nrow = 2, heights = unit.c(unit(sgh, ,             \"in\"), unit(min(slh, 0.2 * pw), \"in\")), newpage = newPage),     },     else {,         p &lt;- grid.arrange(g, newpage = newPage),     },     invisible(p), }</span></td></tr>\n",
       "\t<tr><th scope=row>visible</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>⋯</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 2 × 55\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/AllClasses.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/AnnotationGenome.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/AnnotationPeaks.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser-CRHR1.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser-FKBP5.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser3.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser4-CAMKV.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser4-KCNIP4.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ArchRBrowser4-xiangyu.R & ⋯ & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/MultiModal.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/plot\\_footprints.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ProjectMethods.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/QualityControl.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/RcppExports.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ReproduciblePeakSet.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/RNAIntegration.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/Trajectory.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/ValidationUtils.R & /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti\\_project/ArchR/R/VisualizeData.R\\\\\n",
       "  & <chr> & <lgl> & <fn> & <lgl> & <fn> & <lgl> & <fn> & <lgl> & <fn> & <lgl> & ⋯ & <fn> & <lgl> & <fn> & <lgl> & <fn> & <lgl> & <fn> & <lgl> & <fn> & <lgl>\\\\\n",
       "\\hline\n",
       "\tvalue & rownames & function (genome = NULL) , \\{,     .validInput(input = genome, name = \"genome\", valid = \"character\"),     if (toupper(genome) == \"HG19\" \\textbar{} toupper(genome) == \"HG38\") \\{,         if (suppressWarnings(!require(org.Hs.eg.db))) \\{,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Hs.eg.db\", update = FALSE),         \\},         library(org.Hs.eg.db),         annodb <- org.Hs.eg.db,     \\},     else if (toupper(genome) == \"MM9\" \\textbar{} toupper(genome) == \"MM10\") \\{,         if (suppressWarnings(!require(org.Mm.eg.db))) \\{,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Mm.eg.db\", update = FALSE),         \\},         library(org.Mm.eg.db),         annodb <- org.Mm.eg.db,     \\},     else \\{,         stop(\"Genome not recognized!\"),     \\},     return(annodb), \\} & function (seEnrich = NULL, pal = paletteContinuous(set = \"comet\", ,     n = 100), n = 10, cutOff = 20, pMax = Inf, clusterCols = TRUE, ,     binaryClusterRows = TRUE, labelRows = TRUE, rastr = TRUE, ,     transpose = FALSE, returnMatrix = FALSE, logFile = createLogFile(\"plotEnrichHeatmap\")) , \\{,     .validInput(input = seEnrich, name = \"seEnrich\", valid = c(\"SummarizedExperiment\")),     .validInput(input = pal, name = \"pal\", valid = c(\"character\")),     .validInput(input = n, name = \"n\", valid = c(\"integer\")),     .validInput(input = cutOff, name = \"cutOff\", valid = c(\"numeric\")),     .validInput(input = pMax, name = \"pMax\", valid = c(\"numeric\")),     .validInput(input = clusterCols, name = \"clusterCols\", valid = c(\"boolean\")),     .validInput(input = binaryClusterRows, name = \"binaryClusterRows\", ,         valid = c(\"boolean\")),     .validInput(input = labelRows, name = \"labelRows\", valid = c(\"boolean\")),     .validInput(input = rastr, name = \"rastr\", valid = c(\"boolean\")),     .validInput(input = transpose, name = \"transpose\", valid = c(\"boolean\")),     .validInput(input = returnMatrix, name = \"returnMatrix\", ,         valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotEnrichHeatmap Input-Parameters\", logFile = logFile),     mat <- assays(seEnrich){[}{[}\"mlog10Padj\"{]}{]},     .logThis(mat, \"mat-mlog10Padj\", logFile = logFile),     keep <- lapply(seq\\_len(ncol(mat)), function(x) \\{,         idx <- head(order(mat{[}, x{]}, decreasing = TRUE), n),         rownames(mat){[}idx{[}which(mat{[}idx, x{]} > cutOff){]}{]},     \\}) \\%>\\% unlist \\%>\\% unique,     mat <- mat{[}keep, , drop = FALSE{]},     .logThis(mat, \"mat-mlog10Padj-Filter\", logFile = logFile),     if (nrow(mat) == 0) \\{,         stop(\"No enrichments found for your cutoff!\"),     \\},     passMat <- lapply(seq\\_len(nrow(mat)), function(x) \\{,         (mat{[}x, {]} >= 0.9 * max(mat{[}x, {]})) * 1,     \\}) \\%>\\% Reduce(\"rbind\", .) \\%>\\% data.frame,     colnames(passMat) <- colnames(mat),     .logThis(passMat, \"passMat\", logFile = logFile),     mat{[}mat > pMax{]} <- pMax,     if (nrow(mat) == 0) \\{,         stop(\"No enrichments found for your cutoff!\"),     \\},     mat <- .rowScale(as.matrix(mat), min = 0),     if (pMax != 100) \\{,         rownames(mat{[}{[}1{]}{]}) <- paste0(rownames(mat{[}{[}1{]}{]}), \" (\", ,             round(mat\\$max), \")\"),         rownames(passMat) <- rownames(mat{[}{[}1{]}{]}),     \\},     .logThis(mat, \"mat-mlog10Padj-RowScale\", logFile = logFile),     mat2 <- mat{[}{[}1{]}{]} * 100,     if (nrow(mat2) > 1 \\& ncol(mat2) > 1) \\{,         if (binaryClusterRows) \\{,             bS <- .binarySort(mat2, lmat = passMat{[}rownames(mat2), ,                 colnames(mat2){]}, clusterCols = clusterCols),             mat2 <- bS{[}{[}1{]}{]}{[}, colnames(mat2){]},             clusterRows <- FALSE,             clusterCols <- bS{[}{[}2{]}{]},         \\},         else \\{,             clusterRows <- TRUE,         \\},     \\},     else \\{,         clusterCols <- NULL,         clusterRows <- FALSE,     \\},     if (nrow(mat2) > 100) \\{,         borderColor <- FALSE,     \\},     else \\{,         borderColor <- TRUE,     \\},     .logThis(mat2, \"mat-mlog10Padj-RowScale-Plot\", logFile = logFile),     if (transpose) \\{,         if (!is.null(clusterCols)) \\{,             mat2 <- t(mat2{[}, clusterCols\\$order, drop = FALSE{]}),         \\},         else \\{,             mat2 <- t(mat2),         \\},         if (returnMatrix) \\{,             return(mat2),         \\},         ht <- tryCatch(\\{,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq\\_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) {[}0-Max{]}\"),         \\}, error = function(e) \\{,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq\\_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) {[}0-Max{]}\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         \\}),     \\},     else \\{,         if (returnMatrix) \\{,             return(mat2),         \\},         ht <- tryCatch(\\{,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq\\_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) {[}0-Max{]}\"),         \\}, error = function(e) \\{,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq\\_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) {[}0-Max{]}\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         \\}),     \\},     return(ht), \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , \\{,     .requirePackage(\"patchwork\", installInfo = \"devtools::install\\_github('thomasp85/patchwork')\"),     if (is.null(pal)) \\{,         pal <- paletteDiscrete(values = featureMat\\$Group, set = \"stallion\"),     \\},     if (log2Norm) \\{,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     \\},     else \\{,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     \\},     featurePlot <- ggGroup(x = featureMat\\$Group, y = featureMat{[}, ,         feature{]}, groupOrder = gtools::mixedsort(paste0(unique(featureMat\\$Group))), ,         pal = pal) + facet\\_wrap(x \\textasciitilde{} ., ncol = 1, scales = \"free\\_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme\\_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element\\_blank(), axis.title.y = element\\_blank(), ,         axis.text.y = element\\_blank(), axis.ticks.y = element\\_blank(), ,         strip.text = element\\_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element\\_text(angle = 0), ,         strip.background = element\\_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) \\%in\\% \"bulktrack\")) \\{,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot + plot\\_spacer(),         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     else \\{,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList{[}{[}idx{]}{]} + featurePlot,         plotList{[}idx{]} <- NULL,         for (i in seq\\_along(plotList)) \\{,             p <- p + plotList{[}{[}i{]}{]} + plot\\_spacer() + plot\\_spacer(),         \\},         p <- p + plot\\_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     \\},     p, \\} & ⋯ & function (ArchRProj = NULL, name = \"CombinedDims\", reducedDims = NULL, ,     dimWeights = NULL, dimsToUse = NULL, scaleDims = NULL, corCutOff = 0.75) , \\{,     if (is.null(dimWeights)) \\{,         dimWeights <- rep(1, length(reducedDims)),     \\},     combinedDims <- lapply(seq\\_along(reducedDims), function(x) \\{,         rD <- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims{[}x{]}, ,             dimsToUse = dimsToUse, scaleDims = scaleDims, corCutOff = corCutOff),         cV <- apply(as.matrix(rD), 2, var),         normV <- 1/sqrt(sum(cV)),         rD * normV * dimWeights{[}x{]},     \\}) \\%>\\% Reduce(\"cbind\", .),     ArchRProj@reducedDims{[}{[}name{]}{]} <- SimpleList(matRD = combinedDims, ,         scaleDims = NA, corToDepth = list(scaled = rep(0, ncol(combinedDims)), ,             none = rep(0, ncol(combinedDims)))),     ArchRProj, \\} & function (seFoot = NULL, names = NULL, pal = NULL, flank = 250, ,     flankNorm = 50, normMethod = \"Subtract\", smoothWindow = NULL, ,     baseSize = 6, plot = TRUE, ArchRProj = NULL, plotName = paste0(\"Plot-Footprints-\", ,         normMethod), height = 6, width = 4, addDOC = TRUE, force = FALSE, ,     logFile = createLogFile(\"plotFootprints\")) , \\{,     tstart <- Sys.time(),     if (is.null(names)) \\{,         names <- names(assays(seFoot)),     \\},     if (plot) \\{,         name <- gsub(\"\\textbackslash{}\\textbackslash{}.pdf\", \"\", plotName),         if (is.null(ArchRProj)) \\{,             outDir <- \"Plots\",         \\},         else \\{,             outDir <- file.path(getOutputDirectory(ArchRProj), ,                 \"Plots\"),         \\},         dir.create(outDir, showWarnings = FALSE),         if (addDOC) \\{,             doc <- gsub(\":\", \"-\", stringr::str\\_split(Sys.time(), ,                 pattern = \" \", simplify = TRUE){[}1, 2{]}),             filename <- file.path(outDir, paste0(name, \"\\_Date-\", ,                 Sys.Date(), \"\\_Time-\", doc, \".pdf\")),         \\},         else \\{,             filename <- file.path(outDir, paste0(name, \".pdf\")),         \\},         pdf(filename, width = width, height = height, useDingbats = FALSE),     \\},     ggList <- lapply(seq\\_along(names), function(x) \\{,         gg <- .ggFootprint(seFoot = seFoot, name = names{[}x{]}, ,             pal = pal, smoothWindow = smoothWindow, flank = flank, ,             flankNorm = flankNorm, baseSize = baseSize, normMethod = normMethod, ,             logFile = logFile),         if (plot) \\{,             if (x != 1) \\{,                 grid::grid.newpage(),             \\},             grid::grid.draw(gg),             return(0),         \\},         else \\{,             return(gg),         \\},     \\}),     if (!plot) \\{,         names(ggList) <- names,         ggList,     \\},     else \\{,         dev.off(),         return(invisible(0)),     \\}, \\} & function (ArchRProj = NULL, features = NULL, name = NULL, addRatio = TRUE, ,     threads = getArchRThreads(), logFile = createLogFile(\"addFeatureCounts\")) , \\{,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = features, name = \"features\", valid = c(\"granges\")),     .validInput(input = name, name = \"name\", valid = c(\"character\")),     .validInput(input = addRatio, name = \"addRatio\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     ArrowFiles <- getArrowFiles(ArchRProj),     cellNames <- ArchRProj\\$cellNames,     featuresList <- split(features, seqnames(features)),     h5disableFileLocking(),     countsDF <- .safelapply(seq\\_along(featuresList), function(i) \\{,         chri <- names(featuresList){[}i{]},         cellTotal <- rep(0, length(cellNames)),         names(cellTotal) <- cellNames,         featuresi <- ranges(featuresList{[}{[}i{]}{]}),         .logDiffTime(paste0(\"Counting in \", chri, \" (\", i, \" of \", ,             length(featuresList), \")\"), t1 = tstart, logFile = logFile),         for (j in seq\\_along(ArrowFiles)) \\{,             fragmentsij <- .getFragsFromArrow(ArrowFile = ArrowFiles{[}j{]}, ,                 chr = chri, out = \"IRanges\", cellNames = cellNames),             if (length(fragmentsij) > 0) \\{,                 mcols(fragmentsij)\\$RG@values <- match(mcols(fragmentsij)\\$RG@values, ,                   cellNames),                 for (y in seq\\_len(2)) \\{,                   if (y == 1) \\{,                     temp <- IRanges(start(fragmentsij), width = 1),                   \\},                   else if (y == 2) \\{,                     temp <- IRanges(end(fragmentsij), width = 1),                   \\},                   stopifnot(length(temp) == length(fragmentsij)),                   tabSum <- S4Vectors:::tabulate(mcols(fragmentsij)\\$RG{[}queryHits(findOverlaps(temp, ,                     featuresi)){]}),                   cellTotal{[}seq\\_along(tabSum){]} <- cellTotal{[}seq\\_along(tabSum){]} + ,                     tabSum,                 \\},             \\},             if (j\\%\\%3 == 0) \\{,                 gc(),             \\},         \\},         cellTotal,     \\}, threads = threads) \\%>\\% Reduce(\"rbind\", .),     totalCounts <- colSums(countsDF),     countRatio <- totalCounts/(2 * ArchRProj\\$nFrags),     .logDiffTime(sprintf(\"Adding \\%s to cellColData\", paste0(name, ,         \"Counts\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj, data = totalCounts, ,         cells = names(totalCounts), name = paste0(name, \"Counts\"), ,         force = TRUE),     .logDiffTime(sprintf(\"Adding \\%s to cellColData\", paste0(name, ,         \"Ratio\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj, data = countRatio, ,         cells = names(totalCounts), name = paste0(name, \"Ratio\"), ,         force = TRUE),     ArchRProj, \\} & function (ArchRProj = NULL, groupBy = \"Sample\", chromSizes = getChromSizes(ArchRProj), ,     maxSize = 750, pal = NULL, returnDF = FALSE, threads = getArchRThreads(), ,     logFile = createLogFile(\"plotFragmentSizes\")) , \\{,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = maxSize, name = \"maxSize\", valid = c(\"integer\")),     .validInput(input = returnDF, name = \"returnDF\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotFragmentSizes Input-Parameters\", logFile = logFile),     chr <- paste0(seqnames(chromSizes)),     groups <- getCellColData(ArchRProj = ArchRProj, select = groupBy, ,         drop = FALSE),     uniqGroups <- gtools::mixedsort(unique(groups{[}, 1{]})),     if (threads > 1) \\{,         h5disableFileLocking(),     \\},     dfFS <- .safelapply(seq\\_along(uniqGroups), function(x) \\{,         .logDiffTime(paste0(uniqGroups{[}x{]}, \" Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         cellx <- rownames(groups){[}which(paste0(groups{[}, 1{]}) == ,             uniqGroups{[}x{]}){]},         for (i in seq\\_along(chr)) \\{,             if (i == 1) \\{,                 fsi <- unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes{[}paste0(seqnames(chromSizes)) \\%in\\% ,                     chr{[}i{]}{]}, cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) \\%>\\% width \\%>\\% tabulate(nbins = maxSize),             \\},             else \\{,                 fsi <- fsi + unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes{[}paste0(seqnames(chromSizes)) \\%in\\% ,                     chr{[}i{]}{]}, cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) \\%>\\% width \\%>\\% tabulate(nbins = maxSize),             \\},             .logThis(fsi, paste0(uniqGroups{[}x{]}, \" : FragSizes : \", ,                 chr{[}i{]}), logFile = logFile),         \\},         df <- DataFrame(group = uniqGroups{[}x{]}, fragmentSize = seq\\_along(fsi), ,             fragmentPercent = round(100 * fsi/sum(fsi), 4)),         .logThis(df, paste0(uniqGroups{[}x{]}, \" : Frag DF\"), logFile = logFile),         .logDiffTime(paste0(uniqGroups{[}x{]}, \" Finished Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         df,     \\}, threads = threads) \\%>\\% Reduce(\"rbind\", .),     .logThis(dfFS, paste0(\"All : FragSizes DF\"), logFile = logFile),     .endLogging(logFile = logFile),     if (threads > 1) \\{,         h5enableFileLocking(),     \\},     if (returnDF) \\{,         return(dfFS),     \\},     else \\{,         plotDF <- data.frame(dfFS),         if (is.null(pal)) \\{,             pal <- paletteDiscrete(values = unique(plotDF\\$group)),         \\},         p <- ggplot(plotDF, aes(fragmentSize, fragmentPercent, ,             color = group)) + geom\\_line(size = 1) + theme\\_ArchR() + ,             xlab(\"ATAC-seq Fragment Size (bp)\") + ylab(\"Percentage of Fragments\") + ,             scale\\_color\\_manual(values = pal) + scale\\_y\\_continuous(limits = c(0, ,             max(plotDF\\$fragmentPercent) * 1.05), expand = c(0, ,             0)) + scale\\_x\\_continuous(limits = c(min(plotDF\\$fragmentSize), ,             max(plotDF\\$fragmentSize)), expand = c(0, 0)),         p,     \\}, \\} & function (string\\_vector, id\\_vector, n\\_id, window, w, kmer) , \\{,     .Call(\"\\_ArchR\\_kmerIDFrequencyCpp\", PACKAGE = \"ArchR\", string\\_vector, ,         id\\_vector, n\\_id, window, w, kmer), \\} & function () , \\{,     message(\"Searching For MACS2..\"),     if (.suppressAll(.checkPath(\"macs2\", throwError = FALSE))) \\{,         message(\"Found with \\$path!\"),         return(\"macs2\"),     \\},     message(\"Not Found in \\$PATH\"),     search2 <- suppressWarnings(tryCatch(\\{,         system2(\"pip\", \"show macs2\", stdout = TRUE, stderr = NULL),     \\}, error = function(x) \\{,         \"ERROR\",     \\})),     search3 <- suppressWarnings(tryCatch(\\{,         system2(\"pip3\", \"show macs2\", stdout = TRUE, stderr = NULL),     \\}, error = function(x) \\{,         \"ERROR\",     \\})),     if (length(search2) > 0) \\{,         if (search2{[}1{]} != \"ERROR\") \\{,             path2Install <- gsub(\"Location: \", \"\", search2{[}grep(\"Location\", ,                 search2, ignore.case = TRUE){]}),             path2Bin <- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) \\{,                 message(\"Found with pip!\"),                 return(path2Bin),             \\},         \\},     \\},     message(\"Not Found with pip\"),     if (length(search3) > 0) \\{,         if (search3{[}1{]} != \"ERROR\") \\{,             path2Install <- gsub(\"Location: \", \"\", search3{[}grep(\"Location\", ,                 search3, ignore.case = TRUE){]}),             path2Bin <- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) \\{,                 message(\"Found with pip3!\"),                 return(path2Bin),             \\},         \\},     \\},     message(\"Not Found with pip3\"),     stop(\"Could Not Find Macs2! Please install w/ pip, add to your \\$PATH, or just supply the macs2 path directly and avoid this function!\"), \\} & function (ArchRProj = NULL, useMatrix = \"GeneScoreMatrix\", matrixName = \"GeneIntegrationMatrix\", ,     reducedDims = \"IterativeLSI\", seRNA = NULL, groupATAC = NULL, ,     groupRNA = NULL, groupList = NULL, sampleCellsATAC = 10000, ,     sampleCellsRNA = 10000, embeddingATAC = NULL, embeddingRNA = NULL, ,     dimsToUse = 1:30, scaleDims = NULL, corCutOff = 0.75, plotUMAP = TRUE, ,     UMAPParams = list(n\\_neighbors = 40, min\\_dist = 0.4, metric = \"cosine\", ,         verbose = FALSE), nGenes = 2000, useImputation = TRUE, ,     reduction = \"cca\", addToArrow = TRUE, scaleTo = 10000, genesUse = NULL, ,     nameCell = \"predictedCell\", nameGroup = \"predictedGroup\", ,     nameScore = \"predictedScore\", transferParams = list(), threads = getArchRThreads(), ,     verbose = TRUE, force = FALSE, logFile = createLogFile(\"addGeneIntegrationMatrix\"), ,     ...) , \\{,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = useMatrix, name = \"useMatrix\", valid = c(\"character\")),     .validInput(input = matrixName, name = \"matrixName\", valid = c(\"character\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = seRNA, name = \"seRNA\", valid = c(\"SummarizedExperiment\", ,         \"Seurat\")),     .validInput(input = groupATAC, name = \"groupATAC\", valid = c(\"character\", ,         \"null\")),     .validInput(input = groupRNA, name = \"groupRNA\", valid = c(\"character\")),     .validInput(input = groupList, name = \"groupList\", valid = c(\"list\", ,         \"null\")),     .validInput(input = sampleCellsATAC, name = \"sampleCellsATAC\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = sampleCellsRNA, name = \"sampleCellsRNA\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = embeddingATAC, name = \"embeddingATAC\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = embeddingRNA, name = \"embeddingRNA\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = dimsToUse, name = \"dimsToUse\", valid = c(\"numeric\", ,         \"null\")),     .validInput(input = scaleDims, name = \"scaleDims\", valid = c(\"boolean\", ,         \"null\")),     .validInput(input = plotUMAP, name = \"plotUMAP\", valid = c(\"boolean\")),     .validInput(input = UMAPParams, name = \"UMAPParams\", valid = c(\"list\")),     .validInput(input = nGenes, name = \"nGenes\", valid = c(\"integer\")),     .validInput(input = useImputation, name = \"useImputation\", ,         valid = c(\"boolean\")),     .validInput(input = reduction, name = \"reduction\", valid = c(\"character\")),     .validInput(input = addToArrow, name = \"addToArrow\", valid = c(\"boolean\")),     .validInput(input = scaleTo, name = \"scaleTo\", valid = c(\"numeric\")),     .validInput(input = genesUse, name = \"genesUse\", valid = c(\"character\", ,         \"null\")),     .validInput(input = nameCell, name = \"nameCell\", valid = c(\"character\")),     .validInput(input = nameGroup, name = \"nameGroup\", valid = c(\"character\")),     .validInput(input = nameScore, name = \"nameScore\", valid = c(\"character\")),     .validInput(input = transferParams, name = \"transferParams\", ,         valid = c(\"list\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = verbose, name = \"verbose\", valid = c(\"boolean\")),     .validInput(input = force, name = \"force\", valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logDiffTime(\"Running Seurat's Integration Stuart* et al 2019\", ,         tstart, verbose = verbose, logFile = logFile),     .requirePackage(\"Seurat\", source = \"cran\"),     .logThis(append(args, mget(names(formals()), sys.frame(sys.nframe()))), ,         \"Input-Parameters\", logFile = logFile),     if (is.null(groupList)) \\{,         groupList <- SimpleList(),         groupList{[}{[}1{]}{]} <- SimpleList(ATAC = ArchRProj\\$cellNames, ,             RNA = colnames(seRNA)),     \\},     .logDiffTime(\"Checking ATAC Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (useMatrix \\%ni\\% getAvailableMatrices(ArchRProj)) \\{,         .logMessage(paste0(\"Matrix \", useMatrix, \" does not exist in the provided ArchRProject. See available matrix names from getAvailableMatrices()!\"), ,             logFile = logFile),         stop(\"Matrix name provided to useMatrix does not exist in ArchRProject!\"),     \\},     if (!is.null(groupATAC)) \\{,         dfATAC <- getCellColData(ArchRProj = ArchRProj, select = groupATAC, ,             drop = FALSE),     \\},     nCell <- rep(0, length(ArchRProj\\$cellNames)),     names(nCell) <- ArchRProj\\$cellNames,     groupList <- lapply(seq\\_along(groupList), function(x) \\{,         ATAC <- groupList{[}{[}x{]}{]}\\$ATAC,         if (!is.null(groupATAC)) \\{,             if (any(ATAC \\%in\\% dfATAC{[}, 1{]})) \\{,                 idx <- which(ATAC \\%in\\% dfATAC{[}, 1{]}),                 ATAC2 <- rownames(dfATAC){[}which(dfATAC{[}, 1{]} \\%in\\% ,                   ATAC{[}idx{]}){]},                 if (length(idx) == length(ATAC)) \\{,                   ATAC <- ATAC2,                 \\},                 else \\{,                   ATAC <- c(ATAC{[}-idx{]}, ATAC2),                 \\},             \\},         \\},         SimpleList(ATAC = ATAC, RNA = groupList{[}{[}x{]}{]}\\$RNA),     \\}) \\%>\\% SimpleList,     for (i in seq\\_along(groupList)) \\{,         nCell{[}groupList{[}{[}i{]}{]}\\$ATAC{]} <- nCell{[}groupList{[}{[}i{]}{]}\\$ATAC{]} + ,             1,     \\},     if (!all(nCell == 1)) \\{,         .logMessage(paste0(\"Missing \", length(which(nCell == ,             0)), \" cells. Found \", length(which(nCell > 1)), ,             \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"), ,             logFile = logFile),         stop(\"Missing \", length(which(nCell == 0)), \" cells. Found \", ,             length(which(nCell > 1)), \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"),     \\},     .logDiffTime(\"Checking RNA Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (inherits(seRNA, \"SummarizedExperiment\")) \\{,         seuratRNA <- CreateSeuratObject(counts = assay(seRNA)),         if (groupRNA \\%ni\\% colnames(colData(seRNA))) \\{,             .logMessage(\"groupRNA not in colData of seRNA\", logFile = logFile),             stop(\"groupRNA not in colData of seRNA\"),         \\},         seuratRNA\\$Group <- paste0(colData(seRNA){[}, groupRNA, ,             drop = TRUE{]}),         rm(seRNA),     \\},     else \\{,         if (groupRNA \\%ni\\% colnames(seRNA@meta.data)) \\{,             .logMessage(\"groupRNA not in meta.data of Seurat Object\", ,                 logFile = logFile),             stop(\"groupRNA not in meta.data of Seurat Object\"),         \\},         seuratRNA <- seRNA,         seuratRNA\\$Group <- paste0(seRNA@meta.data{[}, groupRNA{]}),         rm(seRNA),     \\},     if (\"RNA\" \\%in\\% names(seuratRNA@assays)) \\{,         DefaultAssay(seuratRNA) <- \"RNA\",     \\},     else \\{,         stop(\"'RNA' is not present in Seurat Object's Assays! Please make sure that this assay is present!\"),     \\},     gc(),     if (!is.null(groupRNA)) \\{,         dfRNA <- DataFrame(row.names = colnames(seuratRNA), Group = seuratRNA\\$Group),     \\},     groupList <- lapply(seq\\_along(groupList), function(x) \\{,         RNA <- groupList{[}{[}x{]}{]}\\$RNA,         if (!is.null(groupRNA)) \\{,             if (any(RNA \\%in\\% dfRNA{[}, 1{]})) \\{,                 idx <- which(RNA \\%in\\% dfRNA{[}, 1{]}),                 RNA2 <- rownames(dfRNA){[}which(dfRNA{[}, 1{]} \\%in\\% ,                   RNA{[}idx{]}){]},                 if (length(idx) == length(RNA)) \\{,                   RNA <- RNA2,                 \\},                 else \\{,                   RNA <- c(RNA{[}-idx{]}, RNA2),                 \\},             \\},         \\},         SimpleList(ATAC = groupList{[}{[}x{]}{]}\\$ATAC, RNA = RNA),     \\}) \\%>\\% SimpleList,     cellRNA <- unlist(lapply(groupList, function(x) x\\$RNA)),     if (!all(cellRNA \\%in\\% colnames(seuratRNA))) \\{,         .logMessage(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\", ,             logFile = logFile),         stop(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\"),     \\},     seuratRNA <- seuratRNA{[}, unique(cellRNA){]},     seuratRNA <- NormalizeData(object = seuratRNA, verbose = FALSE),     geneDF <- .getFeatureDF(getArrowFiles(ArchRProj), useMatrix),     sumOverlap <- sum(unique(geneDF\\$name) \\%in\\% unique(rownames(seuratRNA))),     if (sumOverlap < 5) \\{,         stop(\"Error not enough overlaps (\", sumOverlap, \") between gene names from gene scores (ArchR) and rna matrix (seRNA)!\"),     \\},     .logDiffTime(paste0(\"Found \", sumOverlap, \" overlapping gene names from gene scores and rna matrix!\"), ,         tstart, verbose = TRUE, logFile = logFile),     .logDiffTime(\"Creating Integration Blocks\", tstart, verbose = verbose, ,         logFile = logFile),     blockList <- SimpleList(),     for (i in seq\\_along(groupList)) \\{,         gLi <- groupList{[}{[}i{]}{]},         if (length(gLi\\$ATAC) > sampleCellsATAC) \\{,             if (!is.null(embeddingATAC)) \\{,                 probATAC <- .getDensity(embeddingATAC{[}gLi\\$ATAC, ,                   1{]}, embeddingATAC{[}gLi\\$ATAC, 2{]})\\$density,                 probATAC <- probATAC/max(probATAC),                 cellsATAC <- gLi\\$ATAC{[}order(probATAC, decreasing = TRUE){]},             \\},             else \\{,                 cellsATAC <- sample(gLi\\$ATAC, length(gLi\\$ATAC)),             \\},             cutoffs <- lapply(seq\\_len(1000), function(x) length(gLi\\$ATAC)/x) \\%>\\% ,                 unlist,             blockSize <- ceiling(min(cutoffs{[}order(abs(cutoffs - ,                 sampleCellsATAC)){[}1{]}{]} + 1, length(gLi\\$ATAC))),             nBlocks <- ceiling(length(gLi\\$ATAC)/blockSize),             blocks <- lapply(seq\\_len(nBlocks), function(x) \\{,                 cellsATAC{[}seq(x, length(cellsATAC), nBlocks){]},             \\}) \\%>\\% SimpleList,         \\},         else \\{,             blocks <- list(gLi\\$ATAC),         \\},         if (!is.null(embeddingRNA)) \\{,             probRNA <- .getDensity(embeddingRNA{[}gLi\\$RNA, 1{]}, ,                 embeddingRNA{[}gLi\\$RNA, 2{]})\\$density,             probRNA <- probRNA/max(probRNA),         \\},         else \\{,             probRNA <- rep(1, length(gLi\\$RNA)),         \\},         blockListi <- lapply(seq\\_along(blocks), function(x) \\{,             SimpleList(ATAC = blocks{[}{[}x{]}{]}, RNA = sample(x = gLi\\$RNA, ,                 size = min(sampleCellsRNA, length(gLi\\$RNA)), ,                 prob = probRNA)),         \\}) \\%>\\% SimpleList,         blockList <- c(blockList, blockListi),     \\},     rm(groupList),     .logDiffTime(\"Prepping Interation Data\", tstart, verbose = verbose, ,         logFile = logFile),     subProj <- ArchRProj,     subProj@imputeWeights <- SimpleList(),     geneDF <- .getFeatureDF(getArrowFiles(subProj), useMatrix),     geneDF <- geneDF{[}geneDF\\$name \\%in\\% rownames(seuratRNA), , ,         drop = FALSE{]},     splitGeneDF <- S4Vectors::split(geneDF, geneDF\\$seqnames),     featureDF <- lapply(splitGeneDF, function(x) \\{,         x\\$idx <- seq\\_len(nrow(x)),         return(x),     \\}) \\%>\\% Reduce(\"rbind\", .),     dfParams <- data.frame(reduction = reduction),     allChr <- unique(featureDF\\$seqnames),     tmpFile <- .tempfile(),     o <- suppressWarnings(file.remove(paste0(tmpFile, \"-IntegrationBlock-\", ,         seq\\_along(blockList), \".h5\"))),     if (threads > 1) \\{,         h5disableFileLocking(),     \\},     rD <- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims, ,         corCutOff = corCutOff, dimsToUse = dimsToUse),     outDir1 <- getOutputDirectory(ArchRProj),     outDir2 <- file.path(outDir1, \"RNAIntegration\"),     outDir3 <- file.path(outDir2, matrixName),     dir.create(outDir1, showWarnings = FALSE),     dir.create(outDir2, showWarnings = FALSE),     dir.create(outDir3, showWarnings = FALSE),     prevFiles <- list.files(outDir3, full.names = TRUE),     prevFiles <- .suppressAll(file.remove(prevFiles)),     tstart <- Sys.time(),     threads2 <- max(ceiling(threads * 0.75), 1),     .logDiffTime(paste0(\"Computing Integration in \", length(blockList), ,         \" Integration Blocks!\"), tstart, verbose = verbose, logFile = logFile),     dfAll <- .safelapply(seq\\_along(blockList), function(i) \\{,         prefix <- sprintf(\"Block (\\%s of \\%s) :\", i, length(blockList)),         .logDiffTime(sprintf(\"\\%s Computing Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         blocki <- blockList{[}{[}i{]}{]},         subProj@cellColData <- subProj@cellColData{[}blocki\\$ATAC, ,             {]},         subProj@sampleColData <- subProj@sampleColData{[}unique(subProj\\$Sample), ,             , drop = FALSE{]},         subRNA <- seuratRNA{[}, blocki\\$RNA{]},         subRNA <- subRNA{[}rownames(subRNA) \\%in\\% geneDF\\$name, {]},         .logDiffTime(sprintf(\"\\%s Identifying Variable Genes\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         subRNA <- FindVariableFeatures(object = subRNA, nfeatures = nGenes, ,             verbose = FALSE),         subRNA <- ScaleData(object = subRNA, verbose = FALSE),         if (is.null(genesUse)) \\{,             genesUse <- VariableFeatures(object = subRNA),         \\},         .logDiffTime(sprintf(\"\\%s Getting GeneScoreMatrix\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         mat <- .getPartialMatrix(getArrowFiles(subProj), featureDF = geneDF{[}geneDF\\$name \\%in\\% ,             genesUse, {]}, threads = 1, cellNames = subProj\\$cellNames, ,             useMatrix = useMatrix, verbose = FALSE),         rownames(mat) <- geneDF{[}geneDF\\$name \\%in\\% genesUse, \"name\"{]},         .logThis(mat, paste0(\"GeneScoreMat-Block-\", i), logFile = logFile),         if (useImputation) \\{,             .logDiffTime(sprintf(\"\\%s Imputing GeneScoreMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             imputeParams <- list(),             imputeParams\\$ArchRProj <- subProj,             imputeParams\\$randomSuffix <- TRUE,             imputeParams\\$reducedDims <- reducedDims,             imputeParams\\$dimsToUse <- dimsToUse,             imputeParams\\$scaleDims <- scaleDims,             imputeParams\\$corCutOff <- corCutOff,             imputeParams\\$threads <- 1,             imputeParams\\$logFile <- logFile,             subProj <- suppressMessages(do.call(addImputeWeights, ,                 imputeParams)),             mat <- suppressMessages(imputeMatrix(mat = mat, imputeWeights = getImputeWeights(subProj), ,                 verbose = FALSE, logFile = logFile)),             o <- suppressWarnings(file.remove(unlist(getImputeWeights(subProj){[}{[}1{]}{]}))),             .logThis(mat, paste0(\"GeneScoreMat-Block-Impute-\", ,                 i), logFile = logFile),         \\},         mat <- log(mat + 1),         seuratATAC <- Seurat::CreateSeuratObject(counts = mat{[}head(seq\\_len(nrow(mat)), ,             5), , drop = FALSE{]}),         seuratATAC{[}{[}\"GeneScore\"{]}{]} <- Seurat::CreateAssayObject(counts = mat),         rm(mat),         DefaultAssay(seuratATAC) <- \"GeneScore\",         seuratATAC <- Seurat::ScaleData(seuratATAC, verbose = FALSE),         .logDiffTime(sprintf(\"\\%s Seurat FindTransferAnchors\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferAnchors <- .retryCatch(\\{,             gc(),             Seurat::FindTransferAnchors(reference = subRNA, query = seuratATAC, ,                 reduction = reduction, features = genesUse, verbose = FALSE, ,                 ...),         \\}, maxAttempts = 2, logFile = logFile),         .logThis(paste0(utils::capture.output(transferAnchors), ,             collapse = \"\\textbackslash{}n\"), paste0(\"transferAnchors-\", i), ,             logFile = logFile),         rDSub <- rD{[}colnames(seuratATAC), , drop = FALSE{]},         .logThis(rDSub, paste0(\"rDSub-\", i), logFile = logFile),         transferParams\\$anchorset <- transferAnchors,         transferParams\\$weight.reduction <- CreateDimReducObject(embeddings = rDSub, ,             key = \"LSI\\_\", assay = DefaultAssay(seuratATAC)),         transferParams\\$verbose <- FALSE,         transferParams\\$dims <- seq\\_len(ncol(rDSub)),         .logDiffTime(sprintf(\"\\%s Seurat TransferData Cell Group Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams\\$refdata <- subRNA\\$Group,         rnaLabels <- do.call(Seurat::TransferData, transferParams),         .logDiffTime(sprintf(\"\\%s Seurat TransferData Cell Names Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams\\$refdata <- colnames(subRNA),         rnaLabels2 <- do.call(Seurat::TransferData, transferParams){[}, ,             1{]},         if (addToArrow) \\{,             .logDiffTime(sprintf(\"\\%s Seurat TransferData GeneMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             transferParams\\$refdata <- GetAssayData(subRNA, assay = \"RNA\", ,                 slot = \"data\"),             gc(),             matchedRNA <- do.call(Seurat::TransferData, transferParams),             matchedRNA <- matchedRNA@data,         \\},         matchDF <- DataFrame(cellNames = colnames(seuratATAC), ,             predictionScore = rnaLabels\\$prediction.score.max, ,             predictedGroup = rnaLabels\\$predicted.id, predictedCell = rnaLabels2),         rownames(matchDF) <- matchDF\\$cellNames,         .logDiffTime(sprintf(\"\\%s Saving TransferAnchors Joint CCA\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         jointCCA <- DataFrame(transferAnchors@object.list{[}{[}1{]}{]}@reductions\\$cca@cell.embeddings),         jointCCA\\$Assay <- ifelse(endsWith(rownames(jointCCA), ,             \"\\_reference\"), \"RNA\", \"ATAC\"),         jointCCA\\$Group <- NA,         jointCCA\\$Score <- NA,         jointCCA{[}paste0(colnames(subRNA), \"\\_reference\"), \"Group\"{]} <- subRNA\\$Group,         jointCCA{[}paste0(matchDF\\$cellNames, \"\\_query\"), \"Group\"{]} <- matchDF\\$predictedGroup,         jointCCA{[}paste0(matchDF\\$cellNames, \"\\_query\"), \"Score\"{]} <- matchDF\\$predictionScore,         .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,             paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),         rm(transferParams, transferAnchors),         gc(),         if (addToArrow) \\{,             .logDiffTime(sprintf(\"\\%s Transferring Paired RNA to Temp File\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             tmpFilei <- paste0(tmpFile, \"-IntegrationBlock-\", ,                 i, \".h5\"),             o <- h5createFile(tmpFilei),             sampleNames <- getCellColData(subProj, \"Sample\"){[}matchDF\\$cellNames, ,                 {]},             uniqueSamples <- unique(sampleNames),             matchedRNA <- .safeSubset(mat = matchedRNA, subsetRows = paste0(featureDF\\$name), ,                 subsetCols = matchDF\\$cellNames),             for (z in seq\\_along(uniqueSamples)) \\{,                 mat <- matchedRNA{[}, which(sampleNames == uniqueSamples{[}z{]}), ,                   drop = FALSE{]},                 Group <- uniqueSamples{[}z{]},                 o <- tryCatch(\\{,                   h5delete(tmpFilei, paste0(Group)),                 \\}, error = function(x) \\{,                 \\}),                 o <- h5createGroup(tmpFilei, paste0(Group)),                 j <- Rle(findInterval(seq(mat@x) - 1, mat@p{[}-1{]}) + ,                   1),                 lengthRle <- length(j@lengths),                 lengthI <- length(mat@i),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/i\"), storage.mode = \"integer\", dims = c(lengthI, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jLengths\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jValues\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/x\"), storage.mode = \"double\", dims = c(lengthI, ,                   1), level = 0)),                 o <- .suppressAll(h5write(obj = mat@i + 1, file = tmpFilei, ,                   name = paste0(Group, \"/i\"))),                 o <- .suppressAll(h5write(obj = j@lengths, file = tmpFilei, ,                   name = paste0(Group, \"/jLengths\"))),                 o <- .suppressAll(h5write(obj = j@values, file = tmpFilei, ,                   name = paste0(Group, \"/jValues\"))),                 o <- .suppressAll(h5write(obj = mat@x, file = tmpFilei, ,                   name = paste0(Group, \"/x\"))),                 o <- .suppressAll(h5write(obj = colnames(mat), ,                   file = tmpFilei, name = paste0(Group, \"/cellNames\"))),             \\},             rm(matchedRNA, mat, j),         \\},         .logDiffTime(sprintf(\"\\%s Completed Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         gc(),         matchDF\\$Block <- Rle(i),         matchDF,     \\}, threads = threads2) \\%>\\% Reduce(\"rbind\", .),     if (plotUMAP) \\{,         for (i in seq\\_along(blockList)) \\{,             o <- tryCatch(\\{,                 prefix <- sprintf(\"Block (\\%s of \\%s) :\", i, length(blockList)),                 .logDiffTime(sprintf(\"\\%s Plotting Joint UMAP\", ,                   prefix), tstart, verbose = verbose, logFile = logFile),                 jointCCA <- readRDS(file.path(outDir3, paste0(\"Save-Block\", ,                   i, \"-JointCCA.rds\"))),                 set.seed(1),                 UMAPParams <- .mergeParams(UMAPParams, list(n\\_neighbors = 40, ,                   min\\_dist = 0.4, metric = \"cosine\", verbose = FALSE)),                 UMAPParams\\$X <- as.data.frame(jointCCA{[}, grep(\"CC\\_\", ,                   colnames(jointCCA)){]}),                 UMAPParams\\$ret\\_nn <- FALSE,                 UMAPParams\\$ret\\_model <- FALSE,                 UMAPParams\\$n\\_threads <- 1,                 uwotUmap <- tryCatch(\\{,                   do.call(uwot::umap, UMAPParams),                 \\}, error = function(e) \\{,                   errorList <- UMAPParams,                   .logError(e, fn = \"uwot::umap\", info = prefix, ,                     errorList = errorList, logFile = logFile),                 \\}),                 jointCCA\\$UMAP1 <- uwotUmap{[}, 1{]},                 jointCCA\\$UMAP2 <- uwotUmap{[}, 2{]},                 .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,                   paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),                 p1 <- ggPoint(x = uwotUmap{[}, 1{]}, y = uwotUmap{[}, ,                   2{]}, color = jointCCA\\$Assay, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by Assay\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element\\_blank(), ,                   axis.ticks.x = element\\_blank(), axis.text.y = element\\_blank(), ,                   axis.ticks.y = element\\_blank()),                 p2 <- ggPoint(x = uwotUmap{[}, 1{]}, y = uwotUmap{[}, ,                   2{]}, color = jointCCA\\$Group, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by scRNA Group\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element\\_blank(), ,                   axis.ticks.x = element\\_blank(), axis.text.y = element\\_blank(), ,                   axis.ticks.y = element\\_blank()),                 pdf(file.path(outDir3, paste0(\"Save-Block\", i, ,                   \"-JointCCA-UMAP.pdf\")), width = 12, height = 6, ,                   useDingbats = FALSE),                 ggAlignPlots(p1, p2, type = \"h\"),                 dev.off(),             \\}, error = function(e) \\{,             \\}),         \\},     \\},     if (addToArrow) \\{,         .logDiffTime(\"Transferring Data to ArrowFiles\", tstart, ,             verbose = verbose, logFile = logFile),         matrixName <- .isProtectedArray(matrixName),         integrationFiles <- paste0(tmpFile, \"-IntegrationBlock-\", ,             seq\\_along(blockList), \".h5\"),         if (!all(file.exists(integrationFiles))) \\{,             .logMessage(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\", ,                 logFile = logFile),             stop(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\"),         \\},         h5list <- .safelapply(seq\\_along(integrationFiles), function(x) \\{,             h5ls(integrationFiles{[}x{]}),         \\}, threads = threads),         ArrowFiles <- getArrowFiles(ArchRProj),         allSamples <- names(ArrowFiles),         o <- .safelapply(seq\\_along(allSamples), function(y) \\{,             sample <- allSamples{[}y{]},             prefix <- sprintf(\"\\%s (\\%s of \\%s)\", sample, y, length(ArrowFiles)),             .logDiffTime(sprintf(\"\\%s Getting GeneIntegrationMatrix From TempFiles!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             sampleIF <- lapply(seq\\_along(h5list), function(x) \\{,                 if (any(h5list{[}{[}x{]}{]}\\$group == paste0(\"/\", sample))) \\{,                   integrationFiles{[}x{]},                 \\},                 else \\{,                   NULL,                 \\},             \\}) \\%>\\% unlist,             sampleMat <- lapply(seq\\_along(sampleIF), function(x) \\{,                 cellNames <- .h5read(sampleIF{[}x{]}, paste0(sample, ,                   \"/cellNames\")),                 mat <- sparseMatrix(i = .h5read(sampleIF{[}x{]}, ,                   paste0(sample, \"/i\")){[}, 1{]}, j = as.vector(Rle(.h5read(sampleIF{[}x{]}, ,                   paste0(sample, \"/jValues\")){[}, 1{]}, .h5read(sampleIF{[}x{]}, ,                   paste0(sample, \"/jLengths\")){[}, 1{]})), x = .h5read(sampleIF{[}x{]}, ,                   paste0(sample, \"/x\")){[}, 1{]}, dims = c(nrow(featureDF), ,                   length(cellNames))),                 colnames(mat) <- cellNames,                 mat,             \\}) \\%>\\% Reduce(\"cbind\", .),             sampleMat@x <- exp(sampleMat@x) - 1,             sampleMat <- .normalizeCols(sampleMat, scaleTo = scaleTo),             sampleMat <- drop0(sampleMat),             rownames(sampleMat) <- paste0(featureDF\\$name),             sampleMat <- sampleMat{[}, ArchRProj\\$cellNames{[}BiocGenerics::which(ArchRProj\\$Sample == ,                 sample){]}, drop = FALSE{]},             o <- .createArrowGroup(ArrowFile = ArrowFiles{[}sample{]}, ,                 group = matrixName, force = force),             o <- .initializeMat(ArrowFile = ArrowFiles{[}sample{]}, ,                 Group = matrixName, Class = \"double\", Units = \"NormCounts\", ,                 cellNames = colnames(sampleMat), params = dfParams, ,                 featureDF = featureDF, force = force),             o <- h5write(obj = dfAll{[}colnames(sampleMat), \"predictionScore\"{]}, ,                 file = ArrowFiles{[}sample{]}, name = paste0(matrixName, ,                   \"/Info/predictionScore\")),             o <- h5write(obj = dfAll{[}colnames(sampleMat), \"predictedGroup\"{]}, ,                 file = ArrowFiles{[}sample{]}, name = paste0(matrixName, ,                   \"/Info/predictedGroup\")),             o <- h5write(obj = dfAll{[}colnames(sampleMat), \"predictedCell\"{]}, ,                 file = ArrowFiles{[}sample{]}, name = paste0(matrixName, ,                   \"/Info/predictedCell\")),             .logDiffTime(sprintf(\"\\%s Adding GeneIntegrationMatrix to ArrowFile!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             for (z in seq\\_along(allChr)) \\{,                 chrz <- allChr{[}z{]},                 .logDiffTime(sprintf(\"Adding GeneIntegrationMatrix to \\%s for Chr (\\%s of \\%s)!\", ,                   sample, z, length(allChr)), tstart, verbose = FALSE, ,                   logFile = logFile),                 idz <- BiocGenerics::which(featureDF\\$seqnames \\%bcin\\% ,                   chrz),                 matz <- sampleMat{[}idz, , drop = FALSE{]},                 stopifnot(identical(paste0(featureDF\\$name{[}idz{]}), ,                   paste0(rownames(matz)))),                 o <- .addMatToArrow(mat = matz, ArrowFile = ArrowFiles{[}sample{]}, ,                   Group = paste0(matrixName, \"/\", chrz), binarize = FALSE, ,                   addColSums = TRUE, addRowSums = TRUE, addRowVarsLog2 = TRUE, ,                   logFile = logFile),                 rm(matz),                 if (z\\%\\%3 == 0 \\textbar{} z == length(allChr)) \\{,                   gc(),                 \\},             \\},             0,         \\}, threads = threads),         o <- suppressWarnings(file.remove(integrationFiles)),     \\},     .logDiffTime(\"Completed Integration with RNA Matrix\", tstart, ,         verbose = verbose, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll\\$cellNames, ,         data = dfAll\\$predictedCell, name = nameCell, force = TRUE),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll\\$cellNames, ,         data = dfAll\\$predictedGroup, name = nameGroup, force = TRUE),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll\\$cellNames, ,         data = dfAll\\$predictionScore, name = nameScore, force = TRUE),     .endLogging(logFile = logFile),     return(ArchRProj), \\} & function (ArchRProj = NULL, useSeqnames = NULL, verbose = TRUE, ,     binarize = FALSE, threads = getArchRThreads(), logFile = createLogFile(\"exportMatrixForSTREAM\")) , \\{,     mat <- getMatrixFromProject(ArchRProj = ArchRProj, useMatrix = \"PeakMatrix\", ,         useSeqnames = useSeqnames, verbose = verbose, binarize = binarize, ,         threads = threads, logFile = logFile),     featureDF <- ArchR:::.getFeatureDF(getArrowFiles(ArchRProj){[}1{]}, ,         \"PeakMatrix\"),     stopifnot(all(featureDF\\$idx == rowData(mat)\\$idx)),     countsDF <- Matrix::summary(assay(mat)),     peaksDF <- data.frame(as.vector(featureDF{[}, 1{]}), featureDF{[}, ,         3{]}, featureDF{[}, 4{]}),     cellsDF <- data.frame(colnames(mat)),     data.table::fwrite(countsDF, file = \"STREAM\\_Counts.tsv.gz\", ,         sep = \"\\textbackslash{}t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(peaksDF, file = \"STREAM\\_Regions.tsv.gz\", ,         sep = \"\\textbackslash{}t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(cellsDF, file = \"STREAM\\_Sample.tsv.gz\", ,         sep = \"\\textbackslash{}t\", row.names = FALSE, col.names = FALSE),     return(0), \\} & function (ArchRProj = NULL) , \\{,     if (!inherits(ArchRProj, \"ArchRProject\")) \\{,         stop(\"Not a valid ArchRProject as input!\"),     \\},     else \\{,         ArchRProj,     \\}, \\} & function (p = NULL, plotWidth = unit(6, \"in\"), plotHeight = unit(6, ,     \"in\"), margin = 0.25, height = 1, it = 0.05, newPage = FALSE) , \\{,     .requirePackage(\"grid\", source = \"cran\"),     .requirePackage(\"gridExtra\", source = \"cran\"),     if (!inherits(plotWidth, \"unit\")) \\{,         plotWidth <- unit(plotWidth, \"in\"),     \\},     if (!inherits(plotHeight, \"unit\")) \\{,         plotHeight <- unit(plotHeight, \"in\"),     \\},     g <- ggplotGrob(p),     legend <- grep(\"guide-box\", g\\$layout\\$name),     if (length(legend) != 0) \\{,         gl <- g\\$grobs{[}{[}legend{]}{]},         g <- ggplotGrob(p + theme(legend.position = \"none\")),     \\},     else \\{,         gl <- NULL,         g <- ggplotGrob(p),     \\},     panels <- grep(\"panel\", g\\$layout\\$name),     panel\\_index\\_w <- unique(g\\$layout\\$l{[}panels{]}),     panel\\_index\\_h <- unique(g\\$layout\\$t{[}panels{]}),     nw <- length(panel\\_index\\_w),     nh <- length(panel\\_index\\_h),     pw <- convertWidth(plotWidth, unitTo = \"in\", valueOnly = TRUE),     ph <- convertWidth(plotHeight, unitTo = \"in\", valueOnly = TRUE),     pw <- pw * 0.95,     ph <- ph * 0.95,     x <- 0,     width <- 1,     sm <- FALSE,     while (!sm) \\{,         x <- x + it,         w <- unit(x * width, \"in\"),         h <- unit(x * height/width, \"in\"),         m <- unit(x * margin/width, \"in\"),         g\\$widths{[}panel\\_index\\_w{]} <- rep(w, nw),         g\\$heights{[}panel\\_index\\_h{]} <- rep(h, nh),         sw <- convertWidth(x = sum(g\\$widths) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sh <- convertHeight(x = sum(g\\$heights) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sm <- sw > pw \\textbar{} sh > ph,     \\},     if (length(legend) != 0) \\{,         sgh <- convertHeight(x = sum(g\\$heights), unitTo = \"in\", ,             valueOnly = TRUE),         sgw <- convertWidth(x = sum(g\\$widths), unitTo = \"in\", ,             valueOnly = TRUE),         slh <- convertHeight(x = sum(gl\\$heights), unitTo = \"in\", ,             valueOnly = TRUE),         slw <- convertWidth(x = sum(gl\\$widths), unitTo = \"in\", ,             valueOnly = TRUE),         size <- 6,         wh <- 0.1,         it <- 0,         while (slh > 0.2 * ph \\textbar{} slw > pw) \\{,             it <- it + 1,             if (it > 3) \\{,                 break,             \\},             size <- size * 0.8,             wh <- wh * 0.8,             gl <- ggplotGrob(p + theme(legend.key.width = unit(wh, ,                 \"cm\"), legend.key.height = unit(wh, \"cm\"), legend.spacing.x = unit(0, ,                 \"cm\"), legend.spacing.y = unit(0, \"cm\"), legend.text = element\\_text(size = max(size, ,                 2))) + guides(fill = guide\\_legend(ncol = 4), ,                 color = guide\\_legend(ncol = 4)))\\$grobs{[}{[}legend{]}{]},             slh <- convertHeight(x = sum(gl\\$heights), unitTo = \"in\", ,                 valueOnly = TRUE),             slw <- convertWidth(x = sum(gl\\$widths), unitTo = \"in\", ,                 valueOnly = TRUE),         \\},         p <- grid.arrange(g, gl, ncol = 1, nrow = 2, heights = unit.c(unit(sgh, ,             \"in\"), unit(min(slh, 0.2 * pw), \"in\")), newpage = newPage),     \\},     else \\{,         p <- grid.arrange(g, newpage = newPage),     \\},     invisible(p), \\}\\\\\n",
       "\tvisible & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & ⋯ & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 2 × 55\n",
       "\n",
       "| <!--/--> | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AllClasses.R &lt;chr&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationGenome.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationPeaks.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-CRHR1.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-FKBP5.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser3.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-CAMKV.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-KCNIP4.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-xiangyu.R &lt;lgl&gt; | ⋯ ⋯ | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/MultiModal.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/plot_footprints.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ProjectMethods.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/QualityControl.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RcppExports.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ReproduciblePeakSet.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RNAIntegration.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/Trajectory.R &lt;lgl&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ValidationUtils.R &lt;fn&gt; | /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/VisualizeData.R &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| value | rownames | function (genome = NULL) , {,     .validInput(input = genome, name = \"genome\", valid = \"character\"),     if (toupper(genome) == \"HG19\" | toupper(genome) == \"HG38\") {,         if (suppressWarnings(!require(org.Hs.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Hs.eg.db\", update = FALSE),         },         library(org.Hs.eg.db),         annodb &lt;- org.Hs.eg.db,     },     else if (toupper(genome) == \"MM9\" | toupper(genome) == \"MM10\") {,         if (suppressWarnings(!require(org.Mm.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Mm.eg.db\", update = FALSE),         },         library(org.Mm.eg.db),         annodb &lt;- org.Mm.eg.db,     },     else {,         stop(\"Genome not recognized!\"),     },     return(annodb), } | function (seEnrich = NULL, pal = paletteContinuous(set = \"comet\", ,     n = 100), n = 10, cutOff = 20, pMax = Inf, clusterCols = TRUE, ,     binaryClusterRows = TRUE, labelRows = TRUE, rastr = TRUE, ,     transpose = FALSE, returnMatrix = FALSE, logFile = createLogFile(\"plotEnrichHeatmap\")) , {,     .validInput(input = seEnrich, name = \"seEnrich\", valid = c(\"SummarizedExperiment\")),     .validInput(input = pal, name = \"pal\", valid = c(\"character\")),     .validInput(input = n, name = \"n\", valid = c(\"integer\")),     .validInput(input = cutOff, name = \"cutOff\", valid = c(\"numeric\")),     .validInput(input = pMax, name = \"pMax\", valid = c(\"numeric\")),     .validInput(input = clusterCols, name = \"clusterCols\", valid = c(\"boolean\")),     .validInput(input = binaryClusterRows, name = \"binaryClusterRows\", ,         valid = c(\"boolean\")),     .validInput(input = labelRows, name = \"labelRows\", valid = c(\"boolean\")),     .validInput(input = rastr, name = \"rastr\", valid = c(\"boolean\")),     .validInput(input = transpose, name = \"transpose\", valid = c(\"boolean\")),     .validInput(input = returnMatrix, name = \"returnMatrix\", ,         valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotEnrichHeatmap Input-Parameters\", logFile = logFile),     mat &lt;- assays(seEnrich)[[\"mlog10Padj\"]],     .logThis(mat, \"mat-mlog10Padj\", logFile = logFile),     keep &lt;- lapply(seq_len(ncol(mat)), function(x) {,         idx &lt;- head(order(mat[, x], decreasing = TRUE), n),         rownames(mat)[idx[which(mat[idx, x] &gt; cutOff)]],     }) %&gt;% unlist %&gt;% unique,     mat &lt;- mat[keep, , drop = FALSE],     .logThis(mat, \"mat-mlog10Padj-Filter\", logFile = logFile),     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     passMat &lt;- lapply(seq_len(nrow(mat)), function(x) {,         (mat[x, ] &gt;= 0.9 * max(mat[x, ])) * 1,     }) %&gt;% Reduce(\"rbind\", .) %&gt;% data.frame,     colnames(passMat) &lt;- colnames(mat),     .logThis(passMat, \"passMat\", logFile = logFile),     mat[mat &gt; pMax] &lt;- pMax,     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     mat &lt;- .rowScale(as.matrix(mat), min = 0),     if (pMax != 100) {,         rownames(mat[[1]]) &lt;- paste0(rownames(mat[[1]]), \" (\", ,             round(mat$max), \")\"),         rownames(passMat) &lt;- rownames(mat[[1]]),     },     .logThis(mat, \"mat-mlog10Padj-RowScale\", logFile = logFile),     mat2 &lt;- mat[[1]] * 100,     if (nrow(mat2) &gt; 1 &amp; ncol(mat2) &gt; 1) {,         if (binaryClusterRows) {,             bS &lt;- .binarySort(mat2, lmat = passMat[rownames(mat2), ,                 colnames(mat2)], clusterCols = clusterCols),             mat2 &lt;- bS[[1]][, colnames(mat2)],             clusterRows &lt;- FALSE,             clusterCols &lt;- bS[[2]],         },         else {,             clusterRows &lt;- TRUE,         },     },     else {,         clusterCols &lt;- NULL,         clusterRows &lt;- FALSE,     },     if (nrow(mat2) &gt; 100) {,         borderColor &lt;- FALSE,     },     else {,         borderColor &lt;- TRUE,     },     .logThis(mat2, \"mat-mlog10Padj-RowScale-Plot\", logFile = logFile),     if (transpose) {,         if (!is.null(clusterCols)) {,             mat2 &lt;- t(mat2[, clusterCols$order, drop = FALSE]),         },         else {,             mat2 &lt;- t(mat2),         },         if (returnMatrix) {,             return(mat2),         },         ht &lt;- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     else {,         if (returnMatrix) {,             return(mat2),         },         ht &lt;- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     return(ht), } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal &lt;- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title &lt;- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title &lt;- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot &lt;- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx &lt;- which(tolower(names(plotList)) == \"bulktrack\"),         p &lt;- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx &lt;- which(tolower(names(plotList)) == \"sctrack\"),         p &lt;- plotList[[idx]] + featurePlot,         plotList[idx] &lt;- NULL,         for (i in seq_along(plotList)) {,             p &lt;- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p &lt;- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, } | ⋯ | function (ArchRProj = NULL, name = \"CombinedDims\", reducedDims = NULL, ,     dimWeights = NULL, dimsToUse = NULL, scaleDims = NULL, corCutOff = 0.75) , {,     if (is.null(dimWeights)) {,         dimWeights &lt;- rep(1, length(reducedDims)),     },     combinedDims &lt;- lapply(seq_along(reducedDims), function(x) {,         rD &lt;- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims[x], ,             dimsToUse = dimsToUse, scaleDims = scaleDims, corCutOff = corCutOff),         cV &lt;- apply(as.matrix(rD), 2, var),         normV &lt;- 1/sqrt(sum(cV)),         rD * normV * dimWeights[x],     }) %&gt;% Reduce(\"cbind\", .),     ArchRProj@reducedDims[[name]] &lt;- SimpleList(matRD = combinedDims, ,         scaleDims = NA, corToDepth = list(scaled = rep(0, ncol(combinedDims)), ,             none = rep(0, ncol(combinedDims)))),     ArchRProj, } | function (seFoot = NULL, names = NULL, pal = NULL, flank = 250, ,     flankNorm = 50, normMethod = \"Subtract\", smoothWindow = NULL, ,     baseSize = 6, plot = TRUE, ArchRProj = NULL, plotName = paste0(\"Plot-Footprints-\", ,         normMethod), height = 6, width = 4, addDOC = TRUE, force = FALSE, ,     logFile = createLogFile(\"plotFootprints\")) , {,     tstart &lt;- Sys.time(),     if (is.null(names)) {,         names &lt;- names(assays(seFoot)),     },     if (plot) {,         name &lt;- gsub(\"\\\\.pdf\", \"\", plotName),         if (is.null(ArchRProj)) {,             outDir &lt;- \"Plots\",         },         else {,             outDir &lt;- file.path(getOutputDirectory(ArchRProj), ,                 \"Plots\"),         },         dir.create(outDir, showWarnings = FALSE),         if (addDOC) {,             doc &lt;- gsub(\":\", \"-\", stringr::str_split(Sys.time(), ,                 pattern = \" \", simplify = TRUE)[1, 2]),             filename &lt;- file.path(outDir, paste0(name, \"_Date-\", ,                 Sys.Date(), \"_Time-\", doc, \".pdf\")),         },         else {,             filename &lt;- file.path(outDir, paste0(name, \".pdf\")),         },         pdf(filename, width = width, height = height, useDingbats = FALSE),     },     ggList &lt;- lapply(seq_along(names), function(x) {,         gg &lt;- .ggFootprint(seFoot = seFoot, name = names[x], ,             pal = pal, smoothWindow = smoothWindow, flank = flank, ,             flankNorm = flankNorm, baseSize = baseSize, normMethod = normMethod, ,             logFile = logFile),         if (plot) {,             if (x != 1) {,                 grid::grid.newpage(),             },             grid::grid.draw(gg),             return(0),         },         else {,             return(gg),         },     }),     if (!plot) {,         names(ggList) &lt;- names,         ggList,     },     else {,         dev.off(),         return(invisible(0)),     }, } | function (ArchRProj = NULL, features = NULL, name = NULL, addRatio = TRUE, ,     threads = getArchRThreads(), logFile = createLogFile(\"addFeatureCounts\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = features, name = \"features\", valid = c(\"granges\")),     .validInput(input = name, name = \"name\", valid = c(\"character\")),     .validInput(input = addRatio, name = \"addRatio\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     ArrowFiles &lt;- getArrowFiles(ArchRProj),     cellNames &lt;- ArchRProj$cellNames,     featuresList &lt;- split(features, seqnames(features)),     h5disableFileLocking(),     countsDF &lt;- .safelapply(seq_along(featuresList), function(i) {,         chri &lt;- names(featuresList)[i],         cellTotal &lt;- rep(0, length(cellNames)),         names(cellTotal) &lt;- cellNames,         featuresi &lt;- ranges(featuresList[[i]]),         .logDiffTime(paste0(\"Counting in \", chri, \" (\", i, \" of \", ,             length(featuresList), \")\"), t1 = tstart, logFile = logFile),         for (j in seq_along(ArrowFiles)) {,             fragmentsij &lt;- .getFragsFromArrow(ArrowFile = ArrowFiles[j], ,                 chr = chri, out = \"IRanges\", cellNames = cellNames),             if (length(fragmentsij) &gt; 0) {,                 mcols(fragmentsij)$RG@values &lt;- match(mcols(fragmentsij)$RG@values, ,                   cellNames),                 for (y in seq_len(2)) {,                   if (y == 1) {,                     temp &lt;- IRanges(start(fragmentsij), width = 1),                   },                   else if (y == 2) {,                     temp &lt;- IRanges(end(fragmentsij), width = 1),                   },                   stopifnot(length(temp) == length(fragmentsij)),                   tabSum &lt;- S4Vectors:::tabulate(mcols(fragmentsij)$RG[queryHits(findOverlaps(temp, ,                     featuresi))]),                   cellTotal[seq_along(tabSum)] &lt;- cellTotal[seq_along(tabSum)] + ,                     tabSum,                 },             },             if (j%%3 == 0) {,                 gc(),             },         },         cellTotal,     }, threads = threads) %&gt;% Reduce(\"rbind\", .),     totalCounts &lt;- colSums(countsDF),     countRatio &lt;- totalCounts/(2 * ArchRProj$nFrags),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Counts\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj, data = totalCounts, ,         cells = names(totalCounts), name = paste0(name, \"Counts\"), ,         force = TRUE),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Ratio\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj, data = countRatio, ,         cells = names(totalCounts), name = paste0(name, \"Ratio\"), ,         force = TRUE),     ArchRProj, } | function (ArchRProj = NULL, groupBy = \"Sample\", chromSizes = getChromSizes(ArchRProj), ,     maxSize = 750, pal = NULL, returnDF = FALSE, threads = getArchRThreads(), ,     logFile = createLogFile(\"plotFragmentSizes\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = maxSize, name = \"maxSize\", valid = c(\"integer\")),     .validInput(input = returnDF, name = \"returnDF\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotFragmentSizes Input-Parameters\", logFile = logFile),     chr &lt;- paste0(seqnames(chromSizes)),     groups &lt;- getCellColData(ArchRProj = ArchRProj, select = groupBy, ,         drop = FALSE),     uniqGroups &lt;- gtools::mixedsort(unique(groups[, 1])),     if (threads &gt; 1) {,         h5disableFileLocking(),     },     dfFS &lt;- .safelapply(seq_along(uniqGroups), function(x) {,         .logDiffTime(paste0(uniqGroups[x], \" Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         cellx &lt;- rownames(groups)[which(paste0(groups[, 1]) == ,             uniqGroups[x])],         for (i in seq_along(chr)) {,             if (i == 1) {,                 fsi &lt;- unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %&gt;% width %&gt;% tabulate(nbins = maxSize),             },             else {,                 fsi &lt;- fsi + unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %&gt;% width %&gt;% tabulate(nbins = maxSize),             },             .logThis(fsi, paste0(uniqGroups[x], \" : FragSizes : \", ,                 chr[i]), logFile = logFile),         },         df &lt;- DataFrame(group = uniqGroups[x], fragmentSize = seq_along(fsi), ,             fragmentPercent = round(100 * fsi/sum(fsi), 4)),         .logThis(df, paste0(uniqGroups[x], \" : Frag DF\"), logFile = logFile),         .logDiffTime(paste0(uniqGroups[x], \" Finished Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         df,     }, threads = threads) %&gt;% Reduce(\"rbind\", .),     .logThis(dfFS, paste0(\"All : FragSizes DF\"), logFile = logFile),     .endLogging(logFile = logFile),     if (threads &gt; 1) {,         h5enableFileLocking(),     },     if (returnDF) {,         return(dfFS),     },     else {,         plotDF &lt;- data.frame(dfFS),         if (is.null(pal)) {,             pal &lt;- paletteDiscrete(values = unique(plotDF$group)),         },         p &lt;- ggplot(plotDF, aes(fragmentSize, fragmentPercent, ,             color = group)) + geom_line(size = 1) + theme_ArchR() + ,             xlab(\"ATAC-seq Fragment Size (bp)\") + ylab(\"Percentage of Fragments\") + ,             scale_color_manual(values = pal) + scale_y_continuous(limits = c(0, ,             max(plotDF$fragmentPercent) * 1.05), expand = c(0, ,             0)) + scale_x_continuous(limits = c(min(plotDF$fragmentSize), ,             max(plotDF$fragmentSize)), expand = c(0, 0)),         p,     }, } | function (string_vector, id_vector, n_id, window, w, kmer) , {,     .Call(\"_ArchR_kmerIDFrequencyCpp\", PACKAGE = \"ArchR\", string_vector, ,         id_vector, n_id, window, w, kmer), } | function () , {,     message(\"Searching For MACS2..\"),     if (.suppressAll(.checkPath(\"macs2\", throwError = FALSE))) {,         message(\"Found with $path!\"),         return(\"macs2\"),     },     message(\"Not Found in $PATH\"),     search2 &lt;- suppressWarnings(tryCatch({,         system2(\"pip\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     search3 &lt;- suppressWarnings(tryCatch({,         system2(\"pip3\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     if (length(search2) &gt; 0) {,         if (search2[1] != \"ERROR\") {,             path2Install &lt;- gsub(\"Location: \", \"\", search2[grep(\"Location\", ,                 search2, ignore.case = TRUE)]),             path2Bin &lt;- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip\"),     if (length(search3) &gt; 0) {,         if (search3[1] != \"ERROR\") {,             path2Install &lt;- gsub(\"Location: \", \"\", search3[grep(\"Location\", ,                 search3, ignore.case = TRUE)]),             path2Bin &lt;- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip3!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip3\"),     stop(\"Could Not Find Macs2! Please install w/ pip, add to your $PATH, or just supply the macs2 path directly and avoid this function!\"), } | function (ArchRProj = NULL, useMatrix = \"GeneScoreMatrix\", matrixName = \"GeneIntegrationMatrix\", ,     reducedDims = \"IterativeLSI\", seRNA = NULL, groupATAC = NULL, ,     groupRNA = NULL, groupList = NULL, sampleCellsATAC = 10000, ,     sampleCellsRNA = 10000, embeddingATAC = NULL, embeddingRNA = NULL, ,     dimsToUse = 1:30, scaleDims = NULL, corCutOff = 0.75, plotUMAP = TRUE, ,     UMAPParams = list(n_neighbors = 40, min_dist = 0.4, metric = \"cosine\", ,         verbose = FALSE), nGenes = 2000, useImputation = TRUE, ,     reduction = \"cca\", addToArrow = TRUE, scaleTo = 10000, genesUse = NULL, ,     nameCell = \"predictedCell\", nameGroup = \"predictedGroup\", ,     nameScore = \"predictedScore\", transferParams = list(), threads = getArchRThreads(), ,     verbose = TRUE, force = FALSE, logFile = createLogFile(\"addGeneIntegrationMatrix\"), ,     ...) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = useMatrix, name = \"useMatrix\", valid = c(\"character\")),     .validInput(input = matrixName, name = \"matrixName\", valid = c(\"character\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = seRNA, name = \"seRNA\", valid = c(\"SummarizedExperiment\", ,         \"Seurat\")),     .validInput(input = groupATAC, name = \"groupATAC\", valid = c(\"character\", ,         \"null\")),     .validInput(input = groupRNA, name = \"groupRNA\", valid = c(\"character\")),     .validInput(input = groupList, name = \"groupList\", valid = c(\"list\", ,         \"null\")),     .validInput(input = sampleCellsATAC, name = \"sampleCellsATAC\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = sampleCellsRNA, name = \"sampleCellsRNA\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = embeddingATAC, name = \"embeddingATAC\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = embeddingRNA, name = \"embeddingRNA\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = dimsToUse, name = \"dimsToUse\", valid = c(\"numeric\", ,         \"null\")),     .validInput(input = scaleDims, name = \"scaleDims\", valid = c(\"boolean\", ,         \"null\")),     .validInput(input = plotUMAP, name = \"plotUMAP\", valid = c(\"boolean\")),     .validInput(input = UMAPParams, name = \"UMAPParams\", valid = c(\"list\")),     .validInput(input = nGenes, name = \"nGenes\", valid = c(\"integer\")),     .validInput(input = useImputation, name = \"useImputation\", ,         valid = c(\"boolean\")),     .validInput(input = reduction, name = \"reduction\", valid = c(\"character\")),     .validInput(input = addToArrow, name = \"addToArrow\", valid = c(\"boolean\")),     .validInput(input = scaleTo, name = \"scaleTo\", valid = c(\"numeric\")),     .validInput(input = genesUse, name = \"genesUse\", valid = c(\"character\", ,         \"null\")),     .validInput(input = nameCell, name = \"nameCell\", valid = c(\"character\")),     .validInput(input = nameGroup, name = \"nameGroup\", valid = c(\"character\")),     .validInput(input = nameScore, name = \"nameScore\", valid = c(\"character\")),     .validInput(input = transferParams, name = \"transferParams\", ,         valid = c(\"list\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = verbose, name = \"verbose\", valid = c(\"boolean\")),     .validInput(input = force, name = \"force\", valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart &lt;- Sys.time(),     .startLogging(logFile = logFile),     .logDiffTime(\"Running Seurat's Integration Stuart* et al 2019\", ,         tstart, verbose = verbose, logFile = logFile),     .requirePackage(\"Seurat\", source = \"cran\"),     .logThis(append(args, mget(names(formals()), sys.frame(sys.nframe()))), ,         \"Input-Parameters\", logFile = logFile),     if (is.null(groupList)) {,         groupList &lt;- SimpleList(),         groupList[[1]] &lt;- SimpleList(ATAC = ArchRProj$cellNames, ,             RNA = colnames(seRNA)),     },     .logDiffTime(\"Checking ATAC Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (useMatrix %ni% getAvailableMatrices(ArchRProj)) {,         .logMessage(paste0(\"Matrix \", useMatrix, \" does not exist in the provided ArchRProject. See available matrix names from getAvailableMatrices()!\"), ,             logFile = logFile),         stop(\"Matrix name provided to useMatrix does not exist in ArchRProject!\"),     },     if (!is.null(groupATAC)) {,         dfATAC &lt;- getCellColData(ArchRProj = ArchRProj, select = groupATAC, ,             drop = FALSE),     },     nCell &lt;- rep(0, length(ArchRProj$cellNames)),     names(nCell) &lt;- ArchRProj$cellNames,     groupList &lt;- lapply(seq_along(groupList), function(x) {,         ATAC &lt;- groupList[[x]]$ATAC,         if (!is.null(groupATAC)) {,             if (any(ATAC %in% dfATAC[, 1])) {,                 idx &lt;- which(ATAC %in% dfATAC[, 1]),                 ATAC2 &lt;- rownames(dfATAC)[which(dfATAC[, 1] %in% ,                   ATAC[idx])],                 if (length(idx) == length(ATAC)) {,                   ATAC &lt;- ATAC2,                 },                 else {,                   ATAC &lt;- c(ATAC[-idx], ATAC2),                 },             },         },         SimpleList(ATAC = ATAC, RNA = groupList[[x]]$RNA),     }) %&gt;% SimpleList,     for (i in seq_along(groupList)) {,         nCell[groupList[[i]]$ATAC] &lt;- nCell[groupList[[i]]$ATAC] + ,             1,     },     if (!all(nCell == 1)) {,         .logMessage(paste0(\"Missing \", length(which(nCell == ,             0)), \" cells. Found \", length(which(nCell &gt; 1)), ,             \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"), ,             logFile = logFile),         stop(\"Missing \", length(which(nCell == 0)), \" cells. Found \", ,             length(which(nCell &gt; 1)), \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"),     },     .logDiffTime(\"Checking RNA Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (inherits(seRNA, \"SummarizedExperiment\")) {,         seuratRNA &lt;- CreateSeuratObject(counts = assay(seRNA)),         if (groupRNA %ni% colnames(colData(seRNA))) {,             .logMessage(\"groupRNA not in colData of seRNA\", logFile = logFile),             stop(\"groupRNA not in colData of seRNA\"),         },         seuratRNA$Group &lt;- paste0(colData(seRNA)[, groupRNA, ,             drop = TRUE]),         rm(seRNA),     },     else {,         if (groupRNA %ni% colnames(seRNA@meta.data)) {,             .logMessage(\"groupRNA not in meta.data of Seurat Object\", ,                 logFile = logFile),             stop(\"groupRNA not in meta.data of Seurat Object\"),         },         seuratRNA &lt;- seRNA,         seuratRNA$Group &lt;- paste0(seRNA@meta.data[, groupRNA]),         rm(seRNA),     },     if (\"RNA\" %in% names(seuratRNA@assays)) {,         DefaultAssay(seuratRNA) &lt;- \"RNA\",     },     else {,         stop(\"'RNA' is not present in Seurat Object's Assays! Please make sure that this assay is present!\"),     },     gc(),     if (!is.null(groupRNA)) {,         dfRNA &lt;- DataFrame(row.names = colnames(seuratRNA), Group = seuratRNA$Group),     },     groupList &lt;- lapply(seq_along(groupList), function(x) {,         RNA &lt;- groupList[[x]]$RNA,         if (!is.null(groupRNA)) {,             if (any(RNA %in% dfRNA[, 1])) {,                 idx &lt;- which(RNA %in% dfRNA[, 1]),                 RNA2 &lt;- rownames(dfRNA)[which(dfRNA[, 1] %in% ,                   RNA[idx])],                 if (length(idx) == length(RNA)) {,                   RNA &lt;- RNA2,                 },                 else {,                   RNA &lt;- c(RNA[-idx], RNA2),                 },             },         },         SimpleList(ATAC = groupList[[x]]$ATAC, RNA = RNA),     }) %&gt;% SimpleList,     cellRNA &lt;- unlist(lapply(groupList, function(x) x$RNA)),     if (!all(cellRNA %in% colnames(seuratRNA))) {,         .logMessage(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\", ,             logFile = logFile),         stop(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\"),     },     seuratRNA &lt;- seuratRNA[, unique(cellRNA)],     seuratRNA &lt;- NormalizeData(object = seuratRNA, verbose = FALSE),     geneDF &lt;- .getFeatureDF(getArrowFiles(ArchRProj), useMatrix),     sumOverlap &lt;- sum(unique(geneDF$name) %in% unique(rownames(seuratRNA))),     if (sumOverlap &lt; 5) {,         stop(\"Error not enough overlaps (\", sumOverlap, \") between gene names from gene scores (ArchR) and rna matrix (seRNA)!\"),     },     .logDiffTime(paste0(\"Found \", sumOverlap, \" overlapping gene names from gene scores and rna matrix!\"), ,         tstart, verbose = TRUE, logFile = logFile),     .logDiffTime(\"Creating Integration Blocks\", tstart, verbose = verbose, ,         logFile = logFile),     blockList &lt;- SimpleList(),     for (i in seq_along(groupList)) {,         gLi &lt;- groupList[[i]],         if (length(gLi$ATAC) &gt; sampleCellsATAC) {,             if (!is.null(embeddingATAC)) {,                 probATAC &lt;- .getDensity(embeddingATAC[gLi$ATAC, ,                   1], embeddingATAC[gLi$ATAC, 2])$density,                 probATAC &lt;- probATAC/max(probATAC),                 cellsATAC &lt;- gLi$ATAC[order(probATAC, decreasing = TRUE)],             },             else {,                 cellsATAC &lt;- sample(gLi$ATAC, length(gLi$ATAC)),             },             cutoffs &lt;- lapply(seq_len(1000), function(x) length(gLi$ATAC)/x) %&gt;% ,                 unlist,             blockSize &lt;- ceiling(min(cutoffs[order(abs(cutoffs - ,                 sampleCellsATAC))[1]] + 1, length(gLi$ATAC))),             nBlocks &lt;- ceiling(length(gLi$ATAC)/blockSize),             blocks &lt;- lapply(seq_len(nBlocks), function(x) {,                 cellsATAC[seq(x, length(cellsATAC), nBlocks)],             }) %&gt;% SimpleList,         },         else {,             blocks &lt;- list(gLi$ATAC),         },         if (!is.null(embeddingRNA)) {,             probRNA &lt;- .getDensity(embeddingRNA[gLi$RNA, 1], ,                 embeddingRNA[gLi$RNA, 2])$density,             probRNA &lt;- probRNA/max(probRNA),         },         else {,             probRNA &lt;- rep(1, length(gLi$RNA)),         },         blockListi &lt;- lapply(seq_along(blocks), function(x) {,             SimpleList(ATAC = blocks[[x]], RNA = sample(x = gLi$RNA, ,                 size = min(sampleCellsRNA, length(gLi$RNA)), ,                 prob = probRNA)),         }) %&gt;% SimpleList,         blockList &lt;- c(blockList, blockListi),     },     rm(groupList),     .logDiffTime(\"Prepping Interation Data\", tstart, verbose = verbose, ,         logFile = logFile),     subProj &lt;- ArchRProj,     subProj@imputeWeights &lt;- SimpleList(),     geneDF &lt;- .getFeatureDF(getArrowFiles(subProj), useMatrix),     geneDF &lt;- geneDF[geneDF$name %in% rownames(seuratRNA), , ,         drop = FALSE],     splitGeneDF &lt;- S4Vectors::split(geneDF, geneDF$seqnames),     featureDF &lt;- lapply(splitGeneDF, function(x) {,         x$idx &lt;- seq_len(nrow(x)),         return(x),     }) %&gt;% Reduce(\"rbind\", .),     dfParams &lt;- data.frame(reduction = reduction),     allChr &lt;- unique(featureDF$seqnames),     tmpFile &lt;- .tempfile(),     o &lt;- suppressWarnings(file.remove(paste0(tmpFile, \"-IntegrationBlock-\", ,         seq_along(blockList), \".h5\"))),     if (threads &gt; 1) {,         h5disableFileLocking(),     },     rD &lt;- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims, ,         corCutOff = corCutOff, dimsToUse = dimsToUse),     outDir1 &lt;- getOutputDirectory(ArchRProj),     outDir2 &lt;- file.path(outDir1, \"RNAIntegration\"),     outDir3 &lt;- file.path(outDir2, matrixName),     dir.create(outDir1, showWarnings = FALSE),     dir.create(outDir2, showWarnings = FALSE),     dir.create(outDir3, showWarnings = FALSE),     prevFiles &lt;- list.files(outDir3, full.names = TRUE),     prevFiles &lt;- .suppressAll(file.remove(prevFiles)),     tstart &lt;- Sys.time(),     threads2 &lt;- max(ceiling(threads * 0.75), 1),     .logDiffTime(paste0(\"Computing Integration in \", length(blockList), ,         \" Integration Blocks!\"), tstart, verbose = verbose, logFile = logFile),     dfAll &lt;- .safelapply(seq_along(blockList), function(i) {,         prefix &lt;- sprintf(\"Block (%s of %s) :\", i, length(blockList)),         .logDiffTime(sprintf(\"%s Computing Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         blocki &lt;- blockList[[i]],         subProj@cellColData &lt;- subProj@cellColData[blocki$ATAC, ,             ],         subProj@sampleColData &lt;- subProj@sampleColData[unique(subProj$Sample), ,             , drop = FALSE],         subRNA &lt;- seuratRNA[, blocki$RNA],         subRNA &lt;- subRNA[rownames(subRNA) %in% geneDF$name, ],         .logDiffTime(sprintf(\"%s Identifying Variable Genes\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         subRNA &lt;- FindVariableFeatures(object = subRNA, nfeatures = nGenes, ,             verbose = FALSE),         subRNA &lt;- ScaleData(object = subRNA, verbose = FALSE),         if (is.null(genesUse)) {,             genesUse &lt;- VariableFeatures(object = subRNA),         },         .logDiffTime(sprintf(\"%s Getting GeneScoreMatrix\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         mat &lt;- .getPartialMatrix(getArrowFiles(subProj), featureDF = geneDF[geneDF$name %in% ,             genesUse, ], threads = 1, cellNames = subProj$cellNames, ,             useMatrix = useMatrix, verbose = FALSE),         rownames(mat) &lt;- geneDF[geneDF$name %in% genesUse, \"name\"],         .logThis(mat, paste0(\"GeneScoreMat-Block-\", i), logFile = logFile),         if (useImputation) {,             .logDiffTime(sprintf(\"%s Imputing GeneScoreMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             imputeParams &lt;- list(),             imputeParams$ArchRProj &lt;- subProj,             imputeParams$randomSuffix &lt;- TRUE,             imputeParams$reducedDims &lt;- reducedDims,             imputeParams$dimsToUse &lt;- dimsToUse,             imputeParams$scaleDims &lt;- scaleDims,             imputeParams$corCutOff &lt;- corCutOff,             imputeParams$threads &lt;- 1,             imputeParams$logFile &lt;- logFile,             subProj &lt;- suppressMessages(do.call(addImputeWeights, ,                 imputeParams)),             mat &lt;- suppressMessages(imputeMatrix(mat = mat, imputeWeights = getImputeWeights(subProj), ,                 verbose = FALSE, logFile = logFile)),             o &lt;- suppressWarnings(file.remove(unlist(getImputeWeights(subProj)[[1]]))),             .logThis(mat, paste0(\"GeneScoreMat-Block-Impute-\", ,                 i), logFile = logFile),         },         mat &lt;- log(mat + 1),         seuratATAC &lt;- Seurat::CreateSeuratObject(counts = mat[head(seq_len(nrow(mat)), ,             5), , drop = FALSE]),         seuratATAC[[\"GeneScore\"]] &lt;- Seurat::CreateAssayObject(counts = mat),         rm(mat),         DefaultAssay(seuratATAC) &lt;- \"GeneScore\",         seuratATAC &lt;- Seurat::ScaleData(seuratATAC, verbose = FALSE),         .logDiffTime(sprintf(\"%s Seurat FindTransferAnchors\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferAnchors &lt;- .retryCatch({,             gc(),             Seurat::FindTransferAnchors(reference = subRNA, query = seuratATAC, ,                 reduction = reduction, features = genesUse, verbose = FALSE, ,                 ...),         }, maxAttempts = 2, logFile = logFile),         .logThis(paste0(utils::capture.output(transferAnchors), ,             collapse = \"\\n\"), paste0(\"transferAnchors-\", i), ,             logFile = logFile),         rDSub &lt;- rD[colnames(seuratATAC), , drop = FALSE],         .logThis(rDSub, paste0(\"rDSub-\", i), logFile = logFile),         transferParams$anchorset &lt;- transferAnchors,         transferParams$weight.reduction &lt;- CreateDimReducObject(embeddings = rDSub, ,             key = \"LSI_\", assay = DefaultAssay(seuratATAC)),         transferParams$verbose &lt;- FALSE,         transferParams$dims &lt;- seq_len(ncol(rDSub)),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Group Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata &lt;- subRNA$Group,         rnaLabels &lt;- do.call(Seurat::TransferData, transferParams),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Names Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata &lt;- colnames(subRNA),         rnaLabels2 &lt;- do.call(Seurat::TransferData, transferParams)[, ,             1],         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Seurat TransferData GeneMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             transferParams$refdata &lt;- GetAssayData(subRNA, assay = \"RNA\", ,                 slot = \"data\"),             gc(),             matchedRNA &lt;- do.call(Seurat::TransferData, transferParams),             matchedRNA &lt;- matchedRNA@data,         },         matchDF &lt;- DataFrame(cellNames = colnames(seuratATAC), ,             predictionScore = rnaLabels$prediction.score.max, ,             predictedGroup = rnaLabels$predicted.id, predictedCell = rnaLabels2),         rownames(matchDF) &lt;- matchDF$cellNames,         .logDiffTime(sprintf(\"%s Saving TransferAnchors Joint CCA\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         jointCCA &lt;- DataFrame(transferAnchors@object.list[[1]]@reductions$cca@cell.embeddings),         jointCCA$Assay &lt;- ifelse(endsWith(rownames(jointCCA), ,             \"_reference\"), \"RNA\", \"ATAC\"),         jointCCA$Group &lt;- NA,         jointCCA$Score &lt;- NA,         jointCCA[paste0(colnames(subRNA), \"_reference\"), \"Group\"] &lt;- subRNA$Group,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Group\"] &lt;- matchDF$predictedGroup,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Score\"] &lt;- matchDF$predictionScore,         .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,             paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),         rm(transferParams, transferAnchors),         gc(),         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Transferring Paired RNA to Temp File\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             tmpFilei &lt;- paste0(tmpFile, \"-IntegrationBlock-\", ,                 i, \".h5\"),             o &lt;- h5createFile(tmpFilei),             sampleNames &lt;- getCellColData(subProj, \"Sample\")[matchDF$cellNames, ,                 ],             uniqueSamples &lt;- unique(sampleNames),             matchedRNA &lt;- .safeSubset(mat = matchedRNA, subsetRows = paste0(featureDF$name), ,                 subsetCols = matchDF$cellNames),             for (z in seq_along(uniqueSamples)) {,                 mat &lt;- matchedRNA[, which(sampleNames == uniqueSamples[z]), ,                   drop = FALSE],                 Group &lt;- uniqueSamples[z],                 o &lt;- tryCatch({,                   h5delete(tmpFilei, paste0(Group)),                 }, error = function(x) {,                 }),                 o &lt;- h5createGroup(tmpFilei, paste0(Group)),                 j &lt;- Rle(findInterval(seq(mat@x) - 1, mat@p[-1]) + ,                   1),                 lengthRle &lt;- length(j@lengths),                 lengthI &lt;- length(mat@i),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/i\"), storage.mode = \"integer\", dims = c(lengthI, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jLengths\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jValues\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/x\"), storage.mode = \"double\", dims = c(lengthI, ,                   1), level = 0)),                 o &lt;- .suppressAll(h5write(obj = mat@i + 1, file = tmpFilei, ,                   name = paste0(Group, \"/i\"))),                 o &lt;- .suppressAll(h5write(obj = j@lengths, file = tmpFilei, ,                   name = paste0(Group, \"/jLengths\"))),                 o &lt;- .suppressAll(h5write(obj = j@values, file = tmpFilei, ,                   name = paste0(Group, \"/jValues\"))),                 o &lt;- .suppressAll(h5write(obj = mat@x, file = tmpFilei, ,                   name = paste0(Group, \"/x\"))),                 o &lt;- .suppressAll(h5write(obj = colnames(mat), ,                   file = tmpFilei, name = paste0(Group, \"/cellNames\"))),             },             rm(matchedRNA, mat, j),         },         .logDiffTime(sprintf(\"%s Completed Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         gc(),         matchDF$Block &lt;- Rle(i),         matchDF,     }, threads = threads2) %&gt;% Reduce(\"rbind\", .),     if (plotUMAP) {,         for (i in seq_along(blockList)) {,             o &lt;- tryCatch({,                 prefix &lt;- sprintf(\"Block (%s of %s) :\", i, length(blockList)),                 .logDiffTime(sprintf(\"%s Plotting Joint UMAP\", ,                   prefix), tstart, verbose = verbose, logFile = logFile),                 jointCCA &lt;- readRDS(file.path(outDir3, paste0(\"Save-Block\", ,                   i, \"-JointCCA.rds\"))),                 set.seed(1),                 UMAPParams &lt;- .mergeParams(UMAPParams, list(n_neighbors = 40, ,                   min_dist = 0.4, metric = \"cosine\", verbose = FALSE)),                 UMAPParams$X &lt;- as.data.frame(jointCCA[, grep(\"CC_\", ,                   colnames(jointCCA))]),                 UMAPParams$ret_nn &lt;- FALSE,                 UMAPParams$ret_model &lt;- FALSE,                 UMAPParams$n_threads &lt;- 1,                 uwotUmap &lt;- tryCatch({,                   do.call(uwot::umap, UMAPParams),                 }, error = function(e) {,                   errorList &lt;- UMAPParams,                   .logError(e, fn = \"uwot::umap\", info = prefix, ,                     errorList = errorList, logFile = logFile),                 }),                 jointCCA$UMAP1 &lt;- uwotUmap[, 1],                 jointCCA$UMAP2 &lt;- uwotUmap[, 2],                 .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,                   paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),                 p1 &lt;- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Assay, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by Assay\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 p2 &lt;- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Group, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by scRNA Group\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 pdf(file.path(outDir3, paste0(\"Save-Block\", i, ,                   \"-JointCCA-UMAP.pdf\")), width = 12, height = 6, ,                   useDingbats = FALSE),                 ggAlignPlots(p1, p2, type = \"h\"),                 dev.off(),             }, error = function(e) {,             }),         },     },     if (addToArrow) {,         .logDiffTime(\"Transferring Data to ArrowFiles\", tstart, ,             verbose = verbose, logFile = logFile),         matrixName &lt;- .isProtectedArray(matrixName),         integrationFiles &lt;- paste0(tmpFile, \"-IntegrationBlock-\", ,             seq_along(blockList), \".h5\"),         if (!all(file.exists(integrationFiles))) {,             .logMessage(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\", ,                 logFile = logFile),             stop(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\"),         },         h5list &lt;- .safelapply(seq_along(integrationFiles), function(x) {,             h5ls(integrationFiles[x]),         }, threads = threads),         ArrowFiles &lt;- getArrowFiles(ArchRProj),         allSamples &lt;- names(ArrowFiles),         o &lt;- .safelapply(seq_along(allSamples), function(y) {,             sample &lt;- allSamples[y],             prefix &lt;- sprintf(\"%s (%s of %s)\", sample, y, length(ArrowFiles)),             .logDiffTime(sprintf(\"%s Getting GeneIntegrationMatrix From TempFiles!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             sampleIF &lt;- lapply(seq_along(h5list), function(x) {,                 if (any(h5list[[x]]$group == paste0(\"/\", sample))) {,                   integrationFiles[x],                 },                 else {,                   NULL,                 },             }) %&gt;% unlist,             sampleMat &lt;- lapply(seq_along(sampleIF), function(x) {,                 cellNames &lt;- .h5read(sampleIF[x], paste0(sample, ,                   \"/cellNames\")),                 mat &lt;- sparseMatrix(i = .h5read(sampleIF[x], ,                   paste0(sample, \"/i\"))[, 1], j = as.vector(Rle(.h5read(sampleIF[x], ,                   paste0(sample, \"/jValues\"))[, 1], .h5read(sampleIF[x], ,                   paste0(sample, \"/jLengths\"))[, 1])), x = .h5read(sampleIF[x], ,                   paste0(sample, \"/x\"))[, 1], dims = c(nrow(featureDF), ,                   length(cellNames))),                 colnames(mat) &lt;- cellNames,                 mat,             }) %&gt;% Reduce(\"cbind\", .),             sampleMat@x &lt;- exp(sampleMat@x) - 1,             sampleMat &lt;- .normalizeCols(sampleMat, scaleTo = scaleTo),             sampleMat &lt;- drop0(sampleMat),             rownames(sampleMat) &lt;- paste0(featureDF$name),             sampleMat &lt;- sampleMat[, ArchRProj$cellNames[BiocGenerics::which(ArchRProj$Sample == ,                 sample)], drop = FALSE],             o &lt;- .createArrowGroup(ArrowFile = ArrowFiles[sample], ,                 group = matrixName, force = force),             o &lt;- .initializeMat(ArrowFile = ArrowFiles[sample], ,                 Group = matrixName, Class = \"double\", Units = \"NormCounts\", ,                 cellNames = colnames(sampleMat), params = dfParams, ,                 featureDF = featureDF, force = force),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictionScore\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictionScore\")),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictedGroup\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedGroup\")),             o &lt;- h5write(obj = dfAll[colnames(sampleMat), \"predictedCell\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedCell\")),             .logDiffTime(sprintf(\"%s Adding GeneIntegrationMatrix to ArrowFile!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             for (z in seq_along(allChr)) {,                 chrz &lt;- allChr[z],                 .logDiffTime(sprintf(\"Adding GeneIntegrationMatrix to %s for Chr (%s of %s)!\", ,                   sample, z, length(allChr)), tstart, verbose = FALSE, ,                   logFile = logFile),                 idz &lt;- BiocGenerics::which(featureDF$seqnames %bcin% ,                   chrz),                 matz &lt;- sampleMat[idz, , drop = FALSE],                 stopifnot(identical(paste0(featureDF$name[idz]), ,                   paste0(rownames(matz)))),                 o &lt;- .addMatToArrow(mat = matz, ArrowFile = ArrowFiles[sample], ,                   Group = paste0(matrixName, \"/\", chrz), binarize = FALSE, ,                   addColSums = TRUE, addRowSums = TRUE, addRowVarsLog2 = TRUE, ,                   logFile = logFile),                 rm(matz),                 if (z%%3 == 0 | z == length(allChr)) {,                   gc(),                 },             },             0,         }, threads = threads),         o &lt;- suppressWarnings(file.remove(integrationFiles)),     },     .logDiffTime(\"Completed Integration with RNA Matrix\", tstart, ,         verbose = verbose, logFile = logFile),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedCell, name = nameCell, force = TRUE),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedGroup, name = nameGroup, force = TRUE),     ArchRProj &lt;- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictionScore, name = nameScore, force = TRUE),     .endLogging(logFile = logFile),     return(ArchRProj), } | function (ArchRProj = NULL, useSeqnames = NULL, verbose = TRUE, ,     binarize = FALSE, threads = getArchRThreads(), logFile = createLogFile(\"exportMatrixForSTREAM\")) , {,     mat &lt;- getMatrixFromProject(ArchRProj = ArchRProj, useMatrix = \"PeakMatrix\", ,         useSeqnames = useSeqnames, verbose = verbose, binarize = binarize, ,         threads = threads, logFile = logFile),     featureDF &lt;- ArchR:::.getFeatureDF(getArrowFiles(ArchRProj)[1], ,         \"PeakMatrix\"),     stopifnot(all(featureDF$idx == rowData(mat)$idx)),     countsDF &lt;- Matrix::summary(assay(mat)),     peaksDF &lt;- data.frame(as.vector(featureDF[, 1]), featureDF[, ,         3], featureDF[, 4]),     cellsDF &lt;- data.frame(colnames(mat)),     data.table::fwrite(countsDF, file = \"STREAM_Counts.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(peaksDF, file = \"STREAM_Regions.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(cellsDF, file = \"STREAM_Sample.tsv.gz\", ,         sep = \"\\t\", row.names = FALSE, col.names = FALSE),     return(0), } | function (ArchRProj = NULL) , {,     if (!inherits(ArchRProj, \"ArchRProject\")) {,         stop(\"Not a valid ArchRProject as input!\"),     },     else {,         ArchRProj,     }, } | function (p = NULL, plotWidth = unit(6, \"in\"), plotHeight = unit(6, ,     \"in\"), margin = 0.25, height = 1, it = 0.05, newPage = FALSE) , {,     .requirePackage(\"grid\", source = \"cran\"),     .requirePackage(\"gridExtra\", source = \"cran\"),     if (!inherits(plotWidth, \"unit\")) {,         plotWidth &lt;- unit(plotWidth, \"in\"),     },     if (!inherits(plotHeight, \"unit\")) {,         plotHeight &lt;- unit(plotHeight, \"in\"),     },     g &lt;- ggplotGrob(p),     legend &lt;- grep(\"guide-box\", g$layout$name),     if (length(legend) != 0) {,         gl &lt;- g$grobs[[legend]],         g &lt;- ggplotGrob(p + theme(legend.position = \"none\")),     },     else {,         gl &lt;- NULL,         g &lt;- ggplotGrob(p),     },     panels &lt;- grep(\"panel\", g$layout$name),     panel_index_w &lt;- unique(g$layout$l[panels]),     panel_index_h &lt;- unique(g$layout$t[panels]),     nw &lt;- length(panel_index_w),     nh &lt;- length(panel_index_h),     pw &lt;- convertWidth(plotWidth, unitTo = \"in\", valueOnly = TRUE),     ph &lt;- convertWidth(plotHeight, unitTo = \"in\", valueOnly = TRUE),     pw &lt;- pw * 0.95,     ph &lt;- ph * 0.95,     x &lt;- 0,     width &lt;- 1,     sm &lt;- FALSE,     while (!sm) {,         x &lt;- x + it,         w &lt;- unit(x * width, \"in\"),         h &lt;- unit(x * height/width, \"in\"),         m &lt;- unit(x * margin/width, \"in\"),         g$widths[panel_index_w] &lt;- rep(w, nw),         g$heights[panel_index_h] &lt;- rep(h, nh),         sw &lt;- convertWidth(x = sum(g$widths) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sh &lt;- convertHeight(x = sum(g$heights) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sm &lt;- sw &gt; pw | sh &gt; ph,     },     if (length(legend) != 0) {,         sgh &lt;- convertHeight(x = sum(g$heights), unitTo = \"in\", ,             valueOnly = TRUE),         sgw &lt;- convertWidth(x = sum(g$widths), unitTo = \"in\", ,             valueOnly = TRUE),         slh &lt;- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,             valueOnly = TRUE),         slw &lt;- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,             valueOnly = TRUE),         size &lt;- 6,         wh &lt;- 0.1,         it &lt;- 0,         while (slh &gt; 0.2 * ph | slw &gt; pw) {,             it &lt;- it + 1,             if (it &gt; 3) {,                 break,             },             size &lt;- size * 0.8,             wh &lt;- wh * 0.8,             gl &lt;- ggplotGrob(p + theme(legend.key.width = unit(wh, ,                 \"cm\"), legend.key.height = unit(wh, \"cm\"), legend.spacing.x = unit(0, ,                 \"cm\"), legend.spacing.y = unit(0, \"cm\"), legend.text = element_text(size = max(size, ,                 2))) + guides(fill = guide_legend(ncol = 4), ,                 color = guide_legend(ncol = 4)))$grobs[[legend]],             slh &lt;- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,                 valueOnly = TRUE),             slw &lt;- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,                 valueOnly = TRUE),         },         p &lt;- grid.arrange(g, gl, ncol = 1, nrow = 2, heights = unit.c(unit(sgh, ,             \"in\"), unit(min(slh, 0.2 * pw), \"in\")), newpage = newPage),     },     else {,         p &lt;- grid.arrange(g, newpage = newPage),     },     invisible(p), } |\n",
       "| visible | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | ⋯ | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AllClasses.R\n",
       "value   rownames                                                                                   \n",
       "visible FALSE                                                                                      \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationGenome.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
       "value   function (genome = NULL) , {,     .validInput(input = genome, name = \"genome\", valid = \"character\"),     if (toupper(genome) == \"HG19\" | toupper(genome) == \"HG38\") {,         if (suppressWarnings(!require(org.Hs.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Hs.eg.db\", update = FALSE),         },         library(org.Hs.eg.db),         annodb <- org.Hs.eg.db,     },     else if (toupper(genome) == \"MM9\" | toupper(genome) == \"MM10\") {,         if (suppressWarnings(!require(org.Mm.eg.db))) {,             message(\"Package does not exist, now trying bioconductor..\"),             BiocManager::install(\"org.Mm.eg.db\", update = FALSE),         },         library(org.Mm.eg.db),         annodb <- org.Mm.eg.db,     },     else {,         stop(\"Genome not recognized!\"),     },     return(annodb), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/AnnotationPeaks.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
       "value   function (seEnrich = NULL, pal = paletteContinuous(set = \"comet\", ,     n = 100), n = 10, cutOff = 20, pMax = Inf, clusterCols = TRUE, ,     binaryClusterRows = TRUE, labelRows = TRUE, rastr = TRUE, ,     transpose = FALSE, returnMatrix = FALSE, logFile = createLogFile(\"plotEnrichHeatmap\")) , {,     .validInput(input = seEnrich, name = \"seEnrich\", valid = c(\"SummarizedExperiment\")),     .validInput(input = pal, name = \"pal\", valid = c(\"character\")),     .validInput(input = n, name = \"n\", valid = c(\"integer\")),     .validInput(input = cutOff, name = \"cutOff\", valid = c(\"numeric\")),     .validInput(input = pMax, name = \"pMax\", valid = c(\"numeric\")),     .validInput(input = clusterCols, name = \"clusterCols\", valid = c(\"boolean\")),     .validInput(input = binaryClusterRows, name = \"binaryClusterRows\", ,         valid = c(\"boolean\")),     .validInput(input = labelRows, name = \"labelRows\", valid = c(\"boolean\")),     .validInput(input = rastr, name = \"rastr\", valid = c(\"boolean\")),     .validInput(input = transpose, name = \"transpose\", valid = c(\"boolean\")),     .validInput(input = returnMatrix, name = \"returnMatrix\", ,         valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotEnrichHeatmap Input-Parameters\", logFile = logFile),     mat <- assays(seEnrich)[[\"mlog10Padj\"]],     .logThis(mat, \"mat-mlog10Padj\", logFile = logFile),     keep <- lapply(seq_len(ncol(mat)), function(x) {,         idx <- head(order(mat[, x], decreasing = TRUE), n),         rownames(mat)[idx[which(mat[idx, x] > cutOff)]],     }) %>% unlist %>% unique,     mat <- mat[keep, , drop = FALSE],     .logThis(mat, \"mat-mlog10Padj-Filter\", logFile = logFile),     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     passMat <- lapply(seq_len(nrow(mat)), function(x) {,         (mat[x, ] >= 0.9 * max(mat[x, ])) * 1,     }) %>% Reduce(\"rbind\", .) %>% data.frame,     colnames(passMat) <- colnames(mat),     .logThis(passMat, \"passMat\", logFile = logFile),     mat[mat > pMax] <- pMax,     if (nrow(mat) == 0) {,         stop(\"No enrichments found for your cutoff!\"),     },     mat <- .rowScale(as.matrix(mat), min = 0),     if (pMax != 100) {,         rownames(mat[[1]]) <- paste0(rownames(mat[[1]]), \" (\", ,             round(mat$max), \")\"),         rownames(passMat) <- rownames(mat[[1]]),     },     .logThis(mat, \"mat-mlog10Padj-RowScale\", logFile = logFile),     mat2 <- mat[[1]] * 100,     if (nrow(mat2) > 1 & ncol(mat2) > 1) {,         if (binaryClusterRows) {,             bS <- .binarySort(mat2, lmat = passMat[rownames(mat2), ,                 colnames(mat2)], clusterCols = clusterCols),             mat2 <- bS[[1]][, colnames(mat2)],             clusterRows <- FALSE,             clusterCols <- bS[[2]],         },         else {,             clusterRows <- TRUE,         },     },     else {,         clusterCols <- NULL,         clusterRows <- FALSE,     },     if (nrow(mat2) > 100) {,         borderColor <- FALSE,     },     else {,         borderColor <- TRUE,     },     .logThis(mat2, \"mat-mlog10Padj-RowScale-Plot\", logFile = logFile),     if (transpose) {,         if (!is.null(clusterCols)) {,             mat2 <- t(mat2[, clusterCols$order, drop = FALSE]),         },         else {,             mat2 <- t(mat2),         },         if (returnMatrix) {,             return(mat2),         },         ht <- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = FALSE, ,                 clusterRows = FALSE, labelRows = TRUE, useRaster = rastr, ,                 fontSizeCols = 6, borderColor = borderColor, ,                 customColLabel = seq_len(ncol(mat2)), showRowDendrogram = FALSE, ,                 draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     else {,         if (returnMatrix) {,             return(mat2),         },         ht <- tryCatch({,             .ArchRHeatmap(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),         }, error = function(e) {,             errorList = list(mat = as.matrix(mat2), scale = FALSE, ,                 limits = c(0, max(mat2)), color = pal, clusterCols = clusterCols, ,                 clusterRows = clusterRows, useRaster = rastr, ,                 borderColor = borderColor, fontSizeRows = 6, ,                 customRowLabel = seq_len(nrow(mat2)), labelCols = TRUE, ,                 showColDendrogram = TRUE, draw = FALSE, name = \"Norm. Enrichment -log10(P-adj) [0-Max]\"),             .logError(e, fn = \".ArchRHeatmap\", info = \"\", errorList = errorList, ,                 logFile = logFile),         }),     },     return(ht), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-CRHR1.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser-FKBP5.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser3.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0.1) + theme(panel.spacing = unit(0, ,         \"lines\"), axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0.35, ,         0.15, 0.35, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-CAMKV.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-KCNIP4.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ArchRBrowser4-xiangyu.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "value   function (plotList = NULL, useMatrix = NULL, featureMat = NULL, ,     log2Norm = FALSE, feature = NULL, pal = NULL, sizes = NULL, ,     baseSize = NULL, facetbaseSize = NULL, borderWidth = NULL, ,     tickWidth = NULL) , {,     .requirePackage(\"patchwork\", installInfo = \"devtools::install_github('thomasp85/patchwork')\"),     if (is.null(pal)) {,         pal <- paletteDiscrete(values = featureMat$Group, set = \"stallion\"),     },     if (log2Norm) {,         title <- paste0(\"Log2 \", useMatrix, \" : \", feature),     },     else {,         title <- paste0(\"Raw \", useMatrix, \" : \", feature),     },     featurePlot <- ggGroup(x = featureMat$Group, y = featureMat[, ,         feature], groupOrder = gtools::mixedsort(paste0(unique(featureMat$Group))), ,         pal = pal) + facet_wrap(x ~ ., ncol = 1, scales = \"free_y\", ,         strip.position = \"right\") + guides(fill = \"none\", colour = \"none\") + ,         theme_ArchR(baseSize = baseSize, baseRectSize = borderWidth, ,             baseLineSize = tickWidth, legendPosition = \"right\", ,             axisTickCm = 0) + theme(panel.spacing = unit(0, \"lines\"), ,         axis.title.x = element_blank(), axis.title.y = element_blank(), ,         axis.text.y = element_blank(), axis.ticks.y = element_blank(), ,         strip.text = element_text(size = facetbaseSize, color = \"black\", ,             margin = margin(0, 0.35, 0, 0.35, \"cm\")), strip.text.y = element_text(angle = 0), ,         strip.background = element_rect(color = \"black\")) + theme(plot.margin = unit(c(0, ,         0.15, 0, 0.15), \"cm\")) + ggtitle(title),     if (any(tolower(names(plotList)) %in% \"bulktrack\")) {,         idx <- which(tolower(names(plotList)) == \"bulktrack\"),         p <- plotList[[idx]] + featurePlot + plot_spacer(),         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     else {,         idx <- which(tolower(names(plotList)) == \"sctrack\"),         p <- plotList[[idx]] + featurePlot,         plotList[idx] <- NULL,         for (i in seq_along(plotList)) {,             p <- p + plotList[[i]] + plot_spacer() + plot_spacer(),         },         p <- p + plot_layout(ncol = 3, widths = c(3, 1, 0.2), ,             heights = sizes),     },     p, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "        ⋯\n",
       "value   ⋯\n",
       "visible ⋯\n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/MultiModal.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             \n",
       "value   function (ArchRProj = NULL, name = \"CombinedDims\", reducedDims = NULL, ,     dimWeights = NULL, dimsToUse = NULL, scaleDims = NULL, corCutOff = 0.75) , {,     if (is.null(dimWeights)) {,         dimWeights <- rep(1, length(reducedDims)),     },     combinedDims <- lapply(seq_along(reducedDims), function(x) {,         rD <- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims[x], ,             dimsToUse = dimsToUse, scaleDims = scaleDims, corCutOff = corCutOff),         cV <- apply(as.matrix(rD), 2, var),         normV <- 1/sqrt(sum(cV)),         rD * normV * dimWeights[x],     }) %>% Reduce(\"cbind\", .),     ArchRProj@reducedDims[[name]] <- SimpleList(matRD = combinedDims, ,         scaleDims = NA, corToDepth = list(scaled = rep(0, ncol(combinedDims)), ,             none = rep(0, ncol(combinedDims)))),     ArchRProj, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/plot_footprints.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "value   function (seFoot = NULL, names = NULL, pal = NULL, flank = 250, ,     flankNorm = 50, normMethod = \"Subtract\", smoothWindow = NULL, ,     baseSize = 6, plot = TRUE, ArchRProj = NULL, plotName = paste0(\"Plot-Footprints-\", ,         normMethod), height = 6, width = 4, addDOC = TRUE, force = FALSE, ,     logFile = createLogFile(\"plotFootprints\")) , {,     tstart <- Sys.time(),     if (is.null(names)) {,         names <- names(assays(seFoot)),     },     if (plot) {,         name <- gsub(\"\\\\\\\\.pdf\", \"\", plotName),         if (is.null(ArchRProj)) {,             outDir <- \"Plots\",         },         else {,             outDir <- file.path(getOutputDirectory(ArchRProj), ,                 \"Plots\"),         },         dir.create(outDir, showWarnings = FALSE),         if (addDOC) {,             doc <- gsub(\":\", \"-\", stringr::str_split(Sys.time(), ,                 pattern = \" \", simplify = TRUE)[1, 2]),             filename <- file.path(outDir, paste0(name, \"_Date-\", ,                 Sys.Date(), \"_Time-\", doc, \".pdf\")),         },         else {,             filename <- file.path(outDir, paste0(name, \".pdf\")),         },         pdf(filename, width = width, height = height, useDingbats = FALSE),     },     ggList <- lapply(seq_along(names), function(x) {,         gg <- .ggFootprint(seFoot = seFoot, name = names[x], ,             pal = pal, smoothWindow = smoothWindow, flank = flank, ,             flankNorm = flankNorm, baseSize = baseSize, normMethod = normMethod, ,             logFile = logFile),         if (plot) {,             if (x != 1) {,                 grid::grid.newpage(),             },             grid::grid.draw(gg),             return(0),         },         else {,             return(gg),         },     }),     if (!plot) {,         names(ggList) <- names,         ggList,     },     else {,         dev.off(),         return(invisible(0)),     }, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ProjectMethods.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       \n",
       "value   function (ArchRProj = NULL, features = NULL, name = NULL, addRatio = TRUE, ,     threads = getArchRThreads(), logFile = createLogFile(\"addFeatureCounts\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = features, name = \"features\", valid = c(\"granges\")),     .validInput(input = name, name = \"name\", valid = c(\"character\")),     .validInput(input = addRatio, name = \"addRatio\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     ArrowFiles <- getArrowFiles(ArchRProj),     cellNames <- ArchRProj$cellNames,     featuresList <- split(features, seqnames(features)),     h5disableFileLocking(),     countsDF <- .safelapply(seq_along(featuresList), function(i) {,         chri <- names(featuresList)[i],         cellTotal <- rep(0, length(cellNames)),         names(cellTotal) <- cellNames,         featuresi <- ranges(featuresList[[i]]),         .logDiffTime(paste0(\"Counting in \", chri, \" (\", i, \" of \", ,             length(featuresList), \")\"), t1 = tstart, logFile = logFile),         for (j in seq_along(ArrowFiles)) {,             fragmentsij <- .getFragsFromArrow(ArrowFile = ArrowFiles[j], ,                 chr = chri, out = \"IRanges\", cellNames = cellNames),             if (length(fragmentsij) > 0) {,                 mcols(fragmentsij)$RG@values <- match(mcols(fragmentsij)$RG@values, ,                   cellNames),                 for (y in seq_len(2)) {,                   if (y == 1) {,                     temp <- IRanges(start(fragmentsij), width = 1),                   },                   else if (y == 2) {,                     temp <- IRanges(end(fragmentsij), width = 1),                   },                   stopifnot(length(temp) == length(fragmentsij)),                   tabSum <- S4Vectors:::tabulate(mcols(fragmentsij)$RG[queryHits(findOverlaps(temp, ,                     featuresi))]),                   cellTotal[seq_along(tabSum)] <- cellTotal[seq_along(tabSum)] + ,                     tabSum,                 },             },             if (j%%3 == 0) {,                 gc(),             },         },         cellTotal,     }, threads = threads) %>% Reduce(\"rbind\", .),     totalCounts <- colSums(countsDF),     countRatio <- totalCounts/(2 * ArchRProj$nFrags),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Counts\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj, data = totalCounts, ,         cells = names(totalCounts), name = paste0(name, \"Counts\"), ,         force = TRUE),     .logDiffTime(sprintf(\"Adding %s to cellColData\", paste0(name, ,         \"Ratio\")), t1 = tstart, verbose = TRUE, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj, data = countRatio, ,         cells = names(totalCounts), name = paste0(name, \"Ratio\"), ,         force = TRUE),     ArchRProj, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/QualityControl.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "value   function (ArchRProj = NULL, groupBy = \"Sample\", chromSizes = getChromSizes(ArchRProj), ,     maxSize = 750, pal = NULL, returnDF = FALSE, threads = getArchRThreads(), ,     logFile = createLogFile(\"plotFragmentSizes\")) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = maxSize, name = \"maxSize\", valid = c(\"integer\")),     .validInput(input = returnDF, name = \"returnDF\", valid = c(\"boolean\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logThis(mget(names(formals()), sys.frame(sys.nframe())), ,         \"plotFragmentSizes Input-Parameters\", logFile = logFile),     chr <- paste0(seqnames(chromSizes)),     groups <- getCellColData(ArchRProj = ArchRProj, select = groupBy, ,         drop = FALSE),     uniqGroups <- gtools::mixedsort(unique(groups[, 1])),     if (threads > 1) {,         h5disableFileLocking(),     },     dfFS <- .safelapply(seq_along(uniqGroups), function(x) {,         .logDiffTime(paste0(uniqGroups[x], \" Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         cellx <- rownames(groups)[which(paste0(groups[, 1]) == ,             uniqGroups[x])],         for (i in seq_along(chr)) {,             if (i == 1) {,                 fsi <- unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %>% width %>% tabulate(nbins = maxSize),             },             else {,                 fsi <- fsi + unlist(suppressMessages(getFragmentsFromProject(ArchRProj = ArchRProj, ,                   subsetBy = chromSizes[paste0(seqnames(chromSizes)) %in% ,                     chr[i]], cellNames = cellx, logFile = logFile)), ,                   use.names = FALSE) %>% width %>% tabulate(nbins = maxSize),             },             .logThis(fsi, paste0(uniqGroups[x], \" : FragSizes : \", ,                 chr[i]), logFile = logFile),         },         df <- DataFrame(group = uniqGroups[x], fragmentSize = seq_along(fsi), ,             fragmentPercent = round(100 * fsi/sum(fsi), 4)),         .logThis(df, paste0(uniqGroups[x], \" : Frag DF\"), logFile = logFile),         .logDiffTime(paste0(uniqGroups[x], \" Finished Computing FragmentSizes (\", ,             x, \" of \", length(uniqGroups), \")!\"), t1 = tstart, ,             logFile = logFile),         df,     }, threads = threads) %>% Reduce(\"rbind\", .),     .logThis(dfFS, paste0(\"All : FragSizes DF\"), logFile = logFile),     .endLogging(logFile = logFile),     if (threads > 1) {,         h5enableFileLocking(),     },     if (returnDF) {,         return(dfFS),     },     else {,         plotDF <- data.frame(dfFS),         if (is.null(pal)) {,             pal <- paletteDiscrete(values = unique(plotDF$group)),         },         p <- ggplot(plotDF, aes(fragmentSize, fragmentPercent, ,             color = group)) + geom_line(size = 1) + theme_ArchR() + ,             xlab(\"ATAC-seq Fragment Size (bp)\") + ylab(\"Percentage of Fragments\") + ,             scale_color_manual(values = pal) + scale_y_continuous(limits = c(0, ,             max(plotDF$fragmentPercent) * 1.05), expand = c(0, ,             0)) + scale_x_continuous(limits = c(min(plotDF$fragmentSize), ,             max(plotDF$fragmentSize)), expand = c(0, 0)),         p,     }, }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RcppExports.R                                                                                           \n",
       "value   function (string_vector, id_vector, n_id, window, w, kmer) , {,     .Call(\"_ArchR_kmerIDFrequencyCpp\", PACKAGE = \"ArchR\", string_vector, ,         id_vector, n_id, window, w, kmer), }\n",
       "visible FALSE                                                                                                                                                                                  \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ReproduciblePeakSet.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               \n",
       "value   function () , {,     message(\"Searching For MACS2..\"),     if (.suppressAll(.checkPath(\"macs2\", throwError = FALSE))) {,         message(\"Found with $path!\"),         return(\"macs2\"),     },     message(\"Not Found in $PATH\"),     search2 <- suppressWarnings(tryCatch({,         system2(\"pip\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     search3 <- suppressWarnings(tryCatch({,         system2(\"pip3\", \"show macs2\", stdout = TRUE, stderr = NULL),     }, error = function(x) {,         \"ERROR\",     })),     if (length(search2) > 0) {,         if (search2[1] != \"ERROR\") {,             path2Install <- gsub(\"Location: \", \"\", search2[grep(\"Location\", ,                 search2, ignore.case = TRUE)]),             path2Bin <- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip\"),     if (length(search3) > 0) {,         if (search3[1] != \"ERROR\") {,             path2Install <- gsub(\"Location: \", \"\", search3[grep(\"Location\", ,                 search3, ignore.case = TRUE)]),             path2Bin <- gsub(\"lib/python/site-packages\", \"bin/macs2\", ,                 path2Install),             if (.suppressAll(.checkPath(path2Bin, throwError = FALSE))) {,                 message(\"Found with pip3!\"),                 return(path2Bin),             },         },     },     message(\"Not Found with pip3\"),     stop(\"Could Not Find Macs2! Please install w/ pip, add to your $PATH, or just supply the macs2 path directly and avoid this function!\"), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/RNAIntegration.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "value   function (ArchRProj = NULL, useMatrix = \"GeneScoreMatrix\", matrixName = \"GeneIntegrationMatrix\", ,     reducedDims = \"IterativeLSI\", seRNA = NULL, groupATAC = NULL, ,     groupRNA = NULL, groupList = NULL, sampleCellsATAC = 10000, ,     sampleCellsRNA = 10000, embeddingATAC = NULL, embeddingRNA = NULL, ,     dimsToUse = 1:30, scaleDims = NULL, corCutOff = 0.75, plotUMAP = TRUE, ,     UMAPParams = list(n_neighbors = 40, min_dist = 0.4, metric = \"cosine\", ,         verbose = FALSE), nGenes = 2000, useImputation = TRUE, ,     reduction = \"cca\", addToArrow = TRUE, scaleTo = 10000, genesUse = NULL, ,     nameCell = \"predictedCell\", nameGroup = \"predictedGroup\", ,     nameScore = \"predictedScore\", transferParams = list(), threads = getArchRThreads(), ,     verbose = TRUE, force = FALSE, logFile = createLogFile(\"addGeneIntegrationMatrix\"), ,     ...) , {,     .validInput(input = ArchRProj, name = \"ArchRProj\", valid = c(\"ArchRProj\")),     .validInput(input = useMatrix, name = \"useMatrix\", valid = c(\"character\")),     .validInput(input = matrixName, name = \"matrixName\", valid = c(\"character\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = seRNA, name = \"seRNA\", valid = c(\"SummarizedExperiment\", ,         \"Seurat\")),     .validInput(input = groupATAC, name = \"groupATAC\", valid = c(\"character\", ,         \"null\")),     .validInput(input = groupRNA, name = \"groupRNA\", valid = c(\"character\")),     .validInput(input = groupList, name = \"groupList\", valid = c(\"list\", ,         \"null\")),     .validInput(input = sampleCellsATAC, name = \"sampleCellsATAC\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = sampleCellsRNA, name = \"sampleCellsRNA\", ,         valid = c(\"integer\", \"null\")),     .validInput(input = embeddingATAC, name = \"embeddingATAC\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = embeddingRNA, name = \"embeddingRNA\", ,         valid = c(\"data.frame\", \"null\")),     .validInput(input = reducedDims, name = \"reducedDims\", valid = c(\"character\")),     .validInput(input = dimsToUse, name = \"dimsToUse\", valid = c(\"numeric\", ,         \"null\")),     .validInput(input = scaleDims, name = \"scaleDims\", valid = c(\"boolean\", ,         \"null\")),     .validInput(input = plotUMAP, name = \"plotUMAP\", valid = c(\"boolean\")),     .validInput(input = UMAPParams, name = \"UMAPParams\", valid = c(\"list\")),     .validInput(input = nGenes, name = \"nGenes\", valid = c(\"integer\")),     .validInput(input = useImputation, name = \"useImputation\", ,         valid = c(\"boolean\")),     .validInput(input = reduction, name = \"reduction\", valid = c(\"character\")),     .validInput(input = addToArrow, name = \"addToArrow\", valid = c(\"boolean\")),     .validInput(input = scaleTo, name = \"scaleTo\", valid = c(\"numeric\")),     .validInput(input = genesUse, name = \"genesUse\", valid = c(\"character\", ,         \"null\")),     .validInput(input = nameCell, name = \"nameCell\", valid = c(\"character\")),     .validInput(input = nameGroup, name = \"nameGroup\", valid = c(\"character\")),     .validInput(input = nameScore, name = \"nameScore\", valid = c(\"character\")),     .validInput(input = transferParams, name = \"transferParams\", ,         valid = c(\"list\")),     .validInput(input = threads, name = \"threads\", valid = c(\"integer\")),     .validInput(input = verbose, name = \"verbose\", valid = c(\"boolean\")),     .validInput(input = force, name = \"force\", valid = c(\"boolean\")),     .validInput(input = logFile, name = \"logFile\", valid = c(\"character\")),     tstart <- Sys.time(),     .startLogging(logFile = logFile),     .logDiffTime(\"Running Seurat's Integration Stuart* et al 2019\", ,         tstart, verbose = verbose, logFile = logFile),     .requirePackage(\"Seurat\", source = \"cran\"),     .logThis(append(args, mget(names(formals()), sys.frame(sys.nframe()))), ,         \"Input-Parameters\", logFile = logFile),     if (is.null(groupList)) {,         groupList <- SimpleList(),         groupList[[1]] <- SimpleList(ATAC = ArchRProj$cellNames, ,             RNA = colnames(seRNA)),     },     .logDiffTime(\"Checking ATAC Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (useMatrix %ni% getAvailableMatrices(ArchRProj)) {,         .logMessage(paste0(\"Matrix \", useMatrix, \" does not exist in the provided ArchRProject. See available matrix names from getAvailableMatrices()!\"), ,             logFile = logFile),         stop(\"Matrix name provided to useMatrix does not exist in ArchRProject!\"),     },     if (!is.null(groupATAC)) {,         dfATAC <- getCellColData(ArchRProj = ArchRProj, select = groupATAC, ,             drop = FALSE),     },     nCell <- rep(0, length(ArchRProj$cellNames)),     names(nCell) <- ArchRProj$cellNames,     groupList <- lapply(seq_along(groupList), function(x) {,         ATAC <- groupList[[x]]$ATAC,         if (!is.null(groupATAC)) {,             if (any(ATAC %in% dfATAC[, 1])) {,                 idx <- which(ATAC %in% dfATAC[, 1]),                 ATAC2 <- rownames(dfATAC)[which(dfATAC[, 1] %in% ,                   ATAC[idx])],                 if (length(idx) == length(ATAC)) {,                   ATAC <- ATAC2,                 },                 else {,                   ATAC <- c(ATAC[-idx], ATAC2),                 },             },         },         SimpleList(ATAC = ATAC, RNA = groupList[[x]]$RNA),     }) %>% SimpleList,     for (i in seq_along(groupList)) {,         nCell[groupList[[i]]$ATAC] <- nCell[groupList[[i]]$ATAC] + ,             1,     },     if (!all(nCell == 1)) {,         .logMessage(paste0(\"Missing \", length(which(nCell == ,             0)), \" cells. Found \", length(which(nCell > 1)), ,             \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"), ,             logFile = logFile),         stop(\"Missing \", length(which(nCell == 0)), \" cells. Found \", ,             length(which(nCell > 1)), \" overlapping cells from ArchRProj in groupList! Cannot have overlapping/missing cells in ATAC input, check 'groupList' argument!\"),     },     .logDiffTime(\"Checking RNA Input\", tstart, verbose = verbose, ,         logFile = logFile),     if (inherits(seRNA, \"SummarizedExperiment\")) {,         seuratRNA <- CreateSeuratObject(counts = assay(seRNA)),         if (groupRNA %ni% colnames(colData(seRNA))) {,             .logMessage(\"groupRNA not in colData of seRNA\", logFile = logFile),             stop(\"groupRNA not in colData of seRNA\"),         },         seuratRNA$Group <- paste0(colData(seRNA)[, groupRNA, ,             drop = TRUE]),         rm(seRNA),     },     else {,         if (groupRNA %ni% colnames(seRNA@meta.data)) {,             .logMessage(\"groupRNA not in meta.data of Seurat Object\", ,                 logFile = logFile),             stop(\"groupRNA not in meta.data of Seurat Object\"),         },         seuratRNA <- seRNA,         seuratRNA$Group <- paste0(seRNA@meta.data[, groupRNA]),         rm(seRNA),     },     if (\"RNA\" %in% names(seuratRNA@assays)) {,         DefaultAssay(seuratRNA) <- \"RNA\",     },     else {,         stop(\"'RNA' is not present in Seurat Object's Assays! Please make sure that this assay is present!\"),     },     gc(),     if (!is.null(groupRNA)) {,         dfRNA <- DataFrame(row.names = colnames(seuratRNA), Group = seuratRNA$Group),     },     groupList <- lapply(seq_along(groupList), function(x) {,         RNA <- groupList[[x]]$RNA,         if (!is.null(groupRNA)) {,             if (any(RNA %in% dfRNA[, 1])) {,                 idx <- which(RNA %in% dfRNA[, 1]),                 RNA2 <- rownames(dfRNA)[which(dfRNA[, 1] %in% ,                   RNA[idx])],                 if (length(idx) == length(RNA)) {,                   RNA <- RNA2,                 },                 else {,                   RNA <- c(RNA[-idx], RNA2),                 },             },         },         SimpleList(ATAC = groupList[[x]]$ATAC, RNA = RNA),     }) %>% SimpleList,     cellRNA <- unlist(lapply(groupList, function(x) x$RNA)),     if (!all(cellRNA %in% colnames(seuratRNA))) {,         .logMessage(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\", ,             logFile = logFile),         stop(\"Found cells for RNA not in colnames(seRNA)! Please retry your input!\"),     },     seuratRNA <- seuratRNA[, unique(cellRNA)],     seuratRNA <- NormalizeData(object = seuratRNA, verbose = FALSE),     geneDF <- .getFeatureDF(getArrowFiles(ArchRProj), useMatrix),     sumOverlap <- sum(unique(geneDF$name) %in% unique(rownames(seuratRNA))),     if (sumOverlap < 5) {,         stop(\"Error not enough overlaps (\", sumOverlap, \") between gene names from gene scores (ArchR) and rna matrix (seRNA)!\"),     },     .logDiffTime(paste0(\"Found \", sumOverlap, \" overlapping gene names from gene scores and rna matrix!\"), ,         tstart, verbose = TRUE, logFile = logFile),     .logDiffTime(\"Creating Integration Blocks\", tstart, verbose = verbose, ,         logFile = logFile),     blockList <- SimpleList(),     for (i in seq_along(groupList)) {,         gLi <- groupList[[i]],         if (length(gLi$ATAC) > sampleCellsATAC) {,             if (!is.null(embeddingATAC)) {,                 probATAC <- .getDensity(embeddingATAC[gLi$ATAC, ,                   1], embeddingATAC[gLi$ATAC, 2])$density,                 probATAC <- probATAC/max(probATAC),                 cellsATAC <- gLi$ATAC[order(probATAC, decreasing = TRUE)],             },             else {,                 cellsATAC <- sample(gLi$ATAC, length(gLi$ATAC)),             },             cutoffs <- lapply(seq_len(1000), function(x) length(gLi$ATAC)/x) %>% ,                 unlist,             blockSize <- ceiling(min(cutoffs[order(abs(cutoffs - ,                 sampleCellsATAC))[1]] + 1, length(gLi$ATAC))),             nBlocks <- ceiling(length(gLi$ATAC)/blockSize),             blocks <- lapply(seq_len(nBlocks), function(x) {,                 cellsATAC[seq(x, length(cellsATAC), nBlocks)],             }) %>% SimpleList,         },         else {,             blocks <- list(gLi$ATAC),         },         if (!is.null(embeddingRNA)) {,             probRNA <- .getDensity(embeddingRNA[gLi$RNA, 1], ,                 embeddingRNA[gLi$RNA, 2])$density,             probRNA <- probRNA/max(probRNA),         },         else {,             probRNA <- rep(1, length(gLi$RNA)),         },         blockListi <- lapply(seq_along(blocks), function(x) {,             SimpleList(ATAC = blocks[[x]], RNA = sample(x = gLi$RNA, ,                 size = min(sampleCellsRNA, length(gLi$RNA)), ,                 prob = probRNA)),         }) %>% SimpleList,         blockList <- c(blockList, blockListi),     },     rm(groupList),     .logDiffTime(\"Prepping Interation Data\", tstart, verbose = verbose, ,         logFile = logFile),     subProj <- ArchRProj,     subProj@imputeWeights <- SimpleList(),     geneDF <- .getFeatureDF(getArrowFiles(subProj), useMatrix),     geneDF <- geneDF[geneDF$name %in% rownames(seuratRNA), , ,         drop = FALSE],     splitGeneDF <- S4Vectors::split(geneDF, geneDF$seqnames),     featureDF <- lapply(splitGeneDF, function(x) {,         x$idx <- seq_len(nrow(x)),         return(x),     }) %>% Reduce(\"rbind\", .),     dfParams <- data.frame(reduction = reduction),     allChr <- unique(featureDF$seqnames),     tmpFile <- .tempfile(),     o <- suppressWarnings(file.remove(paste0(tmpFile, \"-IntegrationBlock-\", ,         seq_along(blockList), \".h5\"))),     if (threads > 1) {,         h5disableFileLocking(),     },     rD <- getReducedDims(ArchRProj = ArchRProj, reducedDims = reducedDims, ,         corCutOff = corCutOff, dimsToUse = dimsToUse),     outDir1 <- getOutputDirectory(ArchRProj),     outDir2 <- file.path(outDir1, \"RNAIntegration\"),     outDir3 <- file.path(outDir2, matrixName),     dir.create(outDir1, showWarnings = FALSE),     dir.create(outDir2, showWarnings = FALSE),     dir.create(outDir3, showWarnings = FALSE),     prevFiles <- list.files(outDir3, full.names = TRUE),     prevFiles <- .suppressAll(file.remove(prevFiles)),     tstart <- Sys.time(),     threads2 <- max(ceiling(threads * 0.75), 1),     .logDiffTime(paste0(\"Computing Integration in \", length(blockList), ,         \" Integration Blocks!\"), tstart, verbose = verbose, logFile = logFile),     dfAll <- .safelapply(seq_along(blockList), function(i) {,         prefix <- sprintf(\"Block (%s of %s) :\", i, length(blockList)),         .logDiffTime(sprintf(\"%s Computing Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         blocki <- blockList[[i]],         subProj@cellColData <- subProj@cellColData[blocki$ATAC, ,             ],         subProj@sampleColData <- subProj@sampleColData[unique(subProj$Sample), ,             , drop = FALSE],         subRNA <- seuratRNA[, blocki$RNA],         subRNA <- subRNA[rownames(subRNA) %in% geneDF$name, ],         .logDiffTime(sprintf(\"%s Identifying Variable Genes\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         subRNA <- FindVariableFeatures(object = subRNA, nfeatures = nGenes, ,             verbose = FALSE),         subRNA <- ScaleData(object = subRNA, verbose = FALSE),         if (is.null(genesUse)) {,             genesUse <- VariableFeatures(object = subRNA),         },         .logDiffTime(sprintf(\"%s Getting GeneScoreMatrix\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         mat <- .getPartialMatrix(getArrowFiles(subProj), featureDF = geneDF[geneDF$name %in% ,             genesUse, ], threads = 1, cellNames = subProj$cellNames, ,             useMatrix = useMatrix, verbose = FALSE),         rownames(mat) <- geneDF[geneDF$name %in% genesUse, \"name\"],         .logThis(mat, paste0(\"GeneScoreMat-Block-\", i), logFile = logFile),         if (useImputation) {,             .logDiffTime(sprintf(\"%s Imputing GeneScoreMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             imputeParams <- list(),             imputeParams$ArchRProj <- subProj,             imputeParams$randomSuffix <- TRUE,             imputeParams$reducedDims <- reducedDims,             imputeParams$dimsToUse <- dimsToUse,             imputeParams$scaleDims <- scaleDims,             imputeParams$corCutOff <- corCutOff,             imputeParams$threads <- 1,             imputeParams$logFile <- logFile,             subProj <- suppressMessages(do.call(addImputeWeights, ,                 imputeParams)),             mat <- suppressMessages(imputeMatrix(mat = mat, imputeWeights = getImputeWeights(subProj), ,                 verbose = FALSE, logFile = logFile)),             o <- suppressWarnings(file.remove(unlist(getImputeWeights(subProj)[[1]]))),             .logThis(mat, paste0(\"GeneScoreMat-Block-Impute-\", ,                 i), logFile = logFile),         },         mat <- log(mat + 1),         seuratATAC <- Seurat::CreateSeuratObject(counts = mat[head(seq_len(nrow(mat)), ,             5), , drop = FALSE]),         seuratATAC[[\"GeneScore\"]] <- Seurat::CreateAssayObject(counts = mat),         rm(mat),         DefaultAssay(seuratATAC) <- \"GeneScore\",         seuratATAC <- Seurat::ScaleData(seuratATAC, verbose = FALSE),         .logDiffTime(sprintf(\"%s Seurat FindTransferAnchors\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferAnchors <- .retryCatch({,             gc(),             Seurat::FindTransferAnchors(reference = subRNA, query = seuratATAC, ,                 reduction = reduction, features = genesUse, verbose = FALSE, ,                 ...),         }, maxAttempts = 2, logFile = logFile),         .logThis(paste0(utils::capture.output(transferAnchors), ,             collapse = \"\\\\n\"), paste0(\"transferAnchors-\", i), ,             logFile = logFile),         rDSub <- rD[colnames(seuratATAC), , drop = FALSE],         .logThis(rDSub, paste0(\"rDSub-\", i), logFile = logFile),         transferParams$anchorset <- transferAnchors,         transferParams$weight.reduction <- CreateDimReducObject(embeddings = rDSub, ,             key = \"LSI_\", assay = DefaultAssay(seuratATAC)),         transferParams$verbose <- FALSE,         transferParams$dims <- seq_len(ncol(rDSub)),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Group Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata <- subRNA$Group,         rnaLabels <- do.call(Seurat::TransferData, transferParams),         .logDiffTime(sprintf(\"%s Seurat TransferData Cell Names Labels\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         transferParams$refdata <- colnames(subRNA),         rnaLabels2 <- do.call(Seurat::TransferData, transferParams)[, ,             1],         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Seurat TransferData GeneMatrix\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             transferParams$refdata <- GetAssayData(subRNA, assay = \"RNA\", ,                 slot = \"data\"),             gc(),             matchedRNA <- do.call(Seurat::TransferData, transferParams),             matchedRNA <- matchedRNA@data,         },         matchDF <- DataFrame(cellNames = colnames(seuratATAC), ,             predictionScore = rnaLabels$prediction.score.max, ,             predictedGroup = rnaLabels$predicted.id, predictedCell = rnaLabels2),         rownames(matchDF) <- matchDF$cellNames,         .logDiffTime(sprintf(\"%s Saving TransferAnchors Joint CCA\", ,             prefix), tstart, verbose = verbose, logFile = logFile),         jointCCA <- DataFrame(transferAnchors@object.list[[1]]@reductions$cca@cell.embeddings),         jointCCA$Assay <- ifelse(endsWith(rownames(jointCCA), ,             \"_reference\"), \"RNA\", \"ATAC\"),         jointCCA$Group <- NA,         jointCCA$Score <- NA,         jointCCA[paste0(colnames(subRNA), \"_reference\"), \"Group\"] <- subRNA$Group,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Group\"] <- matchDF$predictedGroup,         jointCCA[paste0(matchDF$cellNames, \"_query\"), \"Score\"] <- matchDF$predictionScore,         .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,             paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),         rm(transferParams, transferAnchors),         gc(),         if (addToArrow) {,             .logDiffTime(sprintf(\"%s Transferring Paired RNA to Temp File\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             tmpFilei <- paste0(tmpFile, \"-IntegrationBlock-\", ,                 i, \".h5\"),             o <- h5createFile(tmpFilei),             sampleNames <- getCellColData(subProj, \"Sample\")[matchDF$cellNames, ,                 ],             uniqueSamples <- unique(sampleNames),             matchedRNA <- .safeSubset(mat = matchedRNA, subsetRows = paste0(featureDF$name), ,                 subsetCols = matchDF$cellNames),             for (z in seq_along(uniqueSamples)) {,                 mat <- matchedRNA[, which(sampleNames == uniqueSamples[z]), ,                   drop = FALSE],                 Group <- uniqueSamples[z],                 o <- tryCatch({,                   h5delete(tmpFilei, paste0(Group)),                 }, error = function(x) {,                 }),                 o <- h5createGroup(tmpFilei, paste0(Group)),                 j <- Rle(findInterval(seq(mat@x) - 1, mat@p[-1]) + ,                   1),                 lengthRle <- length(j@lengths),                 lengthI <- length(mat@i),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/i\"), storage.mode = \"integer\", dims = c(lengthI, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jLengths\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/jValues\"), storage.mode = \"integer\", dims = c(lengthRle, ,                   1), level = 0)),                 o <- .suppressAll(h5createDataset(tmpFilei, paste0(Group, ,                   \"/x\"), storage.mode = \"double\", dims = c(lengthI, ,                   1), level = 0)),                 o <- .suppressAll(h5write(obj = mat@i + 1, file = tmpFilei, ,                   name = paste0(Group, \"/i\"))),                 o <- .suppressAll(h5write(obj = j@lengths, file = tmpFilei, ,                   name = paste0(Group, \"/jLengths\"))),                 o <- .suppressAll(h5write(obj = j@values, file = tmpFilei, ,                   name = paste0(Group, \"/jValues\"))),                 o <- .suppressAll(h5write(obj = mat@x, file = tmpFilei, ,                   name = paste0(Group, \"/x\"))),                 o <- .suppressAll(h5write(obj = colnames(mat), ,                   file = tmpFilei, name = paste0(Group, \"/cellNames\"))),             },             rm(matchedRNA, mat, j),         },         .logDiffTime(sprintf(\"%s Completed Integration\", prefix), ,             tstart, verbose = verbose, logFile = logFile),         gc(),         matchDF$Block <- Rle(i),         matchDF,     }, threads = threads2) %>% Reduce(\"rbind\", .),     if (plotUMAP) {,         for (i in seq_along(blockList)) {,             o <- tryCatch({,                 prefix <- sprintf(\"Block (%s of %s) :\", i, length(blockList)),                 .logDiffTime(sprintf(\"%s Plotting Joint UMAP\", ,                   prefix), tstart, verbose = verbose, logFile = logFile),                 jointCCA <- readRDS(file.path(outDir3, paste0(\"Save-Block\", ,                   i, \"-JointCCA.rds\"))),                 set.seed(1),                 UMAPParams <- .mergeParams(UMAPParams, list(n_neighbors = 40, ,                   min_dist = 0.4, metric = \"cosine\", verbose = FALSE)),                 UMAPParams$X <- as.data.frame(jointCCA[, grep(\"CC_\", ,                   colnames(jointCCA))]),                 UMAPParams$ret_nn <- FALSE,                 UMAPParams$ret_model <- FALSE,                 UMAPParams$n_threads <- 1,                 uwotUmap <- tryCatch({,                   do.call(uwot::umap, UMAPParams),                 }, error = function(e) {,                   errorList <- UMAPParams,                   .logError(e, fn = \"uwot::umap\", info = prefix, ,                     errorList = errorList, logFile = logFile),                 }),                 jointCCA$UMAP1 <- uwotUmap[, 1],                 jointCCA$UMAP2 <- uwotUmap[, 2],                 .safeSaveRDS(object = jointCCA, file = file.path(outDir3, ,                   paste0(\"Save-Block\", i, \"-JointCCA.rds\"))),                 p1 <- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Assay, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by Assay\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 p2 <- ggPoint(x = uwotUmap[, 1], y = uwotUmap[, ,                   2], color = jointCCA$Group, randomize = TRUE, ,                   size = 0.2, title = paste0(prefix, \" colored by scRNA Group\"), ,                   xlabel = \"UMAP Dimension 1\", ylabel = \"UMAP Dimension 2\", ,                   rastr = TRUE) + theme(axis.text.x = element_blank(), ,                   axis.ticks.x = element_blank(), axis.text.y = element_blank(), ,                   axis.ticks.y = element_blank()),                 pdf(file.path(outDir3, paste0(\"Save-Block\", i, ,                   \"-JointCCA-UMAP.pdf\")), width = 12, height = 6, ,                   useDingbats = FALSE),                 ggAlignPlots(p1, p2, type = \"h\"),                 dev.off(),             }, error = function(e) {,             }),         },     },     if (addToArrow) {,         .logDiffTime(\"Transferring Data to ArrowFiles\", tstart, ,             verbose = verbose, logFile = logFile),         matrixName <- .isProtectedArray(matrixName),         integrationFiles <- paste0(tmpFile, \"-IntegrationBlock-\", ,             seq_along(blockList), \".h5\"),         if (!all(file.exists(integrationFiles))) {,             .logMessage(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\", ,                 logFile = logFile),             stop(\"Something went wrong with integration as not all temporary files containing integrated RNA exist!\"),         },         h5list <- .safelapply(seq_along(integrationFiles), function(x) {,             h5ls(integrationFiles[x]),         }, threads = threads),         ArrowFiles <- getArrowFiles(ArchRProj),         allSamples <- names(ArrowFiles),         o <- .safelapply(seq_along(allSamples), function(y) {,             sample <- allSamples[y],             prefix <- sprintf(\"%s (%s of %s)\", sample, y, length(ArrowFiles)),             .logDiffTime(sprintf(\"%s Getting GeneIntegrationMatrix From TempFiles!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             sampleIF <- lapply(seq_along(h5list), function(x) {,                 if (any(h5list[[x]]$group == paste0(\"/\", sample))) {,                   integrationFiles[x],                 },                 else {,                   NULL,                 },             }) %>% unlist,             sampleMat <- lapply(seq_along(sampleIF), function(x) {,                 cellNames <- .h5read(sampleIF[x], paste0(sample, ,                   \"/cellNames\")),                 mat <- sparseMatrix(i = .h5read(sampleIF[x], ,                   paste0(sample, \"/i\"))[, 1], j = as.vector(Rle(.h5read(sampleIF[x], ,                   paste0(sample, \"/jValues\"))[, 1], .h5read(sampleIF[x], ,                   paste0(sample, \"/jLengths\"))[, 1])), x = .h5read(sampleIF[x], ,                   paste0(sample, \"/x\"))[, 1], dims = c(nrow(featureDF), ,                   length(cellNames))),                 colnames(mat) <- cellNames,                 mat,             }) %>% Reduce(\"cbind\", .),             sampleMat@x <- exp(sampleMat@x) - 1,             sampleMat <- .normalizeCols(sampleMat, scaleTo = scaleTo),             sampleMat <- drop0(sampleMat),             rownames(sampleMat) <- paste0(featureDF$name),             sampleMat <- sampleMat[, ArchRProj$cellNames[BiocGenerics::which(ArchRProj$Sample == ,                 sample)], drop = FALSE],             o <- .createArrowGroup(ArrowFile = ArrowFiles[sample], ,                 group = matrixName, force = force),             o <- .initializeMat(ArrowFile = ArrowFiles[sample], ,                 Group = matrixName, Class = \"double\", Units = \"NormCounts\", ,                 cellNames = colnames(sampleMat), params = dfParams, ,                 featureDF = featureDF, force = force),             o <- h5write(obj = dfAll[colnames(sampleMat), \"predictionScore\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictionScore\")),             o <- h5write(obj = dfAll[colnames(sampleMat), \"predictedGroup\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedGroup\")),             o <- h5write(obj = dfAll[colnames(sampleMat), \"predictedCell\"], ,                 file = ArrowFiles[sample], name = paste0(matrixName, ,                   \"/Info/predictedCell\")),             .logDiffTime(sprintf(\"%s Adding GeneIntegrationMatrix to ArrowFile!\", ,                 prefix), tstart, verbose = verbose, logFile = logFile),             for (z in seq_along(allChr)) {,                 chrz <- allChr[z],                 .logDiffTime(sprintf(\"Adding GeneIntegrationMatrix to %s for Chr (%s of %s)!\", ,                   sample, z, length(allChr)), tstart, verbose = FALSE, ,                   logFile = logFile),                 idz <- BiocGenerics::which(featureDF$seqnames %bcin% ,                   chrz),                 matz <- sampleMat[idz, , drop = FALSE],                 stopifnot(identical(paste0(featureDF$name[idz]), ,                   paste0(rownames(matz)))),                 o <- .addMatToArrow(mat = matz, ArrowFile = ArrowFiles[sample], ,                   Group = paste0(matrixName, \"/\", chrz), binarize = FALSE, ,                   addColSums = TRUE, addRowSums = TRUE, addRowVarsLog2 = TRUE, ,                   logFile = logFile),                 rm(matz),                 if (z%%3 == 0 | z == length(allChr)) {,                   gc(),                 },             },             0,         }, threads = threads),         o <- suppressWarnings(file.remove(integrationFiles)),     },     .logDiffTime(\"Completed Integration with RNA Matrix\", tstart, ,         verbose = verbose, logFile = logFile),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedCell, name = nameCell, force = TRUE),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictedGroup, name = nameGroup, force = TRUE),     ArchRProj <- addCellColData(ArchRProj = ArchRProj, cells = dfAll$cellNames, ,         data = dfAll$predictionScore, name = nameScore, force = TRUE),     .endLogging(logFile = logFile),     return(ArchRProj), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/Trajectory.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    \n",
       "value   function (ArchRProj = NULL, useSeqnames = NULL, verbose = TRUE, ,     binarize = FALSE, threads = getArchRThreads(), logFile = createLogFile(\"exportMatrixForSTREAM\")) , {,     mat <- getMatrixFromProject(ArchRProj = ArchRProj, useMatrix = \"PeakMatrix\", ,         useSeqnames = useSeqnames, verbose = verbose, binarize = binarize, ,         threads = threads, logFile = logFile),     featureDF <- ArchR:::.getFeatureDF(getArrowFiles(ArchRProj)[1], ,         \"PeakMatrix\"),     stopifnot(all(featureDF$idx == rowData(mat)$idx)),     countsDF <- Matrix::summary(assay(mat)),     peaksDF <- data.frame(as.vector(featureDF[, 1]), featureDF[, ,         3], featureDF[, 4]),     cellsDF <- data.frame(colnames(mat)),     data.table::fwrite(countsDF, file = \"STREAM_Counts.tsv.gz\", ,         sep = \"\\\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(peaksDF, file = \"STREAM_Regions.tsv.gz\", ,         sep = \"\\\\t\", row.names = FALSE, col.names = FALSE),     data.table::fwrite(cellsDF, file = \"STREAM_Sample.tsv.gz\", ,         sep = \"\\\\t\", row.names = FALSE, col.names = FALSE),     return(0), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/ValidationUtils.R                                                                                    \n",
       "value   function (ArchRProj = NULL) , {,     if (!inherits(ArchRProj, \"ArchRProject\")) {,         stop(\"Not a valid ArchRProject as input!\"),     },     else {,         ArchRProj,     }, }\n",
       "visible FALSE                                                                                                                                                                               \n",
       "        /gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R/VisualizeData.R                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  \n",
       "value   function (p = NULL, plotWidth = unit(6, \"in\"), plotHeight = unit(6, ,     \"in\"), margin = 0.25, height = 1, it = 0.05, newPage = FALSE) , {,     .requirePackage(\"grid\", source = \"cran\"),     .requirePackage(\"gridExtra\", source = \"cran\"),     if (!inherits(plotWidth, \"unit\")) {,         plotWidth <- unit(plotWidth, \"in\"),     },     if (!inherits(plotHeight, \"unit\")) {,         plotHeight <- unit(plotHeight, \"in\"),     },     g <- ggplotGrob(p),     legend <- grep(\"guide-box\", g$layout$name),     if (length(legend) != 0) {,         gl <- g$grobs[[legend]],         g <- ggplotGrob(p + theme(legend.position = \"none\")),     },     else {,         gl <- NULL,         g <- ggplotGrob(p),     },     panels <- grep(\"panel\", g$layout$name),     panel_index_w <- unique(g$layout$l[panels]),     panel_index_h <- unique(g$layout$t[panels]),     nw <- length(panel_index_w),     nh <- length(panel_index_h),     pw <- convertWidth(plotWidth, unitTo = \"in\", valueOnly = TRUE),     ph <- convertWidth(plotHeight, unitTo = \"in\", valueOnly = TRUE),     pw <- pw * 0.95,     ph <- ph * 0.95,     x <- 0,     width <- 1,     sm <- FALSE,     while (!sm) {,         x <- x + it,         w <- unit(x * width, \"in\"),         h <- unit(x * height/width, \"in\"),         m <- unit(x * margin/width, \"in\"),         g$widths[panel_index_w] <- rep(w, nw),         g$heights[panel_index_h] <- rep(h, nh),         sw <- convertWidth(x = sum(g$widths) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sh <- convertHeight(x = sum(g$heights) + m, unitTo = \"in\", ,             valueOnly = TRUE),         sm <- sw > pw | sh > ph,     },     if (length(legend) != 0) {,         sgh <- convertHeight(x = sum(g$heights), unitTo = \"in\", ,             valueOnly = TRUE),         sgw <- convertWidth(x = sum(g$widths), unitTo = \"in\", ,             valueOnly = TRUE),         slh <- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,             valueOnly = TRUE),         slw <- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,             valueOnly = TRUE),         size <- 6,         wh <- 0.1,         it <- 0,         while (slh > 0.2 * ph | slw > pw) {,             it <- it + 1,             if (it > 3) {,                 break,             },             size <- size * 0.8,             wh <- wh * 0.8,             gl <- ggplotGrob(p + theme(legend.key.width = unit(wh, ,                 \"cm\"), legend.key.height = unit(wh, \"cm\"), legend.spacing.x = unit(0, ,                 \"cm\"), legend.spacing.y = unit(0, \"cm\"), legend.text = element_text(size = max(size, ,                 2))) + guides(fill = guide_legend(ncol = 4), ,                 color = guide_legend(ncol = 4)))$grobs[[legend]],             slh <- convertHeight(x = sum(gl$heights), unitTo = \"in\", ,                 valueOnly = TRUE),             slw <- convertWidth(x = sum(gl$widths), unitTo = \"in\", ,                 valueOnly = TRUE),         },         p <- grid.arrange(g, gl, ncol = 1, nrow = 2, heights = unit.c(unit(sgh, ,             \"in\"), unit(min(slh, 0.2 * pw), \"in\")), newpage = newPage),     },     else {,         p <- grid.arrange(g, newpage = newPage),     },     invisible(p), }\n",
       "visible FALSE                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sapply(list.files('/gpfs/gibbs/pi/gerstein/jz435/ShareZhangLab/PTSD/misc/girgenti_project/ArchR/R',full.name=T),source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5c8dcd3-e29e-4c20-9c2f-c49c2f767d35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "source('../utils/ArchRBrowser-Fig1.R')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "358e85ab-9925-4dcc-a9e7-b1d7d711f7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "SLC17A7 chr19:49428401-49443360\n",
    "GAD2 chr10:26214307-26306558\n",
    "MOG chr6:29655981-29673372\n",
    "OLIG1 chr21:33068144-33074420\n",
    "CLDN5 chr22:19522024-19528545\n",
    "GJA1 chr6:121434692-121450727\n",
    "CX3CR1 chr3:39262494-39282735"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79caa6bf-04a7-4159-a4d0-2421600d8df2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes <- data.frame(getGenes(proj))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93449133-aa31-4c88-9d7b-364fab3971f3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      seqnames    start      end width strand gene_id  symbol\n",
      "21999    chr19 49429401 49442360 12960      -   57030 SLC17A7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2f5ee45242d138-Date-2023-11-20_Time-14-05-52.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-11-20 14:05:52 : Validating Region, 0.004 mins elapsed.\n",
      "\n",
      "2023-11-20 14:05:52 : Adding Bulk Tracks (1 of 1), 0.004 mins elapsed.\n",
      "\n",
      "2023-11-20 14:08:07 : Adding Gene Tracks (1 of 1), 2.261 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand != :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand == :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-11-20 14:08:08 : Plotting, 2.269 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2f5ee45242d138-Date-2023-11-20_Time-14-05-52.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='SLC17A7',] \n",
    "df <- na.omit(df)\n",
    "print(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-1000,end=df$end+1000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.2),\n",
    "    tileSize=250,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "0412585f-b8b3-4016-a195-dce1d6ba4bf6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"SLC17A7_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "bd16297d-ec71-48d5-bd2f-4a9bf7af3d33",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2beed12911fbf-Date-2023-07-21_Time-15-46-40.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-07-21 15:46:40 : Validating Region, 0.005 mins elapsed.\n",
      "\n",
      "2023-07-21 15:46:40 : Adding Bulk Tracks (1 of 1), 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 15:49:06 : Adding Gene Tracks (1 of 1), 2.437 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand != :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-07-21 15:49:06 : Plotting, 2.443 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2beed12911fbf-Date-2023-07-21_Time-15-46-40.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='GAD2',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-2000,end=df$end+2000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.3),\n",
    "    tileSize=750,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "cfebc618-ac50-44a2-9501-fea8cff2a276",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"GAD2_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3eef4d1e-edc8-4743-bc95-c473454b408b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2beed1cd45543-Date-2023-07-21_Time-11-33-51.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-07-21 11:33:51 : Validating Region, 0.005 mins elapsed.\n",
      "\n",
      "2023-07-21 11:33:51 : Adding Bulk Tracks (1 of 1), 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 11:36:21 : Adding Gene Tracks (1 of 1), 2.493 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand != :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand == :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-07-21 11:36:21 : Plotting, 2.499 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2beed1cd45543-Date-2023-07-21_Time-11-33-51.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='MOG',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-1000,end=df$end+1000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.5),\n",
    "    tileSize=250,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "771818cf-dedc-4bb5-948f-ab45d9c7678f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"MOG_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "91494d65-1956-4859-90ab-c6584af9490a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-33ad105ddccf9d-Date-2023-08-31_Time-13-08-22.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-08-31 13:08:23 : Validating Region, 0.006 mins elapsed.\n",
      "\n",
      "2023-08-31 13:08:23 : Adding Bulk Tracks (1 of 1), 0.006 mins elapsed.\n",
      "\n",
      "2023-08-31 13:10:53 : Adding Gene Tracks (1 of 1), 2.516 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand != :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand == :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-08-31 13:10:54 : Plotting, 2.523 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-33ad105ddccf9d-Date-2023-08-31_Time-13-08-22.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='OLIG1',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start - 2000,end=df$end + 2000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.3),\n",
    "    tileSize=250,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d1eed58f-f59c-4911-b227-f5436388f30c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"PDGFRA_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "aaa4a2ed-e10b-4b5c-8f71-2eccd5f27d3a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2beed139431da0-Date-2023-07-21_Time-15-50-32.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-07-21 15:50:33 : Validating Region, 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 15:50:33 : Adding Bulk Tracks (1 of 1), 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 15:52:50 : Adding Gene Tracks (1 of 1), 2.295 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand == :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-07-21 15:52:50 : Plotting, 2.301 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2beed139431da0-Date-2023-07-21_Time-15-50-32.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='CLDN5',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-1000,end=df$end+1000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.2),\n",
    "    tileSize=200,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "32b21d89-3169-476f-aa61-edbdfd0eeedf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"CLDN5_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5d221cc3-f77b-4475-9c76-a78987f90079",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2beed160fbe509-Date-2023-07-21_Time-11-51-13.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-07-21 11:51:13 : Validating Region, 0.008 mins elapsed.\n",
      "\n",
      "2023-07-21 11:51:13 : Adding Bulk Tracks (1 of 1), 0.008 mins elapsed.\n",
      "\n",
      "2023-07-21 11:53:43 : Adding Gene Tracks (1 of 1), 2.493 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand != :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-07-21 11:53:43 : Plotting, 2.499 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2beed160fbe509-Date-2023-07-21_Time-11-51-13.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='GJA1',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-1000,end=df$end+1000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.3),\n",
    "    tileSize=250,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "da1e3aa4-9429-4154-9933-51ee4785c9f1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"GJA1_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6d9e0191-ab23-492c-a3e4-9e18fde3f82b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-2beed12c17ae0e-Date-2023-07-21_Time-12-14-53.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2023-07-21 12:14:54 : Validating Region, 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 12:14:54 : Adding Bulk Tracks (1 of 1), 0.006 mins elapsed.\n",
      "\n",
      "2023-07-21 12:17:25 : Adding Gene Tracks (1 of 1), 2.532 mins elapsed.\n",
      "\n",
      "Warning message in ggrepel::geom_text_repel(data = genesO[which(genesO$strand == :\n",
      "“\u001b[1m\u001b[22mIgnoring unknown aesthetics: \u001b[32mlabel.size\u001b[39m”\n",
      "2023-07-21 12:17:26 : Plotting, 2.537 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-2beed12c17ae0e-Date-2023-07-21_Time-12-14-53.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df <- genes[genes$symbol=='CX3CR1',] \n",
    "df <- na.omit(df)\n",
    "\n",
    "p <- plotBrowserTrack3(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Int_Cluster\", \n",
    "    useGroups = c('EXC','INH','OLI','OPC','END','AST','MIC'),\n",
    "    plotSummary = c(\"bulkTrack\",\"geneTrack\"),\n",
    "    region = GRanges(seqnames = df$seqnames,ranges = IRanges(start=df$start-1000,end=df$end+1000)),\n",
    "    sizes = c(4, 0.5),\n",
    "    ylim2 = c(0, 0.3),\n",
    "    tileSize=250,\n",
    "    pal=c('#b22222','#2E8B57','#5254a3','#aec7e8','#bc80bd','#ffed6f','#7f7f7f')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "eeb55f0e-2e76-4452-8fb4-6b3cd1e06256",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(p, \n",
    "    name = \"CX3CR1_track.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 2.5, height = 4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
